
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.1, mkdocs-material-7.1.10">
    
    
      
        <title>R - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.1fe995fd.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.f1a3b89f.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../css/custom.css">
    
    
      
  



  
  



  <script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
  <script async src="https://www.google-analytics.com/analytics.js"></script>


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    <script>function __prefix(e){return new URL("../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../index.html" title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              R
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../index.html" title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../index.html" class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../applications.html" class="md-nav__link">
        Applications
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Using kb
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_usage.html" class="md-nav__link">
        Introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_ref.html" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_count.html" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      <label class="md-nav__link" for="__nav_4">
        Tutorials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_1" type="checkbox" id="__nav_4_1" >
      
      <label class="md-nav__link" for="__nav_4_1">
        Introduction to single-cell RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../scRNA-seq_intro/python/scRNA-seq_intro.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_2" type="checkbox" id="__nav_4_2" >
      
      <label class="md-nav__link" for="__nav_4_2">
        Preprocessing & Quality Control
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/python/kb_intro_1_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/R/kb_intro_1_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_3" type="checkbox" id="__nav_4_3" >
      
      <label class="md-nav__link" for="__nav_4_3">
        Getting started with analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_getting_started/python/kb_intro_2_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_getting_started/R/kb_intro_2_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" >
      
      <label class="md-nav__link" for="__nav_4_4">
        Building & annotating an atlas
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_building_atlas/python/kb_analysis_0_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_building_atlas/R/kb_analysis_0_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      <label class="md-nav__link" for="__nav_4_5">
        Pseudotime analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/slingshot/kb_slingshot.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      <label class="md-nav__link" for="__nav_4_6">
        Single-nuclei RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_nucleus/python/kb_single_nucleus.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      <label class="md-nav__link" for="__nav_4_7">
        RNA Velocity
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity/python/kb_velocity.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" >
      
      <label class="md-nav__link" for="__nav_4_8">
        Feature barcode & multimodal data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_kite/python/kb_kite.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../download_data/data_download.html" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      <label class="md-nav__link" for="__nav_4_10">
        Build custom index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_custom_index/python/kb_transcriptome_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_11" type="checkbox" id="__nav_4_11" >
      
      <label class="md-nav__link" for="__nav_4_11">
        Build RNA velocity index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity_index/python/kb_velocity_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_12" type="checkbox" id="__nav_4_12" >
      
      <label class="md-nav__link" for="__nav_4_12">
        Aggregate multiple matrices
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_aggregate/python/kb_aggregating_count_matrices.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_13" type="checkbox" id="__nav_4_13" >
      
      <label class="md-nav__link" for="__nav_4_13">
        Process multiple FASTQs
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_multiple_fastqs/python/kb_multiple_files.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_14" type="checkbox" id="__nav_4_14" checked>
      
      <label class="md-nav__link" for="__nav_4_14">
        Processing multi-species data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../python/kb_species_mixing.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          R
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="kb_mixed_species_10x_v2.html" class="md-nav__link md-nav__link--active">
        R
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generate-the-gene-count-matrix" class="md-nav__link">
    Generate the gene count matrix
  </a>
  
    <nav class="md-nav" aria-label="Generate the gene count matrix">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#build-the-kallisto-index" class="md-nav__link">
    Build the kallisto index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#explore-the-data" class="md-nav__link">
    Explore the data
  </a>
  
    <nav class="md-nav" aria-label="Explore the data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#remove-empty-droplets" class="md-nav__link">
    Remove empty droplets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cell-species" class="md-nav__link">
    Cell species
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dimension-reduction" class="md-nav__link">
    Dimension reduction
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_15" type="checkbox" id="__nav_4_15" >
      
      <label class="md-nav__link" for="__nav_4_15">
        Pseudotime analysis (Monocle2)
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/monocle2/kb_monocle2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generate-the-gene-count-matrix" class="md-nav__link">
    Generate the gene count matrix
  </a>
  
    <nav class="md-nav" aria-label="Generate the gene count matrix">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#build-the-kallisto-index" class="md-nav__link">
    Build the kallisto index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#explore-the-data" class="md-nav__link">
    Explore the data
  </a>
  
    <nav class="md-nav" aria-label="Explore the data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#remove-empty-droplets" class="md-nav__link">
    Remove empty droplets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cell-species" class="md-nav__link">
    Cell species
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dimension-reduction" class="md-nav__link">
    Dimension reduction
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/tree/master/tutorials/docs/tutorials/kb_species_mixing/R/kb_mixed_species_10x_v2.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_mixed_species_10x_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<p>In this vignette, we process fastq data from scRNA-seq (10x v2 chemistry) with to make a sparse matrix that can be used in downstream analysis with command line tools <code>kallisto</code> and <code>bustools</code>, as described in the <a href="https://www.biorxiv.org/content/10.1101/673285v1"><code>kallisto bus</code> paper</a>. Then we will start a standard downstream analysis with <code>Seurat</code>. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;pip3 install kb-python&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;wget https://github.com/pachterlab/kallisto/releases/download/v0.46.1/kallisto_linux-v0.46.1.tar.gz&quot;</span><span class="p">)</span>
<span class="nf">untar</span><span class="p">(</span><span class="s">&quot;kallisto_linux-v0.46.1.tar.gz&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;cp kallisto/kallisto /usr/local/bin&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Since <code>Seurat</code> is a large package with many dependencies, it takes a while to install the packages, especially those with compiled code.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;multtest&quot;</span><span class="p">,</span> <span class="s">&quot;DropletUtils&quot;</span><span class="p">,</span> <span class="s">&quot;BUSpaRse&quot;</span><span class="p">),</span> <span class="n">Ncores</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Seurat&quot;</span><span class="p">,</span> <span class="s">&quot;biomartr&quot;</span><span class="p">),</span> <span class="n">Ncores</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Installing</span> <span class="n">package</span> <span class="n">into</span> <span class="err">‘</span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">library</span><span class="err">’</span>
<span class="p">(</span><span class="k">as</span> <span class="err">‘</span><span class="n">lib</span><span class="err">’</span> <span class="k">is</span> <span class="n">unspecified</span><span class="p">)</span>

<span class="n">Bioconductor</span> <span class="n">version</span> <span class="mf">3.10</span> <span class="p">(</span><span class="n">BiocManager</span> <span class="mf">1.30</span><span class="o">.</span><span class="mi">10</span><span class="p">),</span> <span class="n">R</span> <span class="mf">3.6</span><span class="o">.</span><span class="mi">2</span> <span class="p">(</span><span class="mi">2019</span><span class="o">-</span><span class="mi">12</span><span class="o">-</span><span class="mi">12</span><span class="p">)</span>

<span class="n">Installing</span> <span class="n">package</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="s1">&#39;BiocVersion&#39;</span><span class="p">,</span> <span class="s1">&#39;multtest&#39;</span><span class="p">,</span> <span class="s1">&#39;DropletUtils&#39;</span><span class="p">,</span> <span class="s1">&#39;BUSpaRse&#39;</span>

<span class="n">also</span> <span class="n">installing</span> <span class="n">the</span> <span class="n">dependencies</span> <span class="err">‘</span><span class="n">formatR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bit</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">vctrs</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lambda</span><span class="o">.</span><span class="n">r</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">futile</span><span class="o">.</span><span class="n">options</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">matrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bit64</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">blob</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DBI</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">zlibbioc</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bitops</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rhtslib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SummarizedExperiment</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">futile</span><span class="o">.</span><span class="n">logger</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">snow</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">limma</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">locfit</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DelayedArray</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">IRanges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">oo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">methodsS3</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sitmo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RSQLite</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">XML</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocFileCache</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rappdirs</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">XVector</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rtracklayer</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rsamtools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ProtGenerics</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RCurl</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomeInfoDbData</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicAlignments</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocGenerics</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Biobase</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SingleCellExperiment</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">S4Vectors</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocParallel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">edgeR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rhdf5</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">HDF5Array</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">utils</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">dqrng</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">beachmat</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rhdf5lib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">AnnotationDbi</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">AnnotationFilter</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">biomaRt</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Biostrings</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BSgenome</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ensembldb</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomeInfoDb</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicFeatures</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicRanges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plyranges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppParallel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppArmadillo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppProgress</span><span class="err">’</span>


<span class="n">Old</span> <span class="n">packages</span><span class="p">:</span> <span class="s1">&#39;BH&#39;</span><span class="p">,</span> <span class="s1">&#39;broom&#39;</span><span class="p">,</span> <span class="s1">&#39;callr&#39;</span><span class="p">,</span> <span class="s1">&#39;cli&#39;</span><span class="p">,</span> <span class="s1">&#39;curl&#39;</span><span class="p">,</span> <span class="s1">&#39;DBI&#39;</span><span class="p">,</span> <span class="s1">&#39;DT&#39;</span><span class="p">,</span> <span class="s1">&#39;fansi&#39;</span><span class="p">,</span>
  <span class="s1">&#39;farver&#39;</span><span class="p">,</span> <span class="s1">&#39;gh&#39;</span><span class="p">,</span> <span class="s1">&#39;hms&#39;</span><span class="p">,</span> <span class="s1">&#39;knitr&#39;</span><span class="p">,</span> <span class="s1">&#39;mime&#39;</span><span class="p">,</span> <span class="s1">&#39;pillar&#39;</span><span class="p">,</span> <span class="s1">&#39;prettyunits&#39;</span><span class="p">,</span> <span class="s1">&#39;rlang&#39;</span><span class="p">,</span>
  <span class="s1">&#39;rmarkdown&#39;</span><span class="p">,</span> <span class="s1">&#39;rprojroot&#39;</span><span class="p">,</span> <span class="s1">&#39;stringi&#39;</span><span class="p">,</span> <span class="s1">&#39;tidyr&#39;</span><span class="p">,</span> <span class="s1">&#39;tinytex&#39;</span><span class="p">,</span> <span class="s1">&#39;vctrs&#39;</span><span class="p">,</span> <span class="s1">&#39;xfun&#39;</span><span class="p">,</span>
  <span class="s1">&#39;xtable&#39;</span><span class="p">,</span> <span class="s1">&#39;boot&#39;</span><span class="p">,</span> <span class="s1">&#39;foreign&#39;</span><span class="p">,</span> <span class="s1">&#39;MASS&#39;</span>

<span class="n">Installing</span> <span class="n">package</span> <span class="n">into</span> <span class="err">‘</span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">library</span><span class="err">’</span>
<span class="p">(</span><span class="k">as</span> <span class="err">‘</span><span class="n">lib</span><span class="err">’</span> <span class="k">is</span> <span class="n">unspecified</span><span class="p">)</span>

<span class="n">also</span> <span class="n">installing</span> <span class="n">the</span> <span class="n">dependencies</span> <span class="err">‘</span><span class="n">mnormt</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">numDeriv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">TH</span><span class="o">.</span><span class="n">data</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sandwich</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lsei</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bibtex</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gbRd</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sn</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">mvtnorm</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plotrix</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">multcomp</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gtools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gdata</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">caTools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">npsurv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">globals</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">listenv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">zoo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rdpack</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">TFisher</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">mutoss</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">hexbin</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gplots</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gridExtra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">FNN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RSpectra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ape</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">cowplot</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">fitdistrplus</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">future</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">future</span><span class="o">.</span><span class="n">apply</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggrepel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggridges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ica</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">igraph</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">irlba</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">leiden</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lmtest</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">metap</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">pbapply</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plotly</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">png</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RANN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppAnnoy</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">reticulate</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ROCR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rsvd</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rtsne</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sctransform</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SDMTools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">tsne</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">uwot</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppEigen</span><span class="err">’</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">BUSpaRse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DropletUtils</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">biomartr</span><span class="p">)</span>
<span class="nf">theme_set</span><span class="p">(</span><span class="nf">theme_bw</span><span class="p">())</span>
</code></pre></div>
</td></tr></table>
<h2 id="download-data">Download data<a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h2>
<p>The data set we are using here is 1k 1:1 Mixture of Fresh Frozen Human (HEK293T) and Mouse (NIH3T3) Cells from the 10x website. First, we download the fastq files (6.34 GB).</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;./data&quot;</span><span class="p">))</span> <span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;./data&quot;</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;./data/hgmm_1k_fastqs.tar&quot;</span><span class="p">))</span> <span class="p">{</span>
  <span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;http://cf.10xgenomics.com/samples/cell-exp/2.1.0/hgmm_1k/hgmm_1k_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">destfile</span> <span class="o">=</span> <span class="s">&quot;./data/hgmm_1k_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<p>Then untar this file</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">untar</span><span class="p">(</span><span class="s">&quot;./data/hgmm_1k_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">exdir</span> <span class="o">=</span> <span class="s">&quot;./data&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h1 id="generate-the-gene-count-matrix">Generate the gene count matrix<a class="headerlink" href="#generate-the-gene-count-matrix" title="Permanent link">&para;</a></h1>
<h2 id="build-the-kallisto-index">Build the <code>kallisto</code> index<a class="headerlink" href="#build-the-kallisto-index" title="Permanent link">&para;</a></h2>
<p>Here we use <a href="https://pachterlab.github.io/kallisto/about">kallisto</a> to pseudoalign the reads to the transcriptome and then to create the <code>bus</code> file to be converted to a sparse matrix. The first step is to build an index of the transcriptome. This data set has both human and mouse cells, so we need both human and mouse transcriptomes. The transcriptomes downloaded here are from Ensembl version 99, the most recent version as of writing. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;./reference&quot;</span><span class="p">))</span> <span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;./reference&quot;</span><span class="p">)</span>
<span class="c1"># Human transcriptome</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;./reference/hs_cdna.fa.gz&quot;</span><span class="p">))</span> <span class="p">{</span>
  <span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz&quot;</span><span class="p">,</span> 
  <span class="s">&quot;./reference/hs_cdna.fa.gz&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;wget&quot;</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="p">}</span>
<span class="c1"># Mouse transcriptome</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;./reference/mm_cdna.fa.gz&quot;</span><span class="p">))</span> <span class="p">{</span>
  <span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;ftp://ftp.ensembl.org/pub/release-99/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz&quot;</span><span class="p">,</span> 
  <span class="s">&quot;./reference/mm_cdna.fa.gz&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;wget&quot;</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;./reference/hs_mm_tr_index.idx&quot;</span><span class="p">))</span> <span class="p">{</span>
  <span class="nf">system</span><span class="p">(</span><span class="s">&quot;kallisto index -i ./reference/hs_mm_tr_index.idx ./reference/hs_cdna.fa.gz ./reference/mm_cdna.fa.gz&quot;</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tr2g</span> <span class="o">&lt;-</span> <span class="nf">map_dfr</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;./reference/hs_cdna.fa.gz&quot;</span><span class="p">,</span> <span class="s">&quot;./reference/mm_cdna.fa.gz&quot;</span><span class="p">),</span> <span class="n">tr2g_fasta</span><span class="p">)</span>
<span class="nf">save_tr2g_bustools</span><span class="p">(</span><span class="n">tr2g</span><span class="p">,</span> <span class="s">&quot;./reference/hs_mm_tr2g.tsv&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Reading FASTA file.

Reading FASTA file.
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">fastq_fns</span> <span class="o">&lt;-</span> <span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;./data/fastqs&quot;</span><span class="p">,</span> <span class="n">full.name</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="n">fastq_fns</span> <span class="o">&lt;-</span> <span class="n">fastq_fns</span><span class="p">[</span><span class="nf">str_detect</span><span class="p">(</span><span class="n">fastq_fns</span><span class="p">,</span> <span class="s">&quot;R\\d&quot;</span><span class="p">)]</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">fastq_fns</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>'./data/fastqs/hgmm_1k_S1_L001_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L001_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L002_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L002_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L003_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L003_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L004_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L004_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L005_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L005_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L006_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L006_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L007_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L007_R2_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L008_R1_001.fastq.gz'</li>
    <li>'./data/fastqs/hgmm_1k_S1_L008_R2_001.fastq.gz'</li>
</ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;kb count -i reference/hs_mm_tr_index.idx -g reference/hs_mm_tr2g.tsv -x 10xv2 -o output&quot;</span><span class="p">,</span>
<span class="nf">paste</span><span class="p">(</span><span class="n">fastq_fns</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="p">),</span> <span class="n">intern</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>See what the outputs are</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;./output&quot;</span><span class="p">,</span> <span class="n">recursive</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>'10xv2_whitelist.txt'</li>
    <li>'counts_unfiltered/cells_x_genes.barcodes.txt'</li>
    <li>'counts_unfiltered/cells_x_genes.genes.txt'</li>
    <li>'counts_unfiltered/cells_x_genes.mtx'</li>
    <li>'inspect.json'</li>
    <li>'matrix.ec'</li>
    <li>'output.bus'</li>
    <li>'output.unfiltered.bus'</li>
    <li>'run_info.json'</li>
    <li>'transcripts.txt'</li>
</ol>

<p>Explaining the output:</p>
<ul>
<li><code>matrix.ec</code>: A text file with two columns. The first column is the 0 based index of equivalence classes. The second column is the set of transcripts (denoted by 0 based index based on order of appearance in the transcriptome fasta file) present in the corresponding equivalence class.</li>
<li><code>output.bus</code>: The data represented in <code>bus</code> format. This is a binary file, so don't use something like <code>read.table</code> to read it into R.</li>
<li><code>run_info.json</code>: Information about the call to <code>kallisto bus</code>, including the command used, number and percentage of reads pseudoaligned, version of <code>kallisto</code> used, and etc.</li>
<li><code>transcript.txt</code>: A text file with one column, which is the transcripts present in the data, in the same order as in the transcriptome fasta file.</li>
</ul>
<h1 id="explore-the-data">Explore the data<a class="headerlink" href="#explore-the-data" title="Permanent link">&para;</a></h1>
<p>Now we can load the matrix into R for analysis.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="nf">read_count_output</span><span class="p">(</span><span class="s">&quot;./output/counts_unfiltered&quot;</span><span class="p">,</span>
                             <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;cells_x_genes&quot;</span><span class="p">,</span> <span class="n">tcc</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="remove-empty-droplets">Remove empty droplets<a class="headerlink" href="#remove-empty-droplets" title="Permanent link">&para;</a></h2>
<p>Cool, so now we have the sparse matrix. What does it look like?</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>77202</li>
    <li>362207</li>
</ol>

<p>The number of genes is as expected for two species. There're way more cells than we expect, which is about 1000. So what's going on?</p>
<p>How many UMIs per barcode?</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tot_counts</span> <span class="o">&lt;-</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">tot_counts</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    0.00     1.00     1.00    76.04     8.00 74292.00
</code></pre></div>

<p>The vast majority of "cells" have only a few UMI detected. Those are empty droplets. 10x claims to have cell capture rate of up to 65%, but in practice, depending on how many cells are in fact loaded, the rate can be much lower. A commonly used method to estimate the number of empty droplets is barcode ranking knee and inflection points, as those are often assumed to represent transition between two components of a distribution. While more sophisticated method exist (e.g. see <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#detecting-empty-droplets"><code>emptyDrops</code> in <code>DropletUtils</code></a>), for simplicity, we will use the barcode ranking method here. However, whichever way we go, we don't have the ground truth.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Compute barcode rank</span>
<span class="n">bc_rank</span> <span class="o">&lt;-</span> <span class="nf">barcodeRanks</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#&#39; Knee plot for filtering empty droplets</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; Visualizes the inflection point to filter empty droplets. This function plots </span>
<span class="c1">#&#39; different datasets with a different color. Facets can be added after calling</span>
<span class="c1">#&#39; this function with `facet_*` functions.</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; @param bc_rank A `DataFrame` output from `DropletUtil::barcodeRanks`.</span>
<span class="c1">#&#39; @return A ggplot2 object.</span>
<span class="n">knee_plot</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">knee_plt</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">rank</span> <span class="o">=</span> <span class="n">bc_rank</span><span class="p">[[</span><span class="s">&quot;rank&quot;</span><span class="p">]],</span>
                     <span class="n">total</span> <span class="o">=</span> <span class="n">bc_rank</span><span class="p">[[</span><span class="s">&quot;total&quot;</span><span class="p">]])</span> <span class="o">%&gt;%</span> 
    <span class="nf">distinct</span><span class="p">()</span> <span class="o">%&gt;%</span> 
    <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">total</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span>
  <span class="n">annot</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">inflection</span> <span class="o">=</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)[[</span><span class="s">&quot;inflection&quot;</span><span class="p">]],</span>
                  <span class="n">rank_cutoff</span> <span class="o">=</span> <span class="nf">max</span><span class="p">(</span><span class="n">bc_rank</span><span class="o">$</span><span class="n">rank</span><span class="p">[</span><span class="n">bc_rank</span><span class="o">$</span><span class="n">total</span> <span class="o">&gt;</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)[[</span><span class="s">&quot;inflection&quot;</span><span class="p">]]]))</span>
  <span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">knee_plt</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="n">total</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_line</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">geom_hline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="n">inflection</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_vline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="n">rank_cutoff</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">annotation_logticks</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Rank&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Total UMIs&quot;</span><span class="p">)</span>
  <span class="nf">return</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">knee_plot</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_30_0.png" /></p>
<p>The inflection point looks like a reasonable number of cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Filter the matrix</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[,</span> <span class="n">tot_counts</span> <span class="o">&gt;</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span><span class="o">$</span><span class="n">inflection</span><span class="p">]</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">,]</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>37755</li>
    <li>1095</li>
</ol>

<h2 id="cell-species">Cell species<a class="headerlink" href="#cell-species" title="Permanent link">&para;</a></h2>
<p>How many cells are from humans and how many from mice? The number of cells with mixed species indicates doublet rate.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span>
<span class="normal">9</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">gene_species</span> <span class="o">&lt;-</span> <span class="nf">ifelse</span><span class="p">(</span><span class="nf">str_detect</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">),</span> <span class="s">&quot;^ENSMUSG&quot;</span><span class="p">),</span> <span class="s">&quot;mouse&quot;</span><span class="p">,</span> <span class="s">&quot;human&quot;</span><span class="p">)</span>
<span class="n">mouse_inds</span> <span class="o">&lt;-</span> <span class="n">gene_species</span> <span class="o">==</span> <span class="s">&quot;mouse&quot;</span>
<span class="n">human_inds</span> <span class="o">&lt;-</span> <span class="n">gene_species</span> <span class="o">==</span> <span class="s">&quot;human&quot;</span>
<span class="c1"># mark cells as mouse or human</span>
<span class="n">cell_species</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">n_mouse_umi</span> <span class="o">=</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">[</span><span class="n">mouse_inds</span><span class="p">,]),</span>
                       <span class="n">n_human_umi</span> <span class="o">=</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">[</span><span class="n">human_inds</span><span class="p">,]),</span>
                       <span class="n">tot_umi</span> <span class="o">=</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">),</span>
                       <span class="n">prop_mouse</span> <span class="o">=</span> <span class="n">n_mouse_umi</span> <span class="o">/</span> <span class="n">tot_umi</span><span class="p">,</span>
                       <span class="n">prop_human</span> <span class="o">=</span> <span class="n">n_human_umi</span> <span class="o">/</span> <span class="n">tot_umi</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Classify species based on proportion of UMI, with cutoff of 90%</span>
<span class="n">cell_species</span> <span class="o">&lt;-</span> <span class="n">cell_species</span> <span class="o">%&gt;%</span> 
  <span class="nf">mutate</span><span class="p">(</span><span class="n">species</span> <span class="o">=</span> <span class="nf">case_when</span><span class="p">(</span>
    <span class="n">prop_mouse</span> <span class="o">&gt;</span> <span class="m">0.9</span> <span class="o">~</span> <span class="s">&quot;mouse&quot;</span><span class="p">,</span>
    <span class="n">prop_human</span> <span class="o">&gt;</span> <span class="m">0.9</span> <span class="o">~</span> <span class="s">&quot;human&quot;</span><span class="p">,</span>
    <span class="kc">TRUE</span> <span class="o">~</span> <span class="s">&quot;mixed&quot;</span>
  <span class="p">))</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">cell_species</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">n_human_umi</span><span class="p">,</span> <span class="n">n_mouse_umi</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">species</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_36_0.png" /></p>
<p>Great, looks like the vast majority of cells are not mixed.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">cell_species</span> <span class="o">%&gt;%</span> 
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">count</span><span class="p">(</span><span class="n">species</span><span class="p">)</span> <span class="o">%&gt;%</span> 
  <span class="nf">mutate</span><span class="p">(</span><span class="n">proportion</span> <span class="o">=</span> <span class="n">n</span> <span class="o">/</span> <span class="nf">ncol</span><span class="p">(</span><span class="n">res_mat</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<table>
<caption>A tibble: 3 × 3</caption>
<thead>
    <tr><th scope=col>species</th><th scope=col>n</th><th scope=col>proportion</th></tr>
    <tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
    <tr><td>human</td><td>566</td><td>0.516894977</td></tr>
    <tr><td>mixed</td><td>  3</td><td>0.002739726</td></tr>
    <tr><td>mouse</td><td>526</td><td>0.480365297</td></tr>
</tbody>
</table>

<p>Great, only about 0.3% of cells here are doublets, which is lower than the ~1% 10x lists. Doublet rate tends to be lower when cell concentration is lower. However, doublets can still be formed with cells from the same species, so the number of mixed species "cells" is only a lower bound of doublet rate.</p>
<h2 id="dimension-reduction">Dimension reduction<a class="headerlink" href="#dimension-reduction" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">res_mat</span><span class="p">,</span> <span class="n">min.cells</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span> <span class="o">%&gt;%</span> 
  <span class="nf">SCTransform</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Add species to meta data</span>
<span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">AddMetaData</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">metadata</span> <span class="o">=</span> <span class="n">cell_species</span><span class="o">$</span><span class="n">species</span><span class="p">,</span> <span class="n">col.name</span> <span class="o">=</span> <span class="s">&quot;species&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>See how number of total counts and number of genes expressed are distributed.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">VlnPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span> <span class="s">&quot;nFeature_RNA&quot;</span><span class="p">),</span> <span class="n">group.by</span> <span class="o">=</span> <span class="s">&quot;species&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_43_0.png" /></p>
<p>Another QC plot</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">seu</span><span class="o">@</span><span class="n">meta.data</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">nCount_RNA</span><span class="p">,</span> <span class="n">nFeature_RNA</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">species</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Total UMI counts per cell&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Number of genes detected&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_45_0.png" /></p>
<p>The mixed species doublets do look different from human and mouse cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">RunPCA</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">npcs</span> <span class="o">=</span> <span class="m">30</span><span class="p">)</span>
<span class="nf">ElbowPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">ndims</span> <span class="o">=</span> <span class="m">30</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_47_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">PCAPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">group.by</span> <span class="o">=</span> <span class="s">&quot;species&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_48_0.png" /></p>
<p>The first PC separates species, as expected. Also as expected, the doublets are in between human and mouse cells in this plot.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">RunTSNE</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span> <span class="n">check_duplicates</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">TSNEPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">group.by</span> <span class="o">=</span> <span class="s">&quot;species&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_mixed_species_10x_v2_files/kb_mixed_species_10x_v2_50_0.png" /></p>
<p>The species separate, and the few doublets form its own cluster, as expected.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../python/kb_species_mixing.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Python" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Python
            </div>
          </div>
        </a>
      
      
        
        <a href="../../kb_pseudotime/R/monocle2/kb_monocle2.html" class="md-footer__link md-footer__link--next" aria-label="Next: R" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              R
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../assets/javascripts/workers/search.477d984a.min.js", "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.ddd52ceb.min.js"></script>
      
    
  </body>
</html>