
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.1, mkdocs-material-7.1.10">
    
    
      
        <title>Python - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.1fe995fd.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.f1a3b89f.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../css/custom.css">
    
    
      
  



  
  



  <script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
  <script async src="https://www.google-analytics.com/analytics.js"></script>


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    <script>function __prefix(e){return new URL("../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#introduction-to-single-cell-rna-seq-ii-getting-started-with-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../index.html" title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Python
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../index.html" title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../index.html" class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../applications.html" class="md-nav__link">
        Applications
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Using kb
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_usage.html" class="md-nav__link">
        Introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_ref.html" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_count.html" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      <label class="md-nav__link" for="__nav_4">
        Tutorials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_1" type="checkbox" id="__nav_4_1" >
      
      <label class="md-nav__link" for="__nav_4_1">
        Introduction to single-cell RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../scRNA-seq_intro/python/scRNA-seq_intro.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_2" type="checkbox" id="__nav_4_2" >
      
      <label class="md-nav__link" for="__nav_4_2">
        Preprocessing & Quality Control
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/python/kb_intro_1_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/R/kb_intro_1_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_3" type="checkbox" id="__nav_4_3" checked>
      
      <label class="md-nav__link" for="__nav_4_3">
        Getting started with analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Python
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="kb_intro_2_python.html" class="md-nav__link md-nav__link--active">
        Python
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-python-packages" class="md-nav__link">
    Install python packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-an-index" class="md-nav__link">
    Download an index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exercise" class="md-nav__link">
    Exercise
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-for-library-saturation" class="md-nav__link">
    Test for library saturation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-empty-droplets" class="md-nav__link">
    Filter empty droplets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-out-by-mitochondrial-content" class="md-nav__link">
    Filtering out by mitochondrial content
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-out-genes-that-are-not-present-in-any-cells" class="md-nav__link">
    Filter out genes that are not present in any cells
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-count-distributions" class="md-nav__link">
    Visualizing count distributions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#normalize-the-counts" class="md-nav__link">
    Normalize the counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identify-highly-variable-genes" class="md-nav__link">
    Identify highly variable genes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering-and-visualization" class="md-nav__link">
    Clustering and visualization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pca" class="md-nav__link">
    PCA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tsne" class="md-nav__link">
    tSNE
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#umap" class="md-nav__link">
    UMAP
  </a>
  
    <nav class="md-nav" aria-label="UMAP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#discussion" class="md-nav__link">
    Discussion
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../R/kb_intro_2_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" >
      
      <label class="md-nav__link" for="__nav_4_4">
        Building & annotating an atlas
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_building_atlas/python/kb_analysis_0_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_building_atlas/R/kb_analysis_0_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      <label class="md-nav__link" for="__nav_4_5">
        Pseudotime analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/slingshot/kb_slingshot.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      <label class="md-nav__link" for="__nav_4_6">
        Single-nuclei RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_nucleus/python/kb_single_nucleus.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      <label class="md-nav__link" for="__nav_4_7">
        RNA Velocity
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity/python/kb_velocity.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" >
      
      <label class="md-nav__link" for="__nav_4_8">
        Feature barcode & multimodal data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_kite/python/kb_kite.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../download_data/data_download.html" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      <label class="md-nav__link" for="__nav_4_10">
        Build custom index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_custom_index/python/kb_transcriptome_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_11" type="checkbox" id="__nav_4_11" >
      
      <label class="md-nav__link" for="__nav_4_11">
        Build RNA velocity index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity_index/python/kb_velocity_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_12" type="checkbox" id="__nav_4_12" >
      
      <label class="md-nav__link" for="__nav_4_12">
        Aggregate multiple matrices
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_aggregate/python/kb_aggregating_count_matrices.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_13" type="checkbox" id="__nav_4_13" >
      
      <label class="md-nav__link" for="__nav_4_13">
        Process multiple FASTQs
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_multiple_fastqs/python/kb_multiple_files.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_14" type="checkbox" id="__nav_4_14" >
      
      <label class="md-nav__link" for="__nav_4_14">
        Processing multi-species data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_species_mixing/python/kb_species_mixing.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_species_mixing/R/kb_mixed_species_10x_v2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_15" type="checkbox" id="__nav_4_15" >
      
      <label class="md-nav__link" for="__nav_4_15">
        Pseudotime analysis (Monocle2)
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/monocle2/kb_monocle2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-python-packages" class="md-nav__link">
    Install python packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-an-index" class="md-nav__link">
    Download an index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exercise" class="md-nav__link">
    Exercise
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-for-library-saturation" class="md-nav__link">
    Test for library saturation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-empty-droplets" class="md-nav__link">
    Filter empty droplets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-out-by-mitochondrial-content" class="md-nav__link">
    Filtering out by mitochondrial content
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-out-genes-that-are-not-present-in-any-cells" class="md-nav__link">
    Filter out genes that are not present in any cells
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-count-distributions" class="md-nav__link">
    Visualizing count distributions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#normalize-the-counts" class="md-nav__link">
    Normalize the counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identify-highly-variable-genes" class="md-nav__link">
    Identify highly variable genes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering-and-visualization" class="md-nav__link">
    Clustering and visualization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pca" class="md-nav__link">
    PCA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tsne" class="md-nav__link">
    tSNE
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#umap" class="md-nav__link">
    UMAP
  </a>
  
    <nav class="md-nav" aria-label="UMAP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#discussion" class="md-nav__link">
    Discussion
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/tree/master/tutorials/docs/tutorials/kb_getting_started/python/kb_intro_2_python.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/docs/tutorials/kb_getting_started/python/kb_intro_2_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="introduction-to-single-cell-rna-seq-ii-getting-started-with-analysis">Introduction to single-cell RNA-seq II: getting started with analysis<a class="headerlink" href="#introduction-to-single-cell-rna-seq-ii-getting-started-with-analysis" title="Permanent link">&para;</a></h1>
<p>This notebook demonstrates pre-processing and basic analysis of the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE126783">mouse retinal cells GSE126783</a> dataset from <a href="https://doi.org/10.1016/j.immuni.2019.02.007">Koren et al., 2019</a>. Following pre-processing using kallisto and bustools and basic QC, the notebook demonstrates some initial analysis. The approximate running time of the notebook is about 13 minutes.</p>
<p>The notebook was written by Kyung Hoi (Joseph) Min, A. Sina Booeshaghi and Lior Pachter. If you use the methods in this notebook for your analysis please cite the following publications which describe the tools used in the notebook, as well as specific methods they run (these are cited inline in the notebook):</p>
<ul>
<li>Melsted, P., Booeshaghi, A.S. et al. Modular and efficient pre-processing of single-cell RNA-seq. bioRxiv (2019). doi:10.1101/673285</li>
<li>Wolf, F. A., Angere, P. and Theis, F.J. SCANPY: large-scale single-cell gene expression data analysis. Genome Biology (2018). doi:10.1186/s13059-017-1382-0</li>
</ul>
<p>An R notebook implementing the same analysis is available <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_intro_2_R.ipynb">here</a>. See the <a href="https://www.kallistobus.tools/tutorials">kallistobus.tools tutorials</a> site for additional notebooks demonstrating other analyses.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># This is  used to time the running of the notebook
import time
start_time = time.time()
</code></pre></div>
</td></tr></table>
<h3 id="install-python-packages">Install python packages<a class="headerlink" href="#install-python-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># These packages are pre-installed on Google Colab, but are included here to facilitate running this notebook locally
!pip install --quiet matplotlib
!pip install --quiet scikit-learn
!pip install --quiet numpy
!pip install --quiet scipy
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# `kb` is a wrapper for the kallisto and bustools program, and the kb-python package contains the kallisto and bustools executables.
!pip install --quiet kb-python==0.24.1
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">35</span>.<span class="mi">4</span><span class="nv">MB</span> <span class="mi">82</span><span class="nv">kB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">133</span><span class="nv">kB</span> <span class="mi">39</span>.<span class="mi">6</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">51</span><span class="nv">kB</span> <span class="mi">8</span>.<span class="mi">3</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">112</span><span class="nv">kB</span> <span class="mi">45</span>.<span class="mi">3</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[?<span class="mi">25</span><span class="nv">h</span>  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">loompy</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">numpy</span><span class="o">-</span><span class="nv">groupies</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
<span class="nv">CPU</span> <span class="nv">times</span>: <span class="nv">user</span> <span class="mi">107</span> <span class="nv">ms</span>, <span class="nv">sys</span>: <span class="mi">25</span>.<span class="mi">2</span> <span class="nv">ms</span>, <span class="nv">total</span>: <span class="mi">133</span> <span class="nv">ms</span>
<span class="nv">Wall</span> <span class="nv">time</span>: <span class="mi">10</span>.<span class="mi">2</span> <span class="nv">s</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# Install scanpy and other packages needed for single-cell RNA-seq analysis
!pip install --quiet scanpy python-igraph louvain MulticoreTSNE pybiomart
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">2</span>.<span class="mi">0</span><span class="nv">MB</span> <span class="mi">8</span>.<span class="mi">2</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">3</span>.<span class="mi">2</span><span class="nv">MB</span> <span class="mi">50</span>.<span class="mi">3</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">2</span>.<span class="mi">2</span><span class="nv">MB</span> <span class="mi">44</span>.<span class="mi">0</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">81</span><span class="nv">kB</span> <span class="mi">11</span>.<span class="mi">1</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">1</span>.<span class="mi">1</span><span class="nv">MB</span> <span class="mi">48</span>.<span class="mi">9</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">71</span><span class="nv">kB</span> <span class="mi">10</span>.<span class="mi">0</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[?<span class="mi">25</span><span class="nv">h</span>  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">MulticoreTSNE</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">umap</span><span class="o">-</span><span class="nv">learn</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">sinfo</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">pynndescent</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
<span class="nv">CPU</span> <span class="nv">times</span>: <span class="nv">user</span> <span class="mi">108</span> <span class="nv">ms</span>, <span class="nv">sys</span>: <span class="mi">23</span>.<span class="mi">3</span> <span class="nv">ms</span>, <span class="nv">total</span>: <span class="mi">132</span> <span class="nv">ms</span>
<span class="nv">Wall</span> <span class="nv">time</span>: <span class="mi">14</span>.<span class="mi">8</span> <span class="nv">s</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Import packages
import anndata
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scanpy as sc
from sklearn.decomposition import TruncatedSVD
from scipy import sparse, io

matplotlib.rcParams.update({&#39;font.size&#39;: 12})
%config InlineBackend.figure_format = &#39;retina&#39;
</code></pre></div>
</td></tr></table>
<h3 id="download-the-data">Download the data<a class="headerlink" href="#download-the-data" title="Permanent link">&para;</a></h3>
<p><strong>Note:</strong> We use the <code>-O</code> option for <code>wget</code> to rename the files so they can be easily identified. The notebook requires reads in fastq format; the files can be processed in gzip compressed format.</p>
<p>In this example the reads are downloaded from a Box drive; see the <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/data_download.ipynb">data download</a> notebook for information on where to find publicly available single-cell RNA-seq data.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
!wget -q https://caltech.box.com/shared/static/9u2qk1uqu6py03phooe1ti0kjd9v87pu.txt -O checksums.txt
!wget -q https://caltech.box.com/shared/static/w9ww8et5o029s2e3usjzpbq8lpot29rh.gz -O SRR8599150_S1_L001_R1_001.fastq.gz
!wget -q https://caltech.box.com/shared/static/ql00zyvqnpy7bf8ogdoe9zfy907guzy9.gz -O SRR8599150_S1_L001_R2_001.fastq.gz
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>CPU times: user 229 ms, sys: 67.9 ms, total: 297 ms
Wall time: 52.9 s
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># This is formatted as code
</code></pre></div>
</td></tr></table>
<p>Next, we verify the integrity of the files that were downloaded to confirm that they were not corrupted during the download.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!md5sum -c checksums.txt --ignore-missing
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>SRR8599150_S1_L001_R1_001.fastq.gz: OK
SRR8599150_S1_L001_R2_001.fastq.gz: OK
</code></pre></div>

<h3 id="download-an-index">Download an index<a class="headerlink" href="#download-an-index" title="Permanent link">&para;</a></h3>
<p><strong>Note:</strong> See <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_transcriptome_index.ipynb">this notebook</a> for a tutorial on how to build custom transcriptome indices or indices for RNA velocity.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# This downloads a pre-built index for kallisto to use when pseudoaligning the reads
!kb ref -d mouse -i index.idx -g t2g.txt
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">07</span> <span class="mi">18</span><span class="p">:</span><span class="mi">15</span><span class="p">:</span><span class="mi">23</span><span class="p">,</span><span class="mi">071</span><span class="p">]</span>    <span class="n">INFO</span> <span class="n">Downloading</span> <span class="n">files</span> <span class="k">for</span> <span class="n">mouse</span> <span class="n">from</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">caltech</span><span class="o">.</span><span class="n">box</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">shared</span><span class="o">/</span><span class="k">static</span><span class="o">/</span><span class="n">vcaz6cujop0xuapdmz0pplp3aoqc41si</span><span class="o">.</span><span class="n">gz</span> <span class="n">to</span> <span class="n">tmp</span><span class="o">/</span><span class="n">vcaz6cujop0xuapdmz0pplp3aoqc41si</span><span class="o">.</span><span class="n">gz</span>
<span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">07</span><span class="o">-</span><span class="mi">07</span> <span class="mi">18</span><span class="p">:</span><span class="mi">16</span><span class="p">:</span><span class="mi">52</span><span class="p">,</span><span class="mi">467</span><span class="p">]</span>    <span class="n">INFO</span> <span class="n">Extracting</span> <span class="n">files</span> <span class="n">from</span> <span class="n">tmp</span><span class="o">/</span><span class="n">vcaz6cujop0xuapdmz0pplp3aoqc41si</span><span class="o">.</span><span class="n">gz</span>
<span class="n">CPU</span> <span class="n">times</span><span class="p">:</span> <span class="n">user</span> <span class="mi">584</span> <span class="n">ms</span><span class="p">,</span> <span class="n">sys</span><span class="p">:</span> <span class="mi">114</span> <span class="n">ms</span><span class="p">,</span> <span class="n">total</span><span class="p">:</span> <span class="mi">698</span> <span class="n">ms</span>
<span class="n">Wall</span> <span class="n">time</span><span class="p">:</span> <span class="mi">2</span><span class="nb">min</span> <span class="mi">12</span><span class="n">s</span>
</code></pre></div>

<h2 id="pseudoalignment-and-counting">Pseudoalignment and counting<a class="headerlink" href="#pseudoalignment-and-counting" title="Permanent link">&para;</a></h2>
<h3 id="run-kallisto-and-bustools">Run kallisto and bustools<a class="headerlink" href="#run-kallisto-and-bustools" title="Permanent link">&para;</a></h3>
<p>The following command will generate an RNA count matrix of cells (rows) by genes (columns) in H5AD format, which is a binary format used to store <a href="https://anndata.readthedocs.io/en/stable/">Anndata</a> objects. Notice that this requires providing the index and transcript-to-gene mapping downloaded in the previous step to the <code>-i</code> and <code>-g</code> arguments respectively. Also, since the reads were generated with the 10x Genomics Chromium Single Cell v2 Chemistry, the <code>-x 10xv2</code> argument is used. To view other supported technologies, run <code>kb --list</code>.</p>
<p><strong>Note:</strong> To output a <a href="https://linnarssonlab.org/loompy/format/index.html">Loom</a> file instead, replace the <code>--h5ad</code> flag with <code>--loom</code>. To obtain the raw matrix output by <code>kb</code> instead of the H5AD or Loom converted files, omit these flags.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# This step runs `kb` to pseudoalign the reads, and then generate the cells x gene matrix in h5ad format.
!kb count -i index.idx -g t2g.txt -x 10xv2 --h5ad -t 2 \
SRR8599150_S1_L001_R1_001.fastq.gz SRR8599150_S1_L001_R2_001.fastq.gz
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[2021-07-07 18:17:35,074]    INFO Generating BUS file from
[2021-07-07 18:17:35,074]    INFO         SRR8599150_S1_L001_R1_001.fastq.gz
[2021-07-07 18:17:35,074]    INFO         SRR8599150_S1_L001_R2_001.fastq.gz
[2021-07-07 18:19:52,582]    INFO Sorting BUS file ./output.bus to tmp/output.s.bus
[2021-07-07 18:19:55,911]    INFO Whitelist not provided
[2021-07-07 18:19:55,912]    INFO Copying pre-packaged 10XV2 whitelist to .
[2021-07-07 18:19:56,015]    INFO Inspecting BUS file tmp/output.s.bus
[2021-07-07 18:19:57,425]    INFO Correcting BUS records in tmp/output.s.bus to tmp/output.s.c.bus with whitelist ./10xv2_whitelist.txt
[2021-07-07 18:20:11,880]    INFO Sorting BUS file tmp/output.s.c.bus to ./output.unfiltered.bus
[2021-07-07 18:20:14,882]    INFO Generating count matrix ./counts_unfiltered/cells_x_genes from BUS file ./output.unfiltered.bus
[2021-07-07 18:20:16,944]    INFO Converting matrix ./counts_unfiltered/cells_x_genes.mtx to h5ad ./counts_unfiltered/adata.h5ad
CPU times: user 1.05 s, sys: 159 ms, total: 1.2 s
Wall time: 2min 45s
</code></pre></div>

<h3 id="exercise">Exercise<a class="headerlink" href="#exercise" title="Permanent link">&para;</a></h3>
<ul>
<li><a href="https://linnarssonlab.org/loompy/format/index.html">Loom</a> is a alternative format for storing single-cell count matrices. Output a Loom file with kb by replacing the <code>--h5ad</code> flag with <code>--loom</code>, or obtain the raw matrix output by omitting the flags</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# # This runs `kb` to pseudoalign the reads, and then generate the cells x gene matrix in Loom format.
# !kb count -i index.idx -g t2g.txt -x 10xv2 --h5ad -t 2 \
# SRR8599150_S1_L001_R1_001.fastq.gz SRR8599150_S1_L001_R2_001.fastq.gz
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 3.58 µs
</code></pre></div>

<h2 id="basic-qc">Basic QC<a class="headerlink" href="#basic-qc" title="Permanent link">&para;</a></h2>
<p>First, the <em>cells x genes</em> matrix is imported into an H5AD-formatted <a href="https://icb-anndata.readthedocs-hosted.com/en/stable/anndata.AnnData.html">Anndata</a> matrix. Anndata is a convenient format for storing single-cell count matrices in Python.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># import data
adata = anndata.read(&#39;counts_unfiltered/adata.h5ad&#39;)
adata
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">96775</span> <span class="err">×</span> <span class="mi">55421</span>
</code></pre></div>

<p>Represent the cells in 2D with PCA</p>
<p>This is one type of embedding in which cells in higher dimensional gene expression space are represented in two dimensions.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Perform SVD
tsvd = TruncatedSVD(n_components=2)
tsvd.fit(adata.X)
X = tsvd.transform(adata.X)

# Plot the cells in the 2D PCA projection
fig, ax = plt.subplots(figsize=(10, 7))

ax.scatter(X[:,0], X[:,1], alpha=0.5, c=&quot;green&quot;)

plt.axis(&#39;off&#39;)
plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_26_0.png" /></p>
<h3 id="test-for-library-saturation">Test for library saturation<a class="headerlink" href="#test-for-library-saturation" title="Permanent link">&para;</a></h3>
<p>For each cell we ask how many genes did we detect (or see non-zero expression). The idea is that if we have "saturated" our sequencing library then increasing the number of UMI counts (x-axis) will not yield an appreciable increase in the number of genes detected (y-axis).</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Create a plot showing genes detected as a function of UMI counts.
fig, ax = plt.subplots(figsize=(10, 7))

x = np.asarray(adata.X.sum(axis=1))[:,0]
y = np.asarray(np.sum(adata.X&gt;0, axis=1))[:,0]

ax.scatter(x, y, color=&quot;green&quot;, alpha=0.25)
ax.set_xlabel(&quot;UMI Counts&quot;)
ax.set_ylabel(&quot;Genes Detected&quot;)
ax.set_xscale(&#39;log&#39;)
ax.set_yscale(&#39;log&#39;, nonposy=&#39;clip&#39;)

ax.set_xlim((0.5, 4500))
ax.set_ylim((0.5,2000))


plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_29_0.png" /></p>
<h3 id="examine-the-knee-plot">Examine the knee plot<a class="headerlink" href="#examine-the-knee-plot" title="Permanent link">&para;</a></h3>
<p>The "knee plot" was introduced in the Drop-seq paper: 
- Macosko et al., <a href="https://www.cell.com/fulltext/S0092-8674(15)00549-8">Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets</a>, 2015. DOI:10.1016/j.cell.2015.05.002</p>
<p>In this plot cells are ordered by the number of UMI counts associated to them (shown on the <em>x</em>-axis), and the fraction of droplets with at least that number of cells is shown on the <em>y</em>-axis. The idea is that "real" cells have a certain number of UMI counts and that a threshold on the UMI counts filters those cells. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#@title Threshold cells according to knee plot { run: &quot;auto&quot;, vertical-output: true }
cutoff =  200#@param {type:&quot;integer&quot;}
knee = np.sort((np.array(adata.X.sum(axis=1))).flatten())[::-1]
cell_set = np.arange(len(knee))
num_cells = cell_set[knee &gt; cutoff][::-1][0]

fig, ax = plt.subplots(figsize=(10, 7))


ax.loglog(knee, cell_set, linewidth=5, color=&quot;g&quot;)
ax.axvline(x=cutoff, linewidth=3, color=&quot;k&quot;)


ax.axhline(y=num_cells, linewidth=3, color=&quot;k&quot;)

ax.set_xlabel(&quot;UMI Counts&quot;)
ax.set_ylabel(&quot;Set of Barcodes&quot;)

plt.grid(True, which=&quot;both&quot;)
plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_31_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>print(f&quot;{num_cells:,.0f} cells passed the {cutoff} UMI threshold&quot;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="mf">3</span><span class="p">,</span><span class="mf">791</span> <span class="n">cells</span> <span class="n">passed</span> <span class="n">the</span> <span class="mf">200</span> <span class="n">UMI</span> <span class="n">threshold</span>
</code></pre></div>

<p>The knee plot can be used to threshold cells based on the number of UMI counts they contain. The threshold is applied at the "knee", where there is a sharp dropoff in the number of UMIs per cell. In this example we use the nunber 3979 based on the publication describing the data.</p>
<h3 id="filter-empty-droplets">Filter empty droplets<a class="headerlink" href="#filter-empty-droplets" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">96775</span> <span class="err">×</span> <span class="mi">55421</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Filter the cells according to the threshold determined from the knee plot
sc.pp.filter_cells(adata, min_genes=200)
sc.pp.filter_cells(adata, min_counts=knee[expected_num_cells])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">3536</span> <span class="err">×</span> <span class="mi">55421</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span>
</code></pre></div>

<h3 id="filtering-out-by-mitochondrial-content">Filtering out by mitochondrial content<a class="headerlink" href="#filtering-out-by-mitochondrial-content" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>mito_ensembl_ids = sc.queries.mitochondrial_genes(&quot;mmusculus&quot;, attrname=&quot;ensembl_gene_id&quot;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>mito_genes = mito_ensembl_ids[&quot;ensembl_gene_id&quot;].values
# for each cell compute fraction of counts in mito genes vs. all genes
# the `.A1` is only necessary as X is sparse (to transform to a dense array after summing)
adata.obs[&#39;percent_mito&#39;] = np.sum(
    adata[:, mito_genes].X, axis=1).A1 / np.sum(adata.X, axis=1).A1
# add the total counts per cell as observations-annotation to adata
adata.obs[&#39;n_counts&#39;] = adata.X.sum(axis=1).A1
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.scatter(adata, x=&#39;n_counts&#39;, y=&#39;percent_mito&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_41_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata = adata[adata.obs.percent_mito &lt; 0.03]
</code></pre></div>
</td></tr></table>
<h3 id="filter-out-genes-that-are-not-present-in-any-cells">Filter out genes that are not present in any cells<a class="headerlink" href="#filter-out-genes-that-are-not-present-in-any-cells" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_genes(adata, min_cells=3)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="k">var</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">3507</span> <span class="err">×</span> <span class="mi">15420</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;percent_mito&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_cells&#39;</span>
</code></pre></div>

<h3 id="visualizing-count-distributions">Visualizing count distributions<a class="headerlink" href="#visualizing-count-distributions" title="Permanent link">&para;</a></h3>
<p>Examination of the gene count and UMI count distributions is useful QC to evaluate the quality of the library and how deeply it was sequenced.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.violin(adata, [&#39;n_genes&#39;, &#39;n_counts&#39;, &#39;percent_mito&#39;], jitter=0.4, multi_panel=True)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_48_0.png" /></p>
<h2 id="analysis">Analysis<a class="headerlink" href="#analysis" title="Permanent link">&para;</a></h2>
<p>In this part of the tutorial, the cells are clustered by <a href="https://en.wikipedia.org/wiki/Louvain_modularity">Louvain community detection</a>.</p>
<h3 id="normalize-the-counts">Normalize the counts<a class="headerlink" href="#normalize-the-counts" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)
sc.pp.log1p(adata)
</code></pre></div>
</td></tr></table>
<h3 id="identify-highly-variable-genes">Identify highly variable genes<a class="headerlink" href="#identify-highly-variable-genes" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># flavor=&quot;cell_ranger&quot; is consistent with Seurat and flavor=&quot;suerat&quot; is not consistent with Seurat
sc.pp.highly_variable_genes(adata, min_mean=0.01, max_mean=8, min_disp=1, n_top_genes=3000, n_bins=20, flavor=&quot;seurat&quot;)
sc.pl.highly_variable_genes(adata)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_53_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.scale(adata, max_value=10)
</code></pre></div>
</td></tr></table>
<h3 id="clustering-and-visualization">Clustering and visualization<a class="headerlink" href="#clustering-and-visualization" title="Permanent link">&para;</a></h3>
<p>There are many algorithms for clustering cells, and while they have been compared in detail in various benchmarks (see e.g., <a href="https://f1000research.com/articles/7-1141/v2">Duo et al. 2018</a>), there is no univerally agreed upon method. Here we demonstrate clustering using <a href="https://en.wikipedia.org/wiki/Louvain_modularity">Louvain clustering</a>, which is a popular method for clustering single-cell RNA-seq data. The method was published in </p>
<ul>
<li>Blondel, Vincent D; Guillaume, Jean-Loup; Lambiotte, Renaud; Lefebvre, Etienne (9 October 2008). "Fast unfolding of communities in large networks". Journal of Statistical Mechanics: Theory and Experiment. 2008 (10): P10008.</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# Cluster the cells using Louvain clustering
sc.tl.pca(adata, svd_solver=&#39;arpack&#39;, use_highly_variable=True, n_comps=10)
sc.pp.neighbors(adata, n_neighbors=30, n_pcs=10, knn=True)
sc.tl.louvain(adata)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>/usr/local/lib/python3.7/dist-packages/numba/np/ufunc/parallel.py:363: NumbaWarning: The TBB threading layer requires TBB version 2019.5 or later i.e., TBB_INTERFACE_VERSION &gt;= 11005. Found TBB_INTERFACE_VERSION = 9107. The TBB threading layer is disabled.
  warnings.warn(problem)


CPU times: user 18.2 s, sys: 804 ms, total: 19 s
Wall time: 18.8 s
</code></pre></div>

<p>It is useful to revisit the PCA project with points colored by cluster. Previously we computed the PCA projection directly; here we use a function in ScanPy which does the same.</p>
<h3 id="pca">PCA<a class="headerlink" href="#pca" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Perform PCA and plot the projection to the first two dimensions, with points colored according to the Louvain clusters.
fig, ax = plt.subplots(figsize=(10, 7))
sc.pl.pca(adata, color=&#39;louvain&#39;, ax=ax)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_60_0.png" /></p>
<p>The PCA representation is the result a <em>linear</em> map of the data from its ambient dimension, to low dimension (in the case above 2D). While such projections are useful, there are <em>non-linear</em> methods that can capture non-linear geometry in the data.</p>
<h3 id="tsne">tSNE<a class="headerlink" href="#tsne" title="Permanent link">&para;</a></h3>
<p><a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding">t-SNE</a> is a non-linear dimensionality reduction technique described in:</p>
<ul>
<li>Maaten, Laurens van der, and Geoffrey Hinton. "Visualizing data using t-SNE." Journal of machine learning research 9.Nov (2008): 2579-2605.</li>
</ul>
<p>Here it is applied to the 10-dimensional PCA projection of the cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# Visualize cells with t-SNE. The n_pcs parameter sets the number of principal components to project to prior to 
# performing t-SNE
sc.tl.tsne(adata, n_pcs=10)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>CPU times: user 49.6 s, sys: 109 ms, total: 49.7 s
Wall time: 25.4 s
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>fig, ax = plt.subplots(figsize=(10, 7))
sc.pl.tsne(adata, color=&#39;louvain&#39;, ax=ax)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_64_0.png" /></p>
<h2 id="umap">UMAP<a class="headerlink" href="#umap" title="Permanent link">&para;</a></h2>
<p>UMAP stands for Uniform Manifold Approximation and Projection is a non-linear dimensionality reduction techinque described in </p>
<ul>
<li>Leland McInnes and John Healy and James Melville, "UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction" 2018 1802.03426 arXiv stat.ML</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
# Visualize cells with t-SNE. The n_pcs parameter sets the number of principal components to project to prior to 
# performing t-SNE
sc.tl.umap(adata)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>CPU times: user 13 s, sys: 132 ms, total: 13.2 s
Wall time: 12.6 s
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>fig, ax = plt.subplots(figsize=(10, 7))
sc.pl.umap(adata, color=&#39;louvain&#39;, ax=ax)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_python_files/kb_intro_2_python_67_0.png" /></p>
<h3 id="exercises">Exercises<a class="headerlink" href="#exercises" title="Permanent link">&para;</a></h3>
<ul>
<li>The variance explained by each principal component is a measure of how well a projection to that component represents the data. Compute the variance explained by each component.</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Compute and plot the variance explained by the PC subspaces.
# sc.pl.pca_variance_ratio(adata)
</code></pre></div>
</td></tr></table>
<ul>
<li>In the notebook we visualized the data using t-SNE and UMAP. These two techniques can produce different two dimensional embeddings based with different parameters and even different random seeds. Change around these parameters to see how they affect the embedding.</li>
</ul>
<p>For tSNE parameters see https://icb-scanpy.readthedocs-hosted.com/en/stable/api/scanpy.tl.tsne.html</p>
<p>For UMAP parameters see https://icb-scanpy.readthedocs-hosted.com/en/stable/api/scanpy.tl.umap.html</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># %%time
# # Visualize cells with UMAP.
# sc.tl.tsne(adata, 
# n_pcs=None, 
# use_rep=None, 
# perplexity=30, 
# early_exaggeration=12, 
# learning_rate=1000, 
# random_state=0, 
# use_fast_tsne=True, 
# n_jobs=None, 
# copy=False)
# fig, ax = plt.subplots(figsize=(10, 7))
# sc.pl.tsne(adata, color=&#39;louvain&#39;, ax=ax)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># %%time
# # Visualize cells with UMAP.
# sc.tl.umap(adata, 
# min_dist=0.5, 
# spread=1.0, 
# n_components=2, 
# maxiter=None, 
# alpha=1.0, 
# gamma=1.0, 
# negative_sample_rate=5, 
# init_pos=&#39;spectral&#39;, 
# random_state=0, 
# a=None, 
# b=None, 
# copy=False, 
# method=&#39;umap&#39;)
# fig, ax = plt.subplots(figsize=(10, 7))
# sc.pl.umap(adata, color=&#39;louvain&#39;, ax=ax)
</code></pre></div>
</td></tr></table>
<h2 id="discussion">Discussion<a class="headerlink" href="#discussion" title="Permanent link">&para;</a></h2>
<p>This notebook has demonstrated visualization of cells following pre-processing of single-cell RNA-seq data.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Running time of the notebook
print(&quot;{:.2f} minutes&quot;.format((time.time()-start_time)/60))
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="mf">7.70</span> <span class="n">minutes</span>
</code></pre></div>

<p><strong>Feedback</strong>: please report any issues, or submit pull requests for improvements, in the <a href="https://github.com/pachterlab/kallistobustools/blob/master/notebooks/kb_intro_2_python.ipynb">Github repository where this notebook is located</a>.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../../kb_quality_control/R/kb_intro_1_R.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: R" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              R
            </div>
          </div>
        </a>
      
      
        
        <a href="../R/kb_intro_2_R.html" class="md-footer__link md-footer__link--next" aria-label="Next: R" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              R
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../assets/javascripts/workers/search.477d984a.min.js", "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.ddd52ceb.min.js"></script>
      
    
  </body>
</html>