
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.2, mkdocs-material-7.3.0">
    
    
      
        <title>Python - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.8b42a75e.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.3f5d1f46.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../../css/custom.css">
    
    
      
  



  
  



  <script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
  <script async src="https://www.google-analytics.com/analytics.js"></script>


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    <script>function __prefix(e){return new URL("../../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#pre-processing-and-analysis-of-feature-barcode-single-cell-rna-seq-data-with-kite" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Python
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../applications/" class="md-nav__link">
        Applications
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_3">
          Using kb
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_usage/kb_usage/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_usage/kb_ref/" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_usage/kb_count/" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Tutorials
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_1" type="checkbox" id="__nav_4_1" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_1">
          Introduction to single-cell RNA-seq
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../scRNA-seq_intro/python/scRNA-seq_intro/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_2" type="checkbox" id="__nav_4_2" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_2">
          Preprocessing & Quality Control
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_quality_control/python/kb_intro_1_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_quality_control/R/kb_intro_1_R/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_3" type="checkbox" id="__nav_4_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_3">
          Getting started with analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_getting_started/python/kb_intro_2_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_getting_started/R/kb_intro_2_R/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_4">
          Building & annotating an atlas
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_building_atlas/python/kb_analysis_0_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_building_atlas/R/kb_analysis_0_R/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_5">
          Pseudotime analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_pseudotime/R/slingshot/kb_slingshot/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_6">
          Single-nuclei RNA-seq
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_nucleus/python/kb_single_nucleus/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_7">
          RNA Velocity
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_velocity/python/kb_velocity/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4_8">
          Feature barcode & multimodal data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Python
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Python
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pre-processing" class="md-nav__link">
    Pre-processing
  </a>
  
    <nav class="md-nav" aria-label="Pre-processing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb" class="md-nav__link">
    Install kb
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-the-feature-barcode-mismatch-index" class="md-nav__link">
    Build the feature barcode mismatch index
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#generate-a-feature-count-matrix-in-h5ad-format" class="md-nav__link">
    Generate a feature count matrix in H5AD format
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-packages" class="md-nav__link">
    Install packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot-counts" class="md-nav__link">
    Plot counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cluster-with-leiden" class="md-nav__link">
    Cluster with Leiden
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#embedding-and-antibody-quantification" class="md-nav__link">
    Embedding and Antibody Quantification
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../download_data/data_download/" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_10">
          Build custom index
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_custom_index/python/kb_transcriptome_index/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_11" type="checkbox" id="__nav_4_11" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_11">
          Build RNA velocity index
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_velocity_index/python/kb_velocity_index/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_12" type="checkbox" id="__nav_4_12" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_12">
          Aggregate multiple matrices
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_aggregate/python/kb_aggregating_count_matrices/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_13" type="checkbox" id="__nav_4_13" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_13">
          Process multiple FASTQs
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_multiple_fastqs/python/kb_multiple_files/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_14" type="checkbox" id="__nav_4_14" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_14">
          Processing multi-species data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_species_mixing/python/kb_species_mixing/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_species_mixing/R/kb_mixed_species_10x_v2/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_15" type="checkbox" id="__nav_4_15" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_15">
          Pseudotime analysis (Monocle2)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_pseudotime/R/monocle2/kb_monocle2/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pre-processing" class="md-nav__link">
    Pre-processing
  </a>
  
    <nav class="md-nav" aria-label="Pre-processing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb" class="md-nav__link">
    Install kb
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-the-feature-barcode-mismatch-index" class="md-nav__link">
    Build the feature barcode mismatch index
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#generate-a-feature-count-matrix-in-h5ad-format" class="md-nav__link">
    Generate a feature count matrix in H5AD format
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-packages" class="md-nav__link">
    Install packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot-counts" class="md-nav__link">
    Plot counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cluster-with-leiden" class="md-nav__link">
    Cluster with Leiden
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#embedding-and-antibody-quantification" class="md-nav__link">
    Embedding and Antibody Quantification
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/tree/master/tutorials/docs/tutorials/kb_kite/python/kb_kite.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/tutorials/docs/tutorials/kb_kite/python/kb_kite.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="pre-processing-and-analysis-of-feature-barcode-single-cell-rna-seq-data-with-kite">Pre-processing and analysis of feature barcode single-cell RNA-seq data with KITE.<a class="headerlink" href="#pre-processing-and-analysis-of-feature-barcode-single-cell-rna-seq-data-with-kite" title="Permanent link">&para;</a></h1>
<p>In this notebook, we will perform pre-processing and analysis of <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_protein_v3">10x Genomics pbmc_1k_protein_v3</a> feature barcoding dataset using the <strong>Kallisto Indexing and Tag Extraction (KITE)</strong> workflow, implemented with a wrapper called <code>kb</code>. It was developed by Kyung Hoi (Joseph) Min and A. Sina Booeshaghi.</p>
<p>In Feature Barcoding assays, cellular data are recorded as short DNA sequences using procedures adapted from single-cell RNA-seq. The <strong>KITE</strong> workflow generates a "Mismatch Map" containing the sequences of all Feature Barcodes used in the experiment as well as all of their single-base mismatches. The Mismatch Map is used to produce transcipt-to-gene (.t2g) and fasta (.fa) files to be used as inputs for kallisto. An index is made with kallisto index, then kallisto | bustools effectively searches the sequencing data for the sequences in the Mismatch Map.</p>
<h2 id="pre-processing">Pre-processing<a class="headerlink" href="#pre-processing" title="Permanent link">&para;</a></h2>
<h3 id="download-the-data">Download the data<a class="headerlink" href="#download-the-data" title="Permanent link">&para;</a></h3>
<p><strong>Note:</strong> We use the <code>-O</code> option for <code>wget</code> to rename the files to easily identify them.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
!wget -q https://caltech.box.com/shared/static/asmj4nu90ydhsrk3pm7aaxu00cnnfige.txt -O checksums.txt
!wget -q https://caltech.box.com/shared/static/mp2vr3p6dztdyatuag8ir3cektmrztg8.gz -O pbmc_1k_protein_v3_antibody_S2_L001_R1_001.fastq.gz
!wget -q https://caltech.box.com/shared/static/f3payi1za7mn0jfai7vm10sy3yqwgpqh.gz -O pbmc_1k_protein_v3_antibody_S2_L001_R2_001.fastq.gz
!wget -q https://caltech.box.com/shared/static/e112bbczh9o1rl6gfin36bqp0ga7uvdy.gz -O pbmc_1k_protein_v3_antibody_S2_L002_R1_001.fastq.gz
!wget -q https://caltech.box.com/shared/static/3ve2axc8dr8v5nnrhmynrdgpqj6xg42k.gz -O pbmc_1k_protein_v3_antibody_S2_L002_R2_001.fastq.gz
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>CPU times: user 477 ms, sys: 86.2 ms, total: 563 ms
Wall time: 1min
</code></pre></div>

<p>Then, we verify the integrity of the files we downloaded to make sure they were not corrupted during the download.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!md5sum -c checksums.txt --ignore-missing
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>pbmc_1k_protein_v3_antibody_S2_L001_R1_001.fastq.gz: OK
pbmc_1k_protein_v3_antibody_S2_L001_R2_001.fastq.gz: OK
pbmc_1k_protein_v3_antibody_S2_L002_R1_001.fastq.gz: OK
pbmc_1k_protein_v3_antibody_S2_L002_R2_001.fastq.gz: OK
</code></pre></div>

<h3 id="install-kb">Install <code>kb</code><a class="headerlink" href="#install-kb" title="Permanent link">&para;</a></h3>
<p>Install <code>kb</code> for running the kallisto|bustools workflow.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip install --quiet git+https://github.com/pachterlab/kb_python@devel
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">51</span><span class="nv">kB</span> <span class="mi">2</span>.<span class="mi">6</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">13</span>.<span class="mi">2</span><span class="nv">MB</span> <span class="mi">314</span><span class="nv">kB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>████████████████████████████████<span class="o">|</span> <span class="mi">112</span><span class="nv">kB</span> <span class="mi">43</span>.<span class="mi">4</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[?<span class="mi">25</span><span class="nv">h</span>  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">kb</span><span class="o">-</span><span class="nv">python</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">loompy</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">numpy</span><span class="o">-</span><span class="nv">groupies</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
</code></pre></div>

<h3 id="build-the-feature-barcode-mismatch-index">Build the feature barcode mismatch index<a class="headerlink" href="#build-the-feature-barcode-mismatch-index" title="Permanent link">&para;</a></h3>
<p><code>kb</code> is able to generate a FASTA file containing all hamming distance &lt; 2 variants of the feature barcodes and create a kallisto index of these sequences. But it in order to do so, we first need to prepare a TSV containing feature barcode sequences in the first column and the feature barcode names in the second.</p>
<p>First, we download the feature reference file provided by 10x Genomics.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!wget -q http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_protein_v3/pbmc_1k_protein_v3_feature_ref.csv
</code></pre></div>
</td></tr></table>
<p>Let's load it in as a Pandas DataFrame.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import pandas as pd

df = pd.read_csv(&#39;pbmc_1k_protein_v3_feature_ref.csv&#39;)
df
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id</th>
      <th>name</th>
      <th>read</th>
      <th>pattern</th>
      <th>sequence</th>
      <th>feature_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CD3</td>
      <td>CD3_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>AACAAGACCCTTGAG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CD4</td>
      <td>CD4_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>TACCCGTAATAGCGT</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CD8a</td>
      <td>CD8a_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>ATTGGCACTCAGATG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CD14</td>
      <td>CD14_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>GAAAGTCAAAGCACT</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CD15</td>
      <td>CD15_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>ACGAATCAATCTGTG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>5</th>
      <td>CD16</td>
      <td>CD16_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>GTCTTTGTCAGTGCA</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>6</th>
      <td>CD56</td>
      <td>CD56_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>GTTGTCCGACAATAC</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>7</th>
      <td>CD19</td>
      <td>CD19_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>TCAACGCTTGGCTAG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>8</th>
      <td>CD25</td>
      <td>CD25_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>GTGCATTCAACAGTA</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>9</th>
      <td>CD45RA</td>
      <td>CD45RA_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>GATGAGAACAGGTTT</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>10</th>
      <td>CD45RO</td>
      <td>CD45RO_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>TGCATGTCATCGGTG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>11</th>
      <td>PD-1</td>
      <td>PD-1_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>AAGTCGTGAGGCATG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>12</th>
      <td>TIGIT</td>
      <td>TIGIT_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>TGAAGGCTCATTTGT</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>13</th>
      <td>CD127</td>
      <td>CD127_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>ACATTGACGCAACTA</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>14</th>
      <td>IgG2a</td>
      <td>IgG2a_control_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>CTCTATTCAGACCAG</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>15</th>
      <td>IgG1</td>
      <td>IgG1_control_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>ACTCACTGGAGTCTC</td>
      <td>Antibody Capture</td>
    </tr>
    <tr>
      <th>16</th>
      <td>IgG2b</td>
      <td>IgG2b_control_TotalSeqB</td>
      <td>R2</td>
      <td>5PNNNNNNNNNN(BC)NNNNNNNNN</td>
      <td>ATCACATCGTTGCCA</td>
      <td>Antibody Capture</td>
    </tr>
  </tbody>
</table>
</div>

<p>We'll convert this dataframe into a TSV format that <code>kb</code> requires.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>df[[&#39;sequence&#39;, &#39;id&#39;]].to_csv(&#39;features.tsv&#39;, index=None, header=None, sep=&#39;\t&#39;)
!cat features.tsv
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>AACAAGACCCTTGAG CD3
TACCCGTAATAGCGT CD4
ATTGGCACTCAGATG CD8a
GAAAGTCAAAGCACT CD14
ACGAATCAATCTGTG CD15
GTCTTTGTCAGTGCA CD16
GTTGTCCGACAATAC CD56
TCAACGCTTGGCTAG CD19
GTGCATTCAACAGTA CD25
GATGAGAACAGGTTT CD45RA
TGCATGTCATCGGTG CD45RO
AAGTCGTGAGGCATG PD-1
TGAAGGCTCATTTGT TIGIT
ACATTGACGCAACTA CD127
CTCTATTCAGACCAG IgG2a
ACTCACTGGAGTCTC IgG1
ATCACATCGTTGCCA IgG2b
</code></pre></div>

<p>Finally, we use <code>kb</code> to generate the mismatch kallisto index.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!kb ref -i mismatch.idx -f1 mismatch.fa -g t2g.txt --workflow kite features.tsv
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[2021-03-31 23:30:15,970]    INFO Generating mismatch FASTA at mismatch.fa
[2021-03-31 23:30:15,982]    INFO Creating transcript-to-gene mapping at t2g.txt
[2021-03-31 23:30:15,986]    INFO Indexing mismatch.fa to mismatch.idx
</code></pre></div>

<h3 id="generate-a-feature-count-matrix-in-h5ad-format">Generate a feature count matrix in H5AD format<a class="headerlink" href="#generate-a-feature-count-matrix-in-h5ad-format" title="Permanent link">&para;</a></h3>
<p>The following command will generate an RNA count matrix of cells (rows) by genes (columns) in H5AD format, which is a binary format used to store <a href="https://anndata.readthedocs.io/en/stable/">Anndata</a> objects. Notice we are providing the index and transcript-to-gene mapping we generated in the previous step to the <code>-i</code> and <code>-g</code> arguments respectively. Also, these reads were generated with the 10x Genomics Chromium Single Cell v3 Chemistry, hence the <code>-x 10xv3</code> argument. To view other supported technologies, run <code>kb --list</code>.</p>
<p><strong>Note:</strong> If you would like a Loom file instead, replace the <code>--h5ad</code> flag with <code>--loom</code>. If you want to use the raw matrix output by <code>kb</code> instead of their H5AD or Loom converted files, omit these flags.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>%%time
!kb count --h5ad -i mismatch.idx -g t2g.txt -x 10xv3 --workflow kite -t 2 \
pbmc_1k_protein_v3_antibody_S2_L001_R1_001.fastq.gz \
pbmc_1k_protein_v3_antibody_S2_L001_R2_001.fastq.gz \
pbmc_1k_protein_v3_antibody_S2_L002_R1_001.fastq.gz \
pbmc_1k_protein_v3_antibody_S2_L002_R2_001.fastq.gz
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[2021-03-31 23:30:17,474]    INFO Using index mismatch.idx to generate BUS file to . from
[2021-03-31 23:30:17,474]    INFO         pbmc_1k_protein_v3_antibody_S2_L001_R1_001.fastq.gz
[2021-03-31 23:30:17,474]    INFO         pbmc_1k_protein_v3_antibody_S2_L001_R2_001.fastq.gz
[2021-03-31 23:30:17,474]    INFO         pbmc_1k_protein_v3_antibody_S2_L002_R1_001.fastq.gz
[2021-03-31 23:30:17,474]    INFO         pbmc_1k_protein_v3_antibody_S2_L002_R2_001.fastq.gz
[2021-03-31 23:32:02,567]    INFO Sorting BUS file ./output.bus to ./tmp/output.s.bus
[2021-03-31 23:32:18,183]    INFO Whitelist not provided
[2021-03-31 23:32:18,183]    INFO Copying pre-packaged 10XV3 whitelist to .
[2021-03-31 23:32:19,157]    INFO Inspecting BUS file ./tmp/output.s.bus
[2021-03-31 23:32:31,284]    INFO Correcting BUS records in ./tmp/output.s.bus to ./tmp/output.s.c.bus with whitelist ./10xv3_whitelist.txt
[2021-03-31 23:32:49,746]    INFO Sorting BUS file ./tmp/output.s.c.bus to ./output.unfiltered.bus
[2021-03-31 23:33:01,416]    INFO Generating count matrix ./counts_unfiltered/cells_x_features from BUS file ./output.unfiltered.bus
[2021-03-31 23:33:03,924]    INFO Reading matrix ./counts_unfiltered/cells_x_features.mtx
[2021-03-31 23:33:05,698]    INFO Writing matrix to h5ad ./counts_unfiltered/adata.h5ad
CPU times: user 1.3 s, sys: 180 ms, total: 1.48 s
Wall time: 2min 49s
</code></pre></div>

<h2 id="analysis">Analysis<a class="headerlink" href="#analysis" title="Permanent link">&para;</a></h2>
<p>In this part of the tutorial, we will load the RNA count matrix generated by <code>kb count</code> into Python and cluster the cells with Leiden.</p>
<h3 id="install-packages">Install packages<a class="headerlink" href="#install-packages" title="Permanent link">&para;</a></h3>
<p>Google Colab does not come with <code>Scanpy</code>, <code>python-igraph</code>, or <code>louvain</code> (but comes with <code>matplotlib</code>, <code>numpy</code>, <code>pandas</code>, and <code>scipy</code>).</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip --quiet install leidenalg scanpy MulticoreTSNE
</code></pre></div>
</td></tr></table>
<h3 id="import-packages">Import packages<a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import anndata
import numpy as np
import scanpy as sc
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata = anndata.read_h5ad(&#39;counts_unfiltered/adata.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">124716</span> <span class="err">×</span> <span class="mi">17</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;feature_name&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata.obs.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
    <tr>
      <th>barcode</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCCAAGAAACCCA</th>
    </tr>
    <tr>
      <th>AAACCCAAGACGAGGA</th>
    </tr>
    <tr>
      <th>AAACCCAAGAGTGTGT</th>
    </tr>
    <tr>
      <th>AAACCCAAGAGTGTTG</th>
    </tr>
    <tr>
      <th>AAACCCAAGATAGCAC</th>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata.var
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature_name</th>
    </tr>
    <tr>
      <th>feature_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CD3</th>
      <td>CD3</td>
    </tr>
    <tr>
      <th>CD4</th>
      <td>CD4</td>
    </tr>
    <tr>
      <th>CD8a</th>
      <td>CD8a</td>
    </tr>
    <tr>
      <th>CD14</th>
      <td>CD14</td>
    </tr>
    <tr>
      <th>CD15</th>
      <td>CD15</td>
    </tr>
    <tr>
      <th>CD16</th>
      <td>CD16</td>
    </tr>
    <tr>
      <th>CD56</th>
      <td>CD56</td>
    </tr>
    <tr>
      <th>CD19</th>
      <td>CD19</td>
    </tr>
    <tr>
      <th>CD25</th>
      <td>CD25</td>
    </tr>
    <tr>
      <th>CD45RA</th>
      <td>CD45RA</td>
    </tr>
    <tr>
      <th>CD45RO</th>
      <td>CD45RO</td>
    </tr>
    <tr>
      <th>PD-1</th>
      <td>PD-1</td>
    </tr>
    <tr>
      <th>TIGIT</th>
      <td>TIGIT</td>
    </tr>
    <tr>
      <th>CD127</th>
      <td>CD127</td>
    </tr>
    <tr>
      <th>IgG2a</th>
      <td>IgG2a</td>
    </tr>
    <tr>
      <th>IgG1</th>
      <td>IgG1</td>
    </tr>
    <tr>
      <th>IgG2b</th>
      <td>IgG2b</td>
    </tr>
  </tbody>
</table>
</div>

<h3 id="plot-counts">Plot counts<a class="headerlink" href="#plot-counts" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_cells(adata, min_counts=0)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_genes(adata, min_counts=0)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.violin(adata, keys=&#39;n_counts&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_32_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata.obs[&#39;n_countslog&#39;] = np.log1p(adata.obs[&#39;n_counts&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.violin(adata, keys=&#39;n_countslog&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_34_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>adata.obs.index
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Index([&#39;AAACCCAAGAAACCCA&#39;, &#39;AAACCCAAGACGAGGA&#39;, &#39;AAACCCAAGAGTGTGT&#39;,
       &#39;AAACCCAAGAGTGTTG&#39;, &#39;AAACCCAAGATAGCAC&#39;, &#39;AAACCCAAGATGAGTC&#39;,
       &#39;AAACCCAAGATGGTAC&#39;, &#39;AAACCCAAGATTCGTT&#39;, &#39;AAACCCAAGATTTGGG&#39;,
       &#39;AAACCCAAGCAAGCAT&#39;,
       ...
       &#39;TTTGTTGTCGTCGCCT&#39;, &#39;TTTGTTGTCGTTGACG&#39;, &#39;TTTGTTGTCTAACCGG&#39;,
       &#39;TTTGTTGTCTATGTAG&#39;, &#39;TTTGTTGTCTCAACAA&#39;, &#39;TTTGTTGTCTCACTCA&#39;,
       &#39;TTTGTTGTCTCTTCGA&#39;, &#39;TTTGTTGTCTCTTGGT&#39;, &#39;TTTGTTGTCTGCACTT&#39;,
       &#39;TTTGTTGTCTGCGACA&#39;],
      dtype=&#39;object&#39;, name=&#39;barcode&#39;, length=124716)
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_cells(adata, min_counts=1000)
sc.pl.violin(adata, keys=&#39;n_countslog&#39;, title=&quot;kallisto UMI counts&quot;)
adata
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_36_0.png" /></p>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">725</span> <span class="err">×</span> <span class="mi">17</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;feature_name&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span>
</code></pre></div>

<p>Here are violin plots for each Feature Barcode (antibody-oligo conjugates, x-axis) across all cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.violin(adata, keys=list(adata.var.index)[-17:], xlabel=&#39;kallisto&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_38_0.png" /></p>
<h3 id="cluster-with-leiden">Cluster with Leiden<a class="headerlink" href="#cluster-with-leiden" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.normalize_per_cell(adata, counts_per_cell_after=10000)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.neighbors(adata)
sc.tl.umap(adata)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.leiden(adata, resolution=0.05)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(adata, color=&#39;leiden&#39;, palette=&#39;tab10&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_43_0.png" /></p>
<h3 id="embedding-and-antibody-quantification">Embedding and Antibody Quantification<a class="headerlink" href="#embedding-and-antibody-quantification" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(adata, color=adata.var.index)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_45_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.violin(adata, keys=list(adata.var.index[:2]), groupby=&#39;leiden&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_kite_files/kb_kite_46_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>
</code></pre></div>
</td></tr></table>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../../../kb_velocity/python/kb_velocity/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Python" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Python
            </div>
          </div>
        </a>
      
      
        
        <a href="../../../download_data/data_download/" class="md-footer__link md-footer__link--next" aria-label="Next: Find &amp; download data" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Find & download data
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        
          Made with
          <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
            Material for MkDocs
          </a>
        
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../../assets/javascripts/workers/search.f8263e09.min.js", "version": null}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.4fc53ad4.min.js"></script>
      
    
  </body>
</html>