
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.0.7">
    
    
      
        <title>R - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/stylesheets/main.c772ddf0.min.css">
      
        
        <link rel="stylesheet" href="../../../../../assets/stylesheets/palette.7fa14f5b.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../../../css/custom.css">
    
    
      
        
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","kallistobus.tools"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
      
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#setup" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../.." title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              R
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../.." title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" >
      
      <label class="md-nav__link" for="__nav_2">
        Using kb
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../kb_usage/kb_usage/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../kb_usage/kb_ref/" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../kb_usage/kb_count/" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" checked>
      
      <label class="md-nav__link" for="__nav_3">
        Tutorials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_1" type="checkbox" id="__nav_3_1" >
      
      <label class="md-nav__link" for="__nav_3_1">
        Introduction to single-cell RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../scRNA-seq_intro/python/scRNA-seq_intro/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_2" type="checkbox" id="__nav_3_2" >
      
      <label class="md-nav__link" for="__nav_3_2">
        Preprocessing & Quality Control
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_quality_control/python/kb_intro_1_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_quality_control/R/kb_intro_1_R/" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_3" type="checkbox" id="__nav_3_3" >
      
      <label class="md-nav__link" for="__nav_3_3">
        Getting started with analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_getting_started/python/kb_intro_2_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_getting_started/R/kb_intro_2_R/" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_4" type="checkbox" id="__nav_3_4" >
      
      <label class="md-nav__link" for="__nav_3_4">
        Building & annotating an atlas
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_building_atlas/python/kb_analysis_0_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_building_atlas/R/kb_analysis_0_R/" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_5" type="checkbox" id="__nav_3_5" >
      
      <label class="md-nav__link" for="__nav_3_5">
        Pseudotime analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../slingshot/kb_slingshot/" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_6" type="checkbox" id="__nav_3_6" >
      
      <label class="md-nav__link" for="__nav_3_6">
        Single-nuclei RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_nucleus/python/kb_single_nucleus/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_7" type="checkbox" id="__nav_3_7" >
      
      <label class="md-nav__link" for="__nav_3_7">
        RNA Velocity
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_velocity/python/kb_velocity/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_8" type="checkbox" id="__nav_3_8" >
      
      <label class="md-nav__link" for="__nav_3_8">
        Feature barcode & multimodal data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_kite/python/kb_kite/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../download_data/data_download/" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_10" type="checkbox" id="__nav_3_10" >
      
      <label class="md-nav__link" for="__nav_3_10">
        Build custom index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_custom_index/python/kb_transcriptome_index/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_11" type="checkbox" id="__nav_3_11" >
      
      <label class="md-nav__link" for="__nav_3_11">
        Build RNA velocity index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_velocity_index/python/kb_velocity_index/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_12" type="checkbox" id="__nav_3_12" >
      
      <label class="md-nav__link" for="__nav_3_12">
        Aggregate multiple matrices
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_aggregate/python/kb_aggregating_count_matrices/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_13" type="checkbox" id="__nav_3_13" >
      
      <label class="md-nav__link" for="__nav_3_13">
        Process multiple FASTQs
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_multiple_fastqs/python/kb_multiple_files/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_14" type="checkbox" id="__nav_3_14" >
      
      <label class="md-nav__link" for="__nav_3_14">
        Processing multi-species data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_species_mixing/python/kb_species_mixing/" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_species_mixing/R/kb_mixed_species_10x_v2/" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3_15" type="checkbox" id="__nav_3_15" checked>
      
      <label class="md-nav__link" for="__nav_3_15">
        Pseudotime analysis (Monocle2)
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_3_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          R
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        R
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download data
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download data
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/edit/master/docs/tutorials/kb_pseudotime/R/monocle2/kb_monocle2.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_monocle2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h1>
<p>This is the Google colab version of the Monocle 2 notebook on the <a href="https://bustools.github.io/BUS_notebooks_R/monocle2.html">kallisto | bustools R notebook website</a>. This version follows the static version closely, but uses the <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_1k_v3">10xv3 1k E18 mouse neuron dataset</a> to reduce download time and runtime for interactive use here. Google colab gives each notebook a fresh Ubuntu machine with R and some common packages pre-installed. Here we install the packages used in the downstream analysis that are not pre-installed. This takes a while, as some of the packages have several dependencies that have C++ code; it seems that most of time spent installing these packages is spent on compiling C++ code. Package installation might not have taken this long in your experience, since you most likely already have many of the common dependencies installed on your computer, which is not the case here.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;DropletUtils&quot;</span><span class="p">,</span> <span class="s">&quot;monocle&quot;</span><span class="p">,</span> <span class="s">&quot;SingleR&quot;</span><span class="p">,</span> <span class="s">&quot;BUSpaRse&quot;</span><span class="p">,</span> <span class="s">&quot;scater&quot;</span><span class="p">,</span> <span class="s">&quot;scran&quot;</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Installing</span> <span class="n">package</span> <span class="n">into</span> <span class="err">‘</span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">library</span><span class="err">’</span>
<span class="p">(</span><span class="k">as</span> <span class="err">‘</span><span class="n">lib</span><span class="err">’</span> <span class="k">is</span> <span class="n">unspecified</span><span class="p">)</span>

<span class="n">Bioconductor</span> <span class="n">version</span> <span class="mf">3.10</span> <span class="p">(</span><span class="n">BiocManager</span> <span class="mf">1.30</span><span class="o">.</span><span class="mi">10</span><span class="p">),</span> <span class="n">R</span> <span class="mf">3.6</span><span class="o">.</span><span class="mi">2</span> <span class="p">(</span><span class="mi">2019</span><span class="o">-</span><span class="mi">12</span><span class="o">-</span><span class="mi">12</span><span class="p">)</span>

<span class="n">Installing</span> <span class="n">package</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="s1">&#39;BiocVersion&#39;</span><span class="p">,</span> <span class="s1">&#39;DropletUtils&#39;</span><span class="p">,</span> <span class="s1">&#39;monocle&#39;</span><span class="p">,</span> <span class="s1">&#39;SingleR&#39;</span><span class="p">,</span>
  <span class="s1">&#39;BUSpaRse&#39;</span>

<span class="n">also</span> <span class="n">installing</span> <span class="n">the</span> <span class="n">dependencies</span> <span class="err">‘</span><span class="n">formatR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bit</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">vctrs</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lambda</span><span class="o">.</span><span class="n">r</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">futile</span><span class="o">.</span><span class="n">options</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bitops</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">interactiveDisplayBase</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bit64</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">blob</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DBI</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">zlibbioc</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rhtslib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">futile</span><span class="o">.</span><span class="n">logger</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">snow</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">locfit</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">IRanges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">oo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">methodsS3</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sitmo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppEigen</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">FNN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggrepel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gridExtra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sparsesvd</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">docopt</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">graph</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RBGL</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">XML</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RCurl</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RUnit</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppAnnoy</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppHNSW</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">AnnotationHub</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocFileCache</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rappdirs</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RSQLite</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">XVector</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rtracklayer</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rsamtools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ProtGenerics</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomeInfoDbData</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicAlignments</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SingleCellExperiment</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">S4Vectors</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocParallel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">edgeR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rhdf5</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">HDF5Array</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">utils</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">dqrng</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">beachmat</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rhdf5lib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Biobase</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">VGAM</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DDRTree</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">igraph</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocGenerics</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">HSMMSingleCell</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">combinat</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">fastICA</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">irlba</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">matrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">densityClust</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rtsne</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">limma</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">qlcMatrix</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">pheatmap</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">proxy</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">slam</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">viridis</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">biocViews</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RANN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SummarizedExperiment</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DelayedArray</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DelayedMatrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocNeighbors</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ExperimentHub</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">AnnotationDbi</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">AnnotationFilter</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">biomaRt</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Biostrings</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BSgenome</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ensembldb</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomeInfoDb</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicFeatures</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicRanges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plyranges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppParallel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppArmadillo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppProgress</span><span class="err">’</span>


<span class="n">Old</span> <span class="n">packages</span><span class="p">:</span> <span class="s1">&#39;BH&#39;</span><span class="p">,</span> <span class="s1">&#39;broom&#39;</span><span class="p">,</span> <span class="s1">&#39;callr&#39;</span><span class="p">,</span> <span class="s1">&#39;cli&#39;</span><span class="p">,</span> <span class="s1">&#39;curl&#39;</span><span class="p">,</span> <span class="s1">&#39;DBI&#39;</span><span class="p">,</span> <span class="s1">&#39;DT&#39;</span><span class="p">,</span> <span class="s1">&#39;fansi&#39;</span><span class="p">,</span>
  <span class="s1">&#39;farver&#39;</span><span class="p">,</span> <span class="s1">&#39;gh&#39;</span><span class="p">,</span> <span class="s1">&#39;hms&#39;</span><span class="p">,</span> <span class="s1">&#39;knitr&#39;</span><span class="p">,</span> <span class="s1">&#39;mime&#39;</span><span class="p">,</span> <span class="s1">&#39;pillar&#39;</span><span class="p">,</span> <span class="s1">&#39;prettyunits&#39;</span><span class="p">,</span> <span class="s1">&#39;rlang&#39;</span><span class="p">,</span>
  <span class="s1">&#39;rmarkdown&#39;</span><span class="p">,</span> <span class="s1">&#39;rprojroot&#39;</span><span class="p">,</span> <span class="s1">&#39;stringi&#39;</span><span class="p">,</span> <span class="s1">&#39;tidyr&#39;</span><span class="p">,</span> <span class="s1">&#39;tinytex&#39;</span><span class="p">,</span> <span class="s1">&#39;vctrs&#39;</span><span class="p">,</span> <span class="s1">&#39;xfun&#39;</span><span class="p">,</span>
  <span class="s1">&#39;xtable&#39;</span><span class="p">,</span> <span class="s1">&#39;boot&#39;</span><span class="p">,</span> <span class="s1">&#39;foreign&#39;</span><span class="p">,</span> <span class="s1">&#39;MASS&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;wget https://github.com/pachterlab/kallisto/releases/download/v0.46.1/kallisto_linux-v0.46.1.tar.gz&quot;</span><span class="p">)</span>
<span class="nf">untar</span><span class="p">(</span><span class="s">&quot;kallisto_linux-v0.46.1.tar.gz&quot;</span><span class="p">)</span>
<span class="nf">system</span><span class="p">(</span><span class="s">&quot;cp kallisto/kallisto /usr/local/bin&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>To get from fastq files to the gene count matrix, we will use the <code>kb</code> wrapper of <code>kallisto</code> and <code>bustools</code> here, which condenses several commands directly calling <code>kallisto</code> and <code>bustools</code> into 2. See the <a href="https://bustools.github.io/BUS_notebooks_R/monocle2.html">static version of this notebook</a> for instructions of calling <code>kallisto</code> and <code>bustools</code> directly.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;pip3 install kb-python&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">BUSpaRse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DropletUtils</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">scater</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">monocle</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">SingleR</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">theme_set</span><span class="p">(</span><span class="nf">theme_bw</span><span class="p">())</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">SingleCellExperiment</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">SummarizedExperiment</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">GenomicRanges</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">stats4</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">BiocGenerics</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">parallel</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="s1">&#39;BiocGenerics&#39;</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:parallel&#39;</span><span class="p">:</span>

    <span class="n">clusterApply</span><span class="p">,</span> <span class="n">clusterApplyLB</span><span class="p">,</span> <span class="n">clusterCall</span><span class="p">,</span> <span class="n">clusterEvalQ</span><span class="p">,</span>
    <span class="n">clusterExport</span><span class="p">,</span> <span class="n">clusterMap</span><span class="p">,</span> <span class="n">parApply</span><span class="p">,</span> <span class="n">parCapply</span><span class="p">,</span> <span class="n">parLapply</span><span class="p">,</span>
    <span class="n">parLapplyLB</span><span class="p">,</span> <span class="n">parRapply</span><span class="p">,</span> <span class="n">parSapply</span><span class="p">,</span> <span class="n">parSapplyLB</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:stats&#39;</span><span class="p">:</span>

    <span class="n">IQR</span><span class="p">,</span> <span class="n">mad</span><span class="p">,</span> <span class="n">sd</span><span class="p">,</span> <span class="k">var</span><span class="p">,</span> <span class="n">xtabs</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:base&#39;</span><span class="p">:</span>

    <span class="n">anyDuplicated</span><span class="p">,</span> <span class="n">append</span><span class="p">,</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">,</span> <span class="n">basename</span><span class="p">,</span> <span class="n">cbind</span><span class="p">,</span> <span class="n">colnames</span><span class="p">,</span>
    <span class="n">dirname</span><span class="p">,</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">,</span> <span class="n">duplicated</span><span class="p">,</span> <span class="n">eval</span><span class="p">,</span> <span class="n">evalq</span><span class="p">,</span> <span class="n">Filter</span><span class="p">,</span> <span class="n">Find</span><span class="p">,</span> <span class="n">get</span><span class="p">,</span> <span class="n">grep</span><span class="p">,</span>
    <span class="n">grepl</span><span class="p">,</span> <span class="n">intersect</span><span class="p">,</span> <span class="k">is</span><span class="o">.</span><span class="n">unsorted</span><span class="p">,</span> <span class="n">lapply</span><span class="p">,</span> <span class="n">Map</span><span class="p">,</span> <span class="n">mapply</span><span class="p">,</span> <span class="k">match</span><span class="p">,</span> <span class="n">mget</span><span class="p">,</span>
    <span class="n">order</span><span class="p">,</span> <span class="n">paste</span><span class="p">,</span> <span class="n">pmax</span><span class="p">,</span> <span class="n">pmax</span><span class="o">.</span><span class="n">int</span><span class="p">,</span> <span class="n">pmin</span><span class="p">,</span> <span class="n">pmin</span><span class="o">.</span><span class="n">int</span><span class="p">,</span> <span class="n">Position</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span>
    <span class="n">rbind</span><span class="p">,</span> <span class="n">Reduce</span><span class="p">,</span> <span class="n">rownames</span><span class="p">,</span> <span class="n">sapply</span><span class="p">,</span> <span class="n">setdiff</span><span class="p">,</span> <span class="n">sort</span><span class="p">,</span> <span class="n">table</span><span class="p">,</span> <span class="n">tapply</span><span class="p">,</span>
    <span class="n">union</span><span class="p">,</span> <span class="n">unique</span><span class="p">,</span> <span class="n">unsplit</span><span class="p">,</span> <span class="n">which</span><span class="p">,</span> <span class="n">which</span><span class="o">.</span><span class="n">max</span><span class="p">,</span> <span class="n">which</span><span class="o">.</span><span class="n">min</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">S4Vectors</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="s1">&#39;S4Vectors&#39;</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:base&#39;</span><span class="p">:</span>

    <span class="n">expand</span><span class="o">.</span><span class="n">grid</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">IRanges</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">GenomeInfoDb</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">Biobase</span>

<span class="n">Welcome</span> <span class="n">to</span> <span class="n">Bioconductor</span>

    <span class="n">Vignettes</span> <span class="n">contain</span> <span class="n">introductory</span> <span class="n">material</span><span class="p">;</span> <span class="n">view</span> <span class="n">with</span>
    <span class="s1">&#39;browseVignettes()&#39;</span><span class="o">.</span> <span class="n">To</span> <span class="n">cite</span> <span class="n">Bioconductor</span><span class="p">,</span> <span class="n">see</span>
    <span class="s1">&#39;citation(&quot;Biobase&quot;)&#39;</span><span class="p">,</span> <span class="ow">and</span> <span class="k">for</span> <span class="n">packages</span> <span class="s1">&#39;citation(&quot;pkgname&quot;)&#39;</span><span class="o">.</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">DelayedArray</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">matrixStats</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="s1">&#39;matrixStats&#39;</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:Biobase&#39;</span><span class="p">:</span>

    <span class="n">anyMissing</span><span class="p">,</span> <span class="n">rowMedians</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">BiocParallel</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="s1">&#39;DelayedArray&#39;</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:matrixStats&#39;</span><span class="p">:</span>

    <span class="n">colMaxs</span><span class="p">,</span> <span class="n">colMins</span><span class="p">,</span> <span class="n">colRanges</span><span class="p">,</span> <span class="n">rowMaxs</span><span class="p">,</span> <span class="n">rowMins</span><span class="p">,</span> <span class="n">rowRanges</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:base&#39;</span><span class="p">:</span>

    <span class="n">aperm</span><span class="p">,</span> <span class="n">apply</span><span class="p">,</span> <span class="n">rowsum</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">ggplot2</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">Matrix</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="s1">&#39;Matrix&#39;</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="s1">&#39;package:S4Vectors&#39;</span><span class="p">:</span>

    <span class="n">expand</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">VGAM</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">splines</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">DDRTree</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">irlba</span>

<span class="err">──</span> <span class="err"></span><span class="p">[</span><span class="mi">1</span><span class="n">mAttaching</span> <span class="n">packages</span><span class="err"></span><span class="p">[</span><span class="mi">22</span><span class="n">m</span> <span class="err">───────────────────────────────────────</span> <span class="n">tidyverse</span> <span class="mf">1.3</span><span class="o">.</span><span class="mi">0</span> <span class="err">──</span>

<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtibble</span> <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">2.1</span><span class="o">.</span><span class="mi">3</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">0.8</span><span class="o">.</span><span class="mi">3</span>
<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.0</span><span class="o">.</span><span class="mi">0</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mstringr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.4</span><span class="o">.</span><span class="mi">0</span>
<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mreadr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.3</span><span class="o">.</span><span class="mi">1</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mforcats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">0.4</span><span class="o">.</span><span class="mi">0</span>
<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mpurrr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">0.3</span><span class="o">.</span><span class="mi">3</span>

<span class="err">──</span> <span class="err"></span><span class="p">[</span><span class="mi">1</span><span class="n">mConflicts</span><span class="err"></span><span class="p">[</span><span class="mi">22</span><span class="n">m</span> <span class="err">──────────────────────────────────────────</span> <span class="n">tidyverse_conflicts</span><span class="p">()</span> <span class="err">──</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mcollapse</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>   <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">collapse</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mcombine</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>    <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mBiobase</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">combine</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mBiocGenerics</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">combine</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mcount</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>      <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mmatrixStats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">count</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mdesc</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>       <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">desc</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mexpand</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mMatrix</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">expand</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mS4Vectors</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">expand</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mfill</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>       <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mVGAM</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">fill</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mfilter</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mstats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">filter</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mfirst</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>      <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mS4Vectors</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">first</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mlag</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>        <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mstats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">lag</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mpack</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>       <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mMatrix</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">pack</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mggplot2</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mPosition</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mBiocGenerics</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">Position</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mbase</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">Position</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mpurrr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mreduce</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mGenomicRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">reduce</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">reduce</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mrename</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mS4Vectors</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">rename</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mpurrr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">msimplify</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>   <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mDelayedArray</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">simplify</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mslice</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>      <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">slice</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">munpack</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mMatrix</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">unpack</span><span class="p">()</span>
</code></pre></div>

<h2 id="download-data">Download data<a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Download data</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;./data&quot;</span><span class="p">))</span> <span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;./data&quot;</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;./data/neuron_1k_v3_fastqs.tar&quot;</span><span class="p">))</span> <span class="p">{</span>
  <span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;http://cf.10xgenomics.com/samples/cell-exp/3.0.0/neuron_1k_v3/neuron_1k_v3_fastqs.tar&quot;</span><span class="p">,</span> 
      <span class="s">&quot;./data/neuron_1k_v3_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;wget&quot;</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">untar</span><span class="p">(</span><span class="s">&quot;./data/neuron_1k_v3_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">exdir</span> <span class="o">=</span> <span class="s">&quot;./data&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h1 id="generate-the-gene-count-matrix">Generate the gene count matrix<a class="headerlink" href="#generate-the-gene-count-matrix" title="Permanent link">&para;</a></h1>
<h2 id="build-the-kallisto-index">Build the <code>kallisto</code> index<a class="headerlink" href="#build-the-kallisto-index" title="Permanent link">&para;</a></h2>
<p>Here we use <a href="https://pachterlab.github.io/kallisto/about">kallisto</a> to pseudoalign the reads to the transcriptome and then to create the <code>bus</code> file to be converted to a sparse matrix. The first step is to build an index of the mouse transcriptome. The transcriptome downloaded here is Ensembl version 99, the most recent version as of writing.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Mouse transcriptome</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;./reference&quot;</span><span class="p">))</span> <span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;./reference&quot;</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;./reference/mm_cdna99.fa.gz&quot;</span><span class="p">))</span> <span class="p">{</span>
  <span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;ftp://ftp.ensembl.org/pub/release-99/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz&quot;</span><span class="p">,</span> 
  <span class="s">&quot;./reference/mm_cdna99.fa.gz&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;wget&quot;</span><span class="p">,</span> <span class="n">quiet</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;kallisto index -i ./reference/mm_tr_index99.idx ./reference/mm_cdna99.fa.gz&quot;</span><span class="p">,</span> <span class="n">intern</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h4 id="map-transcripts-to-genes">Map transcripts to genes<a class="headerlink" href="#map-transcripts-to-genes" title="Permanent link">&para;</a></h4>
<p>For the sparse matrix, most people are interested in how many UMIs per gene per cell, we here we will quantify this from the <code>bus</code> output, and to do so, we need to find which gene corresponds to each transcript. Remember in the output of <code>kallisto bus</code>, there's the file <code>transcripts.txt</code>. Those are the transcripts in the transcriptome index. </p>
<p>Remember that we downloaded transcriptome FASTA file from Ensembl just now. In FASTA files, each entry is a sequence with a name. In Ensembl FASTA files, the sequence name has genome annotation of the corresponding sequence, so we can extract transcript IDs and corresponding gene IDs and gene names from there.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tr2g</span> <span class="o">&lt;-</span> <span class="nf">tr2g_fasta</span><span class="p">(</span><span class="s">&quot;./reference/mm_cdna99.fa.gz&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Reading FASTA file.
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">tr2g</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table>
<caption>A tibble: 6 × 3</caption>
<thead>
    <tr><th scope=col>transcript</th><th scope=col>gene</th><th scope=col>gene_name</th></tr>
    <tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
    <tr><td>ENSMUST00000177564.1</td><td>ENSMUSG00000096176.1</td><td>Trdd2  </td></tr>
    <tr><td>ENSMUST00000196221.1</td><td>ENSMUSG00000096749.2</td><td>Trdd1  </td></tr>
    <tr><td>ENSMUST00000179664.1</td><td>ENSMUSG00000096749.2</td><td>Trdd1  </td></tr>
    <tr><td>ENSMUST00000178537.1</td><td>ENSMUSG00000095668.1</td><td>Trbd1  </td></tr>
    <tr><td>ENSMUST00000178862.1</td><td>ENSMUSG00000094569.1</td><td>Trbd2  </td></tr>
    <tr><td>ENSMUST00000179520.1</td><td>ENSMUSG00000094028.1</td><td>Ighd4-1</td></tr>
</tbody>
</table>

<p><code>bustools</code> requires <code>tr2g</code> to be written into a tab delimited file of a specific format: No headers, first column is transcript ID, and second column is the corresponding gene ID. Transcript IDs must be in the same order as in the <code>kallisto</code> index.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Write tr2g to format required by bustools</span>
<span class="nf">save_tr2g_bustools</span><span class="p">(</span><span class="n">tr2g</span><span class="p">,</span> <span class="n">file_save</span> <span class="o">=</span> <span class="s">&quot;./reference/tr2g_mm99.tsv&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="using-the-kb-wrapper">Using the <code>kb</code> wrapper<a class="headerlink" href="#using-the-kb-wrapper" title="Permanent link">&para;</a></h2>
<p>With <code>kallisto</code> and <code>bustools</code>, it takes several commands to go from fastq files to the spliced and unspliced matrices, which is quite cumbersome. So a wrapper called <code>kb</code> was written to condense those steps to one. The command line tool <code>kb</code> can be installed with</p>
<p>Then we can use the following command to generate the spliced and unspliced matrices:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;chmod -R 777 data/&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">fn</span> <span class="o">&lt;-</span> <span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;data/neuron_1k_v3_fastqs&quot;</span><span class="p">,</span> <span class="n">full.names</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="n">fn</span> <span class="o">&lt;-</span> <span class="n">fn</span><span class="p">[</span><span class="nf">str_detect</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;R\\d_&quot;</span><span class="p">)]</span>
<span class="n">fn</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>'data/neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R1_001.fastq.gz'</li>
    <li>'data/neuron_1k_v3_fastqs/neuron_1k_v3_S1_L001_R2_001.fastq.gz'</li>
    <li>'data/neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R1_001.fastq.gz'</li>
    <li>'data/neuron_1k_v3_fastqs/neuron_1k_v3_S1_L002_R2_001.fastq.gz'</li>
</ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;kb count -i reference/mm_tr_index99.idx -g reference/tr2g_mm99.tsv -x 10xv3 -o output&quot;</span><span class="p">,</span>
<span class="nf">paste</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s">&quot; &quot;</span><span class="p">)),</span> <span class="n">intern</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h1 id="preprocessing">Preprocessing<a class="headerlink" href="#preprocessing" title="Permanent link">&para;</a></h1>
<p>Now we can load the matrix into R for analysis.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="nf">read_count_output</span><span class="p">(</span><span class="s">&quot;./output/counts_unfiltered&quot;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;cells_x_genes&quot;</span><span class="p">,</span> <span class="n">tcc</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="remove-empty-droplets">Remove empty droplets<a class="headerlink" href="#remove-empty-droplets" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>36711</li>
    <li>399524</li>
</ol>

<p>The number of genes seems reasonable. The number of barcodes is way larger than the expected ~10k.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tot_counts</span> <span class="o">&lt;-</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">tot_counts</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
     0.00      1.00      1.00     40.47      5.00 154216.00
</code></pre></div>

<p>The vast majority of "cells" have only no or just a few UMI detected. Those are empty droplets. 10x claims to have cell capture rate of up to 65%, but in practice, depending on how many cells are in fact loaded, the rate can be much lower. A commonly used method to estimate the number of empty droplets is barcode ranking knee and inflection points, as those are often assumed to represent transition between two components of a distribution. While more sophisticated methods exist (e.g. see <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#detecting-empty-droplets"><code>emptyDrops</code> in <code>DropletUtils</code></a>), for simplicity, we will use the barcode ranking method here. However, whichever way we go, we don't have the ground truth.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Compute barcode rank</span>
<span class="n">bc_rank</span> <span class="o">&lt;-</span> <span class="nf">barcodeRanks</span><span class="p">(</span><span class="n">res_mat</span><span class="p">,</span> <span class="n">lower</span> <span class="o">=</span> <span class="m">1000</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#&#39; Knee plot for filtering empty droplets</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; Visualizes the inflection point to filter empty droplets. This function plots </span>
<span class="c1">#&#39; different datasets with a different color. Facets can be added after calling</span>
<span class="c1">#&#39; this function with `facet_*` functions.</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; @param bc_rank A `DataFrame` output from `DropletUtil::barcodeRanks`.</span>
<span class="c1">#&#39; @return A ggplot2 object.</span>
<span class="n">knee_plot</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">knee_plt</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">rank</span> <span class="o">=</span> <span class="n">bc_rank</span><span class="p">[[</span><span class="s">&quot;rank&quot;</span><span class="p">]],</span>
                     <span class="n">total</span> <span class="o">=</span> <span class="n">bc_rank</span><span class="p">[[</span><span class="s">&quot;total&quot;</span><span class="p">]])</span> <span class="o">%&gt;%</span> 
    <span class="nf">distinct</span><span class="p">()</span> <span class="o">%&gt;%</span> 
    <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">total</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span>
  <span class="n">annot</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">inflection</span> <span class="o">=</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)[[</span><span class="s">&quot;inflection&quot;</span><span class="p">]],</span>
                  <span class="n">rank_cutoff</span> <span class="o">=</span> <span class="nf">max</span><span class="p">(</span><span class="n">bc_rank</span><span class="o">$</span><span class="n">rank</span><span class="p">[</span><span class="n">bc_rank</span><span class="o">$</span><span class="n">total</span> <span class="o">&gt;</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)[[</span><span class="s">&quot;inflection&quot;</span><span class="p">]]]))</span>
  <span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">knee_plt</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="n">total</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_line</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">geom_hline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="n">inflection</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_vline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="n">rank_cutoff</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">annotation_logticks</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Rank&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Total UMIs&quot;</span><span class="p">)</span>
  <span class="nf">return</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">knee_plot</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_32_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Filter the matrix</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[,</span> <span class="n">tot_counts</span> <span class="o">&gt;</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span><span class="o">$</span><span class="n">inflection</span><span class="p">]</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">,]</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<ol class=list-inline>
    <li>20198</li>
    <li>1210</li>
</ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">str_remove</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">),</span> <span class="s">&quot;\\.\\d+&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Now the number of cells is closer to expectation.</p>
<h2 id="cell-type-inference">Cell type inference<a class="headerlink" href="#cell-type-inference" title="Permanent link">&para;</a></h2>
<p>Monocle 2 only infers one trajectory for the entire dataset, so non-neuronal cells like endothelial cells and erythrocytes may be mistaken as highly differentiated cells from the neuronal lineage. So we will remove cell types not of the neural or glial lineages. Cell types are also helpful to orient the trajectory; neuronal progenitor cells must come before neurons. Here cell type inference is done programatically with <a href="https://github.com/dviraran/SingleR"><code>SingleR</code></a>, which compares gene expression profiles of individual cells to bulk RNA-seq data of purified known cell types.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">mouse.rnaseq</span> <span class="o">&lt;-</span> <span class="nf">MouseRNAseqData</span><span class="p">(</span><span class="n">ensembl</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="n">sce</span> <span class="o">&lt;-</span> <span class="nf">SingleCellExperiment</span><span class="p">(</span><span class="n">assays</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">res_mat</span><span class="p">))</span>
<span class="n">sce</span> <span class="o">&lt;-</span> <span class="nf">logNormCounts</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Then <code>SingleR</code> will assign each cell a label based on Spearman correlation with known cell types from bulk RNA-seq. These are meanings of the acronyms:</p>
<ul>
<li>OPCs: Oligodendrocyte progenitor cells</li>
<li>NPCs: Neural progenitor cells</li>
<li>aNSCs: Active neural stem cells</li>
<li>qNSCs: Quiescent neural stem cells</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">annots</span> <span class="o">&lt;-</span> <span class="nf">SingleR</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">ref</span> <span class="o">=</span> <span class="n">mouse.rnaseq</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="nf">colData</span><span class="p">(</span><span class="n">mouse.rnaseq</span><span class="p">)</span><span class="o">$</span><span class="n">label.fine</span><span class="p">,</span>
<span class="n">de.method</span> <span class="o">=</span> <span class="s">&quot;wilcox&quot;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;single&quot;</span><span class="p">,</span> <span class="n">BPPARAM</span> <span class="o">=</span> <span class="nf">MulticoreParam</span><span class="p">(</span><span class="m">4</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">annots</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">DataFrame</span> <span class="s s-Atom">with</span> <span class="mi">1210</span> <span class="s s-Atom">rows</span> <span class="s s-Atom">and</span> <span class="mi">5</span> <span class="s s-Atom">columns</span>
                                                                        <span class="s s-Atom">scores</span>
                                                                      <span class="o">&lt;</span><span class="s s-Atom">matrix</span><span class="o">&gt;</span>
<span class="nv">AAACGCTGTAATGTGA</span>    <span class="mf">0.0601901483760862</span><span class="s s-Atom">:</span><span class="mf">0.242986808015721</span><span class="s s-Atom">:</span><span class="mf">0.177660752587073</span><span class="s s-Atom">:...</span>
<span class="nv">AAACGCTGTCCTGGGT</span>      <span class="mf">0.11153546550616</span><span class="s s-Atom">:</span><span class="mf">0.321858508786419</span><span class="s s-Atom">:</span><span class="mf">0.206383027494298</span><span class="s s-Atom">:...</span>
<span class="nv">AAAGAACCAGGACATG</span>    <span class="mf">0.0585569289411892</span><span class="s s-Atom">:</span><span class="mf">0.225038532259566</span><span class="s s-Atom">:</span><span class="mf">0.177381655702582</span><span class="s s-Atom">:...</span>
<span class="nv">AAAGGTACACACGGTC</span>       <span class="mf">0.137923692580242</span><span class="s s-Atom">:</span><span class="mf">0.3506611435018</span><span class="s s-Atom">:</span><span class="mf">0.281933499704427</span><span class="s s-Atom">:...</span>
<span class="nv">AAAGGTATCACCATAG</span> <span class="o">-</span><span class="mf">0.00511350381259525</span><span class="s s-Atom">:</span><span class="mf">0.166880227412372</span><span class="s s-Atom">:</span><span class="mf">0.0975899954539976</span><span class="s s-Atom">:...</span>
<span class="p">...</span>                                                                        <span class="p">...</span>
<span class="nv">TTTGACTTCGTTCAGA</span>     <span class="mf">0.100933655338997</span><span class="s s-Atom">:</span><span class="mf">0.327948884202836</span><span class="s s-Atom">:</span><span class="mf">0.262011719832591</span><span class="s s-Atom">:...</span>
<span class="nv">TTTGATCTCCATAGGT</span>     <span class="mf">0.0656536744266451</span><span class="s s-Atom">:</span><span class="mf">0.280702755051776</span><span class="s s-Atom">:</span><span class="mf">0.23375860874247</span><span class="s s-Atom">:...</span>
<span class="nv">TTTGGAGAGGCTAACG</span>    <span class="mf">0.0681422805975054</span><span class="s s-Atom">:</span><span class="mf">0.234772340818692</span><span class="s s-Atom">:</span><span class="mf">0.176292333613646</span><span class="s s-Atom">:...</span>
<span class="nv">TTTGGTTAGTAATCCC</span>     <span class="mf">0.125780004617712</span><span class="s s-Atom">:</span><span class="mf">0.315213344287222</span><span class="s s-Atom">:</span><span class="mf">0.258722018820061</span><span class="s s-Atom">:...</span>
<span class="nv">TTTGTTGGTATGGAAT</span>    <span class="mf">0.0512710432754663</span><span class="s s-Atom">:</span><span class="mf">0.231028330364938</span><span class="s s-Atom">:</span><span class="mf">0.177598025280379</span><span class="s s-Atom">:...</span>
                 <span class="s s-Atom">first</span><span class="p">.</span><span class="s s-Atom">labels</span>                        <span class="s s-Atom">tuning</span><span class="p">.</span><span class="s s-Atom">scores</span>      <span class="s s-Atom">labels</span>
                  <span class="o">&lt;</span><span class="s s-Atom">character</span><span class="o">&gt;</span>                          <span class="o">&lt;</span><span class="nv">DataFrame</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="s s-Atom">character</span><span class="o">&gt;</span>
<span class="nv">AAACGCTGTAATGTGA</span>        <span class="s s-Atom">aNSCs</span>  <span class="mf">0.233804530306167</span><span class="s s-Atom">:</span><span class="mf">0.219682141602258</span>       <span class="s s-Atom">aNSCs</span>
<span class="nv">AAACGCTGTCCTGGGT</span>         <span class="nv">NPCs</span>                                  <span class="mi">1</span><span class="s s-Atom">:</span><span class="mi">1</span>       <span class="s s-Atom">aNSCs</span>
<span class="nv">AAAGAACCAGGACATG</span>      <span class="nv">Neurons</span>                                  <span class="mi">1</span><span class="s s-Atom">:</span><span class="mi">1</span>       <span class="s s-Atom">aNSCs</span>
<span class="nv">AAAGGTACACACGGTC</span>         <span class="nv">NPCs</span>                                  <span class="mi">1</span><span class="s s-Atom">:</span><span class="mi">1</span>       <span class="s s-Atom">aNSCs</span>
<span class="nv">AAAGGTATCACCATAG</span>         <span class="nv">NPCs</span>                              <span class="mf">0.5</span><span class="s s-Atom">:</span><span class="mf">0.5</span>       <span class="s s-Atom">aNSCs</span>
<span class="p">...</span>                       <span class="p">...</span>                                  <span class="p">...</span>         <span class="p">...</span>
<span class="nv">TTTGACTTCGTTCAGA</span>      <span class="nv">Neurons</span>                                  <span class="mi">1</span><span class="s s-Atom">:</span><span class="mi">1</span>       <span class="s s-Atom">aNSCs</span>
<span class="nv">TTTGATCTCCATAGGT</span>      <span class="nv">Neurons</span>                              <span class="mf">0.5</span><span class="s s-Atom">:</span><span class="mf">0.5</span>       <span class="s s-Atom">aNSCs</span>
<span class="nv">TTTGGAGAGGCTAACG</span>        <span class="s s-Atom">aNSCs</span> <span class="mf">0.209611614011623</span><span class="s s-Atom">:</span><span class="mf">0.0580258853185659</span>     <span class="nv">Neurons</span>
<span class="nv">TTTGGTTAGTAATCCC</span>      <span class="nv">Neurons</span> <span class="mf">0.187703782257732</span><span class="p">:-</span><span class="mf">0.195976600339204</span>     <span class="nv">Neurons</span>
<span class="nv">TTTGTTGGTATGGAAT</span>         <span class="nv">OPCs</span> <span class="mf">0.299839391397521</span><span class="p">:-</span><span class="mf">0.199361864824359</span>     <span class="nv">Neurons</span>
                 <span class="s s-Atom">pruned</span><span class="p">.</span><span class="s s-Atom">labels</span>
                   <span class="o">&lt;</span><span class="s s-Atom">character</span><span class="o">&gt;</span>
<span class="nv">AAACGCTGTAATGTGA</span>         <span class="s s-Atom">aNSCs</span>
<span class="nv">AAACGCTGTCCTGGGT</span>         <span class="s s-Atom">aNSCs</span>
<span class="nv">AAAGAACCAGGACATG</span>         <span class="s s-Atom">aNSCs</span>
<span class="nv">AAAGGTACACACGGTC</span>         <span class="s s-Atom">aNSCs</span>
<span class="nv">AAAGGTATCACCATAG</span>         <span class="s s-Atom">aNSCs</span>
<span class="p">...</span>                        <span class="p">...</span>
<span class="nv">TTTGACTTCGTTCAGA</span>         <span class="s s-Atom">aNSCs</span>
<span class="nv">TTTGATCTCCATAGGT</span>         <span class="s s-Atom">aNSCs</span>
<span class="nv">TTTGGAGAGGCTAACG</span>       <span class="nv">Neurons</span>
<span class="nv">TTTGGTTAGTAATCCC</span>       <span class="nv">Neurons</span>
<span class="nv">TTTGTTGGTATGGAAT</span>       <span class="nv">Neurons</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">inds</span> <span class="o">&lt;-</span> <span class="n">annots</span><span class="o">$</span><span class="n">pruned.labels</span> <span class="o">%in%</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;NPCs&quot;</span><span class="p">,</span> <span class="s">&quot;Neurons&quot;</span><span class="p">,</span> <span class="s">&quot;OPCs&quot;</span><span class="p">,</span> <span class="s">&quot;Oligodendrocytes&quot;</span><span class="p">,</span> 
                                    <span class="s">&quot;qNSCs&quot;</span><span class="p">,</span> <span class="s">&quot;aNSCs&quot;</span><span class="p">,</span> <span class="s">&quot;Astrocytes&quot;</span><span class="p">,</span> <span class="s">&quot;Ependymal&quot;</span><span class="p">)</span>
<span class="c1"># Only keep these cell types</span>
<span class="n">cells_use</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">annots</span><span class="p">)[</span><span class="n">inds</span><span class="p">]</span>
<span class="n">sce</span> <span class="o">&lt;-</span> <span class="n">sce</span><span class="p">[,</span> <span class="n">cells_use</span><span class="p">]</span>
<span class="n">sce</span><span class="o">$</span><span class="n">cell_type</span> <span class="o">&lt;-</span> <span class="n">annots</span><span class="o">$</span><span class="n">pruned.labels</span><span class="p">[</span><span class="n">inds</span><span class="p">]</span>
</code></pre></div>
</td></tr></table>
<h2 id="qc">QC<a class="headerlink" href="#qc" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="o">&lt;-</span> <span class="nf">perCellQCMetrics</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">colData</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">cbind</span><span class="p">(</span><span class="nf">colData</span><span class="p">(</span><span class="n">sce</span><span class="p">),</span> <span class="n">df</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">colData</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>DataFrame with 1146 rows and 8 columns
                   cell_type       sum  detected   percent_top_50
                 &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;        &lt;numeric&gt;
AAACGCTGTAATGTGA       aNSCs      7072      3256 19.0328054298643
AAACGCTGTCCTGGGT       aNSCs      9458      4101 12.8991330090928
AAAGAACCAGGACATG       aNSCs      5338      2577 20.7006369426752
AAAGGTACACACGGTC       aNSCs     16892      5111 18.2098034572579
AAAGGTATCACCATAG       aNSCs      1397       882 24.9105225483178
...                      ...       ...       ...              ...
TTTGACTTCGTTCAGA       aNSCs     26203      6091 21.4021295271534
TTTGATCTCCATAGGT       aNSCs      8120      3625 16.7980295566502
TTTGGAGAGGCTAACG     Neurons      5483      2737 21.8675907349991
TTTGGTTAGTAATCCC     Neurons     24381      5927 20.2288667404946
TTTGTTGGTATGGAAT     Neurons      3718      2324 14.6046261430877
                  percent_top_100  percent_top_200  percent_top_500     total
                        &lt;numeric&gt;        &lt;numeric&gt;        &lt;numeric&gt; &lt;numeric&gt;
AAACGCTGTAATGTGA 25.8484162895928 34.1911764705882  49.335407239819      7072
AAACGCTGTCCTGGGT  19.221822795517 27.6485514908014 42.5882850496934      9458
AAAGAACCAGGACATG 27.5196702884976 36.2495316597977 52.6227051330086      5338
AAAGGTACACACGGTC 26.5095903386218 35.6973715368222 49.5204830689084     16892
AAAGGTATCACCATAG 35.5762347888332 49.8926270579814  72.655690765927      1397
...                           ...              ...              ...       ...
TTTGACTTCGTTCAGA 28.4814715872228  37.358317749876 50.7422814181582     26203
TTTGATCTCCATAGGT 23.5098522167488 32.0320197044335 47.2044334975369      8120
TTTGGAGAGGCTAACG  28.269195695787 36.6405252598942 52.0335582710195      5483
TTTGGTTAGTAATCCC 27.7183052376851 36.2659447930766 49.9733398958205     24381
TTTGTTGGTATGGAAT 20.6293706293706 29.4513179128564 46.9338353953739      3718
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plotColData</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;sum&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_45_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plotColData</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;detected&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_46_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plotColData</span><span class="p">(</span><span class="n">sce</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s">&quot;sum&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;detected&quot;</span><span class="p">,</span> <span class="n">colour_by</span> <span class="o">=</span> <span class="s">&quot;cell_type&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">annotation_logticks</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_47_0.png" /></p>
<h1 id="monocle-2">Monocle 2<a class="headerlink" href="#monocle-2" title="Permanent link">&para;</a></h1>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Construct CellDataSet object</span>
<span class="n">pd</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">cell_id</span> <span class="o">=</span> <span class="n">cells_use</span><span class="p">,</span> 
                 <span class="n">cell_type</span> <span class="o">=</span> <span class="n">annots</span><span class="o">$</span><span class="n">pruned.labels</span><span class="p">[</span><span class="n">inds</span><span class="p">],</span>
                 <span class="n">row.names</span> <span class="o">=</span> <span class="n">cells_use</span><span class="p">)</span>
<span class="n">pd</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;AnnotatedDataFrame&quot;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="p">)</span>
<span class="n">fd</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">sce</span><span class="p">),</span> 
                 <span class="n">gene_short_name</span> <span class="o">=</span> <span class="n">tr2g</span><span class="o">$</span><span class="n">gene_name</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">sce</span><span class="p">),</span> <span class="n">tr2g</span><span class="o">$</span><span class="n">gene</span><span class="p">)],</span>
                 <span class="n">row.names</span> <span class="o">=</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">sce</span><span class="p">))</span>
<span class="n">fd</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;AnnotatedDataFrame&quot;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">fd</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">newCellDataSet</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">sce</span><span class="p">),</span> <span class="n">phenoData</span> <span class="o">=</span> <span class="n">pd</span><span class="p">,</span> <span class="n">featureData</span> <span class="o">=</span> <span class="n">fd</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Size factor and dispersion will be used to normalize data and select genes for clustering.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">estimateSizeFactors</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">estimateDispersions</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Removing 16 outliers
</code></pre></div>

<p>Genes that aren't highly expressed enough will not be used for clustering, since they may not give meaningful signal and would only add noise.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">disp_table</span> <span class="o">&lt;-</span> <span class="nf">dispersionTable</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
<span class="n">clustering_genes</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">disp_table</span><span class="p">,</span> <span class="n">mean_expression</span> <span class="o">&gt;=</span> <span class="m">0.1</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">setOrderingFilter</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">clustering_genes</span><span class="o">$</span><span class="n">gene_id</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">reduceDimension</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">num_dim</span> <span class="o">=</span> <span class="m">40</span><span class="p">,</span> <span class="n">reduction_method</span> <span class="o">=</span> <span class="s">&#39;tSNE&#39;</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">clusterCells</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;louvain&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plot_cell_clusters</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;tSNE1&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;tSNE2&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_55_0.png" /></p>
<p>See where the annotated cell types are</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plot_cell_clusters</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">color_by</span> <span class="o">=</span> <span class="s">&quot;cell_type&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_color_brewer</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;cell type&quot;</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;qual&quot;</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="s">&quot;Set2&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;tSNE1&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;tSNE2&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;right&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">guides</span><span class="p">(</span><span class="n">color</span> <span class="o">=</span> <span class="nf">guide_legend</span><span class="p">(</span><span class="n">override.aes</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">3</span><span class="p">)))</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_57_0.png" /></p>
<p>Genes likely to be informative of ordering of cells along the pseudotime trajectory will be selected for pseudotime inference.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">diff_genes</span> <span class="o">&lt;-</span> <span class="nf">differentialGeneTest</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">fullModelFormulaStr</span> <span class="o">=</span> <span class="s">&quot;~ Cluster + cell_type&quot;</span><span class="p">,</span>
                                   <span class="n">cores</span> <span class="o">=</span> <span class="m">4</span><span class="p">)</span>
<span class="c1"># Use top 3000 differentially expressed genes</span>
<span class="n">ordering_genes</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="nf">subset</span><span class="p">(</span><span class="n">diff_genes</span><span class="p">,</span> <span class="n">qval</span> <span class="o">&lt;</span> <span class="m">1e-3</span><span class="p">))[</span><span class="nf">order</span><span class="p">(</span><span class="n">diff_genes</span><span class="o">$</span><span class="n">qval</span><span class="p">)][</span><span class="m">1</span><span class="o">:</span><span class="m">3000</span><span class="p">]</span>
<span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">setOrderingFilter</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">ordering_genes</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Here Monocle 2 will first project the data to 2 dimensions with <code>DDRTree</code>, and then do trajectory inference (<code>orderCells</code>).</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">reduceDimension</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">max_components</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&#39;DDRTree&#39;</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">orderCells</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>See what the trajectory looks like. This projection is <code>DDRTree</code>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plot_cell_trajectory</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">color_by</span> <span class="o">=</span> <span class="s">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">cell_size</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_color_brewer</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;cell type&quot;</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;qual&quot;</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="s">&quot;Set2&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_63_0.png" /></p>
<p>In the <a href="https://www.biorxiv.org/content/10.1101/673285v1">kallisto | bustools paper</a>, I used <code>slingshot</code> for pseudotime analysis (Supplementary Figure 6.5) of this dataset, and found two neuronal end points. The result from Monocle 2 here also shows two main branches. Also, as expected, the stem cells are at the very beginning of the trajectory.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plot_cell_trajectory</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">color_by</span> <span class="o">=</span> <span class="s">&quot;Pseudotime&quot;</span><span class="p">,</span> <span class="n">cell_size</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_color_viridis_c</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_65_0.png" /></p>
<p>The pseudotime values are inverted.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">orderCells</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">reverse</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">plot_cell_trajectory</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">color_by</span> <span class="o">=</span> <span class="s">&quot;Pseudotime&quot;</span><span class="p">,</span> <span class="n">cell_size</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_color_viridis_c</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_monocle2_files/kb_monocle2_67_0.png" /></p>
<p>Monocle 2 can also be used to find genes differentially expressed along the pseudotime trajectory and clusters of such genes. See <a href="https://davetang.org/muse/2017/10/01/getting-started-monocle/">David Tang's excellent Monocle 2 tutorial</a> for how to use these functionalities.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        <a href="../../../../kb_species_mixing/R/kb_mixed_species_10x_v2/" class="md-footer__link md-footer__link--prev" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              R
            </div>
          </div>
        </a>
      
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../../../../../assets/javascripts/workers/search.fe42c31b.min.js", "version": null}</script>
    
    
      <script src="../../../../../assets/javascripts/bundle.65ce87ac.min.js"></script>
      
    
  </body>
</html>