
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.1, mkdocs-material-7.1.9">
    
    
      
        <title>R - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.ca7ac06f.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.f1a3b89f.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../css/custom.css">
    
    
      
  



  
  



  <script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
  <script async src="https://www.google-analytics.com/analytics.js"></script>


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    <script>function __prefix(e){return new URL("../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#analysis-of-single-cell-rna-seq-data-building-and-annotating-an-atlas" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../index.html" title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              R
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../index.html" title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../index.html" class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../applications.html" class="md-nav__link">
        Applications
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Using kb
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_usage.html" class="md-nav__link">
        Introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_ref.html" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_count.html" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      <label class="md-nav__link" for="__nav_4">
        Tutorials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_1" type="checkbox" id="__nav_4_1" >
      
      <label class="md-nav__link" for="__nav_4_1">
        Introduction to single-cell RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../scRNA-seq_intro/python/scRNA-seq_intro.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_2" type="checkbox" id="__nav_4_2" >
      
      <label class="md-nav__link" for="__nav_4_2">
        Preprocessing & Quality Control
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/python/kb_intro_1_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/R/kb_intro_1_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_3" type="checkbox" id="__nav_4_3" >
      
      <label class="md-nav__link" for="__nav_4_3">
        Getting started with analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_getting_started/python/kb_intro_2_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_getting_started/R/kb_intro_2_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" checked>
      
      <label class="md-nav__link" for="__nav_4_4">
        Building & annotating an atlas
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../python/kb_analysis_0_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          R
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="kb_analysis_0_R.html" class="md-nav__link md-nav__link--active">
        R
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-r-packages" class="md-nav__link">
    Install R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb-python" class="md-nav__link">
    Install kb-python
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-an-index" class="md-nav__link">
    Download an index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-for-library-saturation" class="md-nav__link">
    Test for library saturation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filter" class="md-nav__link">
    Filter
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#normalize" class="md-nav__link">
    Normalize
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#highly-expressed-genes" class="md-nav__link">
    Highly expressed genes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scaling-the-data" class="md-nav__link">
    Scaling the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#principal-component-analysis" class="md-nav__link">
    Principal component analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#determining-dimensionality" class="md-nav__link">
    Determining dimensionality
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-neighborhood-graph" class="md-nav__link">
    The neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umap-and-t-sne" class="md-nav__link">
    UMAP and t-SNE
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#finding-differentially-expressed-features-cluster-biomarkers" class="md-nav__link">
    Finding differentially expressed features (cluster biomarkers)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assigning-cell-type-identity-to-clusters" class="md-nav__link">
    Assigning cell type identity to clusters
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      <label class="md-nav__link" for="__nav_4_5">
        Pseudotime analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/slingshot/kb_slingshot.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      <label class="md-nav__link" for="__nav_4_6">
        Single-nuclei RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_nucleus/python/kb_single_nucleus.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      <label class="md-nav__link" for="__nav_4_7">
        RNA Velocity
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity/python/kb_velocity.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" >
      
      <label class="md-nav__link" for="__nav_4_8">
        Feature barcode & multimodal data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_kite/python/kb_kite.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../download_data/data_download.html" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      <label class="md-nav__link" for="__nav_4_10">
        Build custom index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_custom_index/python/kb_transcriptome_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_11" type="checkbox" id="__nav_4_11" >
      
      <label class="md-nav__link" for="__nav_4_11">
        Build RNA velocity index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity_index/python/kb_velocity_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_12" type="checkbox" id="__nav_4_12" >
      
      <label class="md-nav__link" for="__nav_4_12">
        Aggregate multiple matrices
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_aggregate/python/kb_aggregating_count_matrices.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_13" type="checkbox" id="__nav_4_13" >
      
      <label class="md-nav__link" for="__nav_4_13">
        Process multiple FASTQs
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_multiple_fastqs/python/kb_multiple_files.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_14" type="checkbox" id="__nav_4_14" >
      
      <label class="md-nav__link" for="__nav_4_14">
        Processing multi-species data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_species_mixing/python/kb_species_mixing.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_species_mixing/R/kb_mixed_species_10x_v2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_15" type="checkbox" id="__nav_4_15" >
      
      <label class="md-nav__link" for="__nav_4_15">
        Pseudotime analysis (Monocle2)
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/monocle2/kb_monocle2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-r-packages" class="md-nav__link">
    Install R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb-python" class="md-nav__link">
    Install kb-python
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-an-index" class="md-nav__link">
    Download an index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-for-library-saturation" class="md-nav__link">
    Test for library saturation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filter" class="md-nav__link">
    Filter
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#normalize" class="md-nav__link">
    Normalize
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#highly-expressed-genes" class="md-nav__link">
    Highly expressed genes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scaling-the-data" class="md-nav__link">
    Scaling the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#principal-component-analysis" class="md-nav__link">
    Principal component analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#determining-dimensionality" class="md-nav__link">
    Determining dimensionality
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-neighborhood-graph" class="md-nav__link">
    The neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umap-and-t-sne" class="md-nav__link">
    UMAP and t-SNE
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#finding-differentially-expressed-features-cluster-biomarkers" class="md-nav__link">
    Finding differentially expressed features (cluster biomarkers)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assigning-cell-type-identity-to-clusters" class="md-nav__link">
    Assigning cell type identity to clusters
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/tree/master/tutorials/docs/tutorials/kb_building_atlas/R/kb_analysis_0_R.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/tutorials/docs/tutorials/kb_building_atlas/R/kb_analysis_0_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="analysis-of-single-cell-rna-seq-data-building-and-annotating-an-atlas">Analysis of single-cell RNA-seq data: building and annotating an atlas<a class="headerlink" href="#analysis-of-single-cell-rna-seq-data-building-and-annotating-an-atlas" title="Permanent link">&para;</a></h1>
<p>This R notebook pre-processes the <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3">pbmc_1k v3 dataset</a> from 10X Genomics with kallisto and bustools using <code>kb</code>, and then performs an analysis of the cell types and their marker genes.</p>
<p>The notebook was written by A. Sina Booeshaghi, Lambda Lu and Lior Pachter and is based on three noteboks:
- The kallisto | bustools <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_1_minute_intro.ipynb#scrollTo=wtwMjIjjCMcD">Introduction to single-cell RNA-seq I</a> notebook.
- The kallisto | bustools <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_standard.ipynb#scrollTo=ijU_u6uj3Sio">Introduction to single-cell RNA-seq II</a> notebook.
- The Seurat <a href="https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html">Guided Clustering Tutorial</a>.</p>
<p>If you use the methods in this notebook for your analysis please cite the following publications which describe the tools used in the notebook:</p>
<ul>
<li>Melsted, P., Booeshaghi, A.S. et al. Modular and efficient pre-processing of single-cell RNA-seq. bioRxiv (2019). doi:10.1101/673285</li>
<li>Stuart, Butler et al. Comprehensive Integration of Single-cell Data. Cell (2019). doi:10.1016/j.cell.2019.05.031</li>
</ul>
<p>A Python notebook implementing the same analysis is available <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_analysis_0_python.ipynb#scrollTo=3kZG9UyUPE_B">here</a>. See the <a href="https://www.kallistobus.tools/tutorials">kallistobus.tools tutorials</a> site for additional notebooks demonstrating other analyses.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># This is used to time the running of the notebook</span>
<span class="n">start_time</span> <span class="o">&lt;-</span> <span class="nf">Sys.time</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<h3 id="install-r-packages">Install R packages<a class="headerlink" href="#install-r-packages" title="Permanent link">&para;</a></h3>
<p>A large fraction of the running time of this notebook is in installing the Seurat R package, since it has lots of dependencies and many of them use Rcpp which results in the need to compile lots of C++ code. Compilation is required because CRAN does not distribute binaries for Linux, which is the operating system here.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system.time</span><span class="p">({</span>
  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;Seurat&quot;</span><span class="p">,</span> <span class="n">Ncpus</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
<span class="p">})</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Installing</span> <span class="n">package</span> <span class="n">into</span> <span class="err">‘</span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">library</span><span class="err">’</span>
<span class="p">(</span><span class="k">as</span> <span class="err">‘</span><span class="n">lib</span><span class="err">’</span> <span class="k">is</span> <span class="n">unspecified</span><span class="p">)</span>

<span class="n">also</span> <span class="n">installing</span> <span class="n">the</span> <span class="n">dependencies</span> <span class="err">‘</span><span class="n">bitops</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gtools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">caTools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sass</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">jquerylib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sitmo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">globals</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">listenv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">parallelly</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plyr</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">zoo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gplots</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">reshape2</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gridExtra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppArmadillo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">httpuv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">xtable</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sourcetools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bslib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">data</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">utils</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">sparse</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">abind</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">tensor</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">goftest</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">deldir</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">polyclip</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">FNN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RSpectra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">dqrng</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">cowplot</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">fitdistrplus</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">future</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">future</span><span class="o">.</span><span class="n">apply</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggrepel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggridges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ica</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">igraph</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">irlba</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">leiden</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lmtest</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">matrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">miniUI</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">patchwork</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">pbapply</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plotly</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">png</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RANN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppAnnoy</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">reticulate</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ROCR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rtsne</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">scattermore</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sctransform</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SeuratObject</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">shiny</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">core</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">geom</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">uwot</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppEigen</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppProgress</span><span class="err">’</span>





    <span class="n">user</span>   <span class="n">system</span>  <span class="n">elapsed</span> 
<span class="mf">1923.767</span>  <span class="mf">192.391</span> <span class="mf">1168.018</span>
</code></pre></div>

<p>The package installation took 20 minutes (elapsed) which is 40% of the running time of the entire notebook. The user time is nearly twice the elapsed time here because 2 cores were used to install all those packages.</p>
<h3 id="install-kb-python">Install kb-python<a class="headerlink" href="#install-kb-python" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Install kb (includes installing kallisto and bustools)</span>
<span class="nf">system</span><span class="p">(</span><span class="s">&quot;pip3 install kb-python&quot;</span><span class="p">,</span> <span class="n">intern</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Collecting kb-python'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/29/94/855ed1c11110a65a466cd95a6fef64958bad055f2678270b80a32e42cdb1/kb_python-0.25.1-py3-none-any.whl (59.1MB)'</span></li><li>'Requirement already satisfied: tqdm&gt;=4.39.0 in /usr/local/lib/python3.7/dist-packages (from kb-python) (4.41.1)'</li><li>'Collecting loompy&gt;=3.0.6'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/36/52/74ed37ae5988522fbf87b856c67c4f80700e6452410b4cd80498c5f416f9/loompy-3.0.6.tar.gz (41kB)'</span></li><li>'Requirement already satisfied: nbformat&gt;=4.4.0 in /usr/local/lib/python3.7/dist-packages (from kb-python) (5.1.2)'</li><li>'Requirement already satisfied: h5py&gt;=2.10.0 in /usr/local/lib/python3.7/dist-packages (from kb-python) (2.10.0)'</li><li>'Requirement already satisfied: requests&gt;=2.19.0 in /usr/local/lib/python3.7/dist-packages (from kb-python) (2.23.0)'</li><li>'Requirement already satisfied: nbconvert&gt;=5.6.0 in /usr/local/lib/python3.7/dist-packages (from kb-python) (5.6.1)'</li><li>'Requirement already satisfied: numpy&gt;=1.17.2 in /usr/local/lib/python3.7/dist-packages (from kb-python) (1.19.5)'</li><li>'Collecting anndata&gt;=0.6.22.post1'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/81/b1/743cc79f89d9db6dccbfb7e6000795acb218a6c6320b7a2337cad99bd047/anndata-0.7.5-py3-none-any.whl (119kB)'</span></li><li>'Collecting scanpy&gt;=1.4.4.post1'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/c1/f4/a7848e6f990cf5bcbedade93702baf3e99ae704714563fe9bdceb3d597c7/scanpy-1.7.1-py3-none-any.whl (10.3MB)'</span></li><li>'Requirement already satisfied: Jinja2&gt;2.10.1 in /usr/local/lib/python3.7/dist-packages (from kb-python) (2.11.3)'</li><li>'Collecting plotly&gt;=4.5.0'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/1f/f6/bd3c17c8003b6641df1228e80e1acac97ed8402635e46c2571f8e1ef63af/plotly-4.14.3-py2.py3-none-any.whl (13.2MB)'</span></li><li>'Requirement already satisfied: scikit-learn&gt;=0.21.3 in /usr/local/lib/python3.7/dist-packages (from kb-python) (0.22.2.post1)'</li><li>'Requirement already satisfied: scipy in /usr/local/lib/python3.7/dist-packages (from loompy&gt;=3.0.6-&gt;kb-python) (1.4.1)'</li><li>'Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from loompy&gt;=3.0.6-&gt;kb-python) (54.2.0)'</li><li>'Requirement already satisfied: numba in /usr/local/lib/python3.7/dist-packages (from loompy&gt;=3.0.6-&gt;kb-python) (0.51.2)'</li><li>'Requirement already satisfied: click in /usr/local/lib/python3.7/dist-packages (from loompy&gt;=3.0.6-&gt;kb-python) (7.1.2)'</li><li>'Collecting numpy-groupies'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/99/eb/fc72b507219957cffdf2c5952e396cc04a30c2223e2fd789f4a744ffc52f/numpy_groupies-0.9.13.tar.gz (109kB)'</span></li><li>'Requirement already satisfied: jsonschema!=2.5.0,&gt;=2.4 in /usr/local/lib/python3.7/dist-packages (from nbformat&gt;=4.4.0-&gt;kb-python) (2.6.0)'</li><li>'Requirement already satisfied: traitlets&gt;=4.1 in /usr/local/lib/python3.7/dist-packages (from nbformat&gt;=4.4.0-&gt;kb-python) (5.0.5)'</li><li>'Requirement already satisfied: jupyter-core in /usr/local/lib/python3.7/dist-packages (from nbformat&gt;=4.4.0-&gt;kb-python) (4.7.1)'</li><li>'Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.7/dist-packages (from nbformat&gt;=4.4.0-&gt;kb-python) (0.2.0)'</li><li>'Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from h5py&gt;=2.10.0-&gt;kb-python) (1.15.0)'</li><li>'Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,&lt;1.26,&gt;=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests&gt;=2.19.0-&gt;kb-python) (1.24.3)'</li><li>'Requirement already satisfied: idna&lt;3,&gt;=2.5 in /usr/local/lib/python3.7/dist-packages (from requests&gt;=2.19.0-&gt;kb-python) (2.10)'</li><li>'Requirement already satisfied: certifi&gt;=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests&gt;=2.19.0-&gt;kb-python) (2020.12.5)'</li><li>'Requirement already satisfied: chardet&lt;4,&gt;=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests&gt;=2.19.0-&gt;kb-python) (3.0.4)'</li><li>'Requirement already satisfied: entrypoints&gt;=0.2.2 in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (0.3)'</li><li>'Requirement already satisfied: pygments in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (2.6.1)'</li><li>'Requirement already satisfied: bleach in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (3.3.0)'</li><li>'Requirement already satisfied: pandocfilters&gt;=1.4.1 in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (1.4.3)'</li><li>'Requirement already satisfied: testpath in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (0.4.4)'</li><li>'Requirement already satisfied: mistune&lt;2,&gt;=0.8.1 in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (0.8.4)'</li><li>'Requirement already satisfied: defusedxml in /usr/local/lib/python3.7/dist-packages (from nbconvert&gt;=5.6.0-&gt;kb-python) (0.7.1)'</li><li>'Requirement already satisfied: packaging in /usr/local/lib/python3.7/dist-packages (from anndata&gt;=0.6.22.post1-&gt;kb-python) (20.9)'</li><li>'Requirement already satisfied: importlib-metadata&gt;=0.7; python_version &lt; "3.8" in /usr/local/lib/python3.7/dist-packages (from anndata&gt;=0.6.22.post1-&gt;kb-python) (3.8.1)'</li><li>'Requirement already satisfied: pandas!=1.1,&gt;=1.0 in /usr/local/lib/python3.7/dist-packages (from anndata&gt;=0.6.22.post1-&gt;kb-python) (1.1.5)'</li><li>'Requirement already satisfied: natsort in /usr/local/lib/python3.7/dist-packages (from anndata&gt;=0.6.22.post1-&gt;kb-python) (5.5.0)'</li><li>'Collecting umap-learn&gt;=0.3.10'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/75/69/85e7f950bb75792ad5d666d86c5f3e62eedbb942848e7e3126513af9999c/umap-learn-0.5.1.tar.gz (80kB)'</span></li><li>'Requirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (1.0.1)'</li><li>'Requirement already satisfied: seaborn in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (0.11.1)'</li><li>'Requirement already satisfied: statsmodels&gt;=0.10.0rc2 in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (0.10.2)'</li><li>'Requirement already satisfied: patsy in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (0.5.1)'</li><li>'Requirement already satisfied: matplotlib&gt;=3.1.2 in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (3.2.2)'</li><li>'Collecting sinfo'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/e1/4c/aef8456284f1a1c3645b938d9ca72388c9c4878e6e67b8a349c7d22fac78/sinfo-0.3.1.tar.gz'</span></li><li>'Collecting legacy-api-wrap'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/a4/68/da997bc56bb69dcdcee4054f0bc42266909307b905389fbc54c9158f42da/legacy_api_wrap-1.2-py3-none-any.whl'</span></li><li>'Requirement already satisfied: tables in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (3.4.4)'</li><li>'Requirement already satisfied: networkx&gt;=2.3 in /usr/local/lib/python3.7/dist-packages (from scanpy&gt;=1.4.4.post1-&gt;kb-python) (2.5)'</li><li>'Requirement already satisfied: MarkupSafe&gt;=0.23 in /usr/local/lib/python3.7/dist-packages (from Jinja2&gt;2.10.1-&gt;kb-python) (1.1.1)'</li><li>'Requirement already satisfied: retrying&gt;=1.3.3 in /usr/local/lib/python3.7/dist-packages (from plotly&gt;=4.5.0-&gt;kb-python) (1.3.3)'</li><li>'Requirement already satisfied: llvmlite&lt;0.35,&gt;=0.34.0.dev0 in /usr/local/lib/python3.7/dist-packages (from numba-&gt;loompy&gt;=3.0.6-&gt;kb-python) (0.34.0)'</li><li>'Requirement already satisfied: webencodings in /usr/local/lib/python3.7/dist-packages (from bleach-&gt;nbconvert&gt;=5.6.0-&gt;kb-python) (0.5.1)'</li><li>'Requirement already satisfied: pyparsing&gt;=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging-&gt;anndata&gt;=0.6.22.post1-&gt;kb-python) (2.4.7)'</li><li>'Requirement already satisfied: typing-extensions&gt;=3.6.4; python_version &lt; "3.8" in /usr/local/lib/python3.7/dist-packages (from importlib-metadata&gt;=0.7; python_version &lt; "3.8"-&gt;anndata&gt;=0.6.22.post1-&gt;kb-python) (3.7.4.3)'</li><li>'Requirement already satisfied: zipp&gt;=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata&gt;=0.7; python_version &lt; "3.8"-&gt;anndata&gt;=0.6.22.post1-&gt;kb-python) (3.4.1)'</li><li>'Requirement already satisfied: python-dateutil&gt;=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas!=1.1,&gt;=1.0-&gt;anndata&gt;=0.6.22.post1-&gt;kb-python) (2.8.1)'</li><li>'Requirement already satisfied: pytz&gt;=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas!=1.1,&gt;=1.0-&gt;anndata&gt;=0.6.22.post1-&gt;kb-python) (2018.9)'</li><li>'Collecting pynndescent&gt;=0.5'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/af/65/8189298dd3a05bbad716ee8e249764ff8800e365d8dc652ad2192ca01b4a/pynndescent-0.5.2.tar.gz (1.1MB)'</span></li><li>'Requirement already satisfied: cycler&gt;=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib&gt;=3.1.2-&gt;scanpy&gt;=1.4.4.post1-&gt;kb-python) (0.10.0)'</li><li>'Requirement already satisfied: kiwisolver&gt;=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib&gt;=3.1.2-&gt;scanpy&gt;=1.4.4.post1-&gt;kb-python) (1.3.1)'</li><li>'Collecting stdlib_list'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/7a/b1/52f59dcf31ead2f0ceff8976288449608d912972b911f55dff712cef5719/stdlib_list-0.8.0-py3-none-any.whl (63kB)'</span></li><li>'Collecting get-version&gt;=2.0.4'</li><li><span style=white-space:pre-wrap>'  Downloading https://files.pythonhosted.org/packages/23/48/7610e884e62fff2183e7bc8592397c39a020267fb5147905fcd3f9cc820c/get_version-2.1-py3-none-any.whl (43kB)'</span></li><li>'Requirement already satisfied: numexpr&gt;=2.5.2 in /usr/local/lib/python3.7/dist-packages (from tables-&gt;scanpy&gt;=1.4.4.post1-&gt;kb-python) (2.7.3)'</li><li>'Requirement already satisfied: decorator&gt;=4.3.0 in /usr/local/lib/python3.7/dist-packages (from networkx&gt;=2.3-&gt;scanpy&gt;=1.4.4.post1-&gt;kb-python) (4.4.2)'</li><li>'Building wheels for collected packages: loompy, numpy-groupies, umap-learn, sinfo, pynndescent'</li><li><span style=white-space:pre-wrap>'  Building wheel for loompy (setup.py): started'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for loompy (setup.py): finished with status \'done\''</span></li><li><span style=white-space:pre-wrap>'  Created wheel for loompy: filename=loompy-3.0.6-cp37-none-any.whl size=47896 sha256=9a9ab62e41414913f9bbfb2eef13596a38bc9dd39316cabab892f51143ed4d10'</span></li><li><span style=white-space:pre-wrap>'  Stored in directory: /root/.cache/pip/wheels/f9/a4/90/5a98ad83419732b0fba533b81a2a52ba3dbe230a936ca4cdc9'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for numpy-groupies (setup.py): started'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for numpy-groupies (setup.py): finished with status \'done\''</span></li><li><span style=white-space:pre-wrap>'  Created wheel for numpy-groupies: filename=numpy_groupies-0.9.13-cp37-none-any.whl size=24068 sha256=bb4d7f7f96af13b58c827bf94f338dd2559b83522f750c17546667d33088ca2e'</span></li><li><span style=white-space:pre-wrap>'  Stored in directory: /root/.cache/pip/wheels/ef/97/d7/270bc85eb8b1b84629caac97a3900bff23edb7f834f6ed729e'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for umap-learn (setup.py): started'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for umap-learn (setup.py): finished with status \'done\''</span></li><li><span style=white-space:pre-wrap>'  Created wheel for umap-learn: filename=umap_learn-0.5.1-cp37-none-any.whl size=76569 sha256=627ba759b3e6d1787390e59be46a8ee88ea2f8e3961907069df003122d8e052c'</span></li><li><span style=white-space:pre-wrap>'  Stored in directory: /root/.cache/pip/wheels/ad/df/d5/a3691296ff779f25cd1cf415a3af954b987fb53111e3392cf4'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for sinfo (setup.py): started'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for sinfo (setup.py): finished with status \'done\''</span></li><li><span style=white-space:pre-wrap>'  Created wheel for sinfo: filename=sinfo-0.3.1-cp37-none-any.whl size=7012 sha256=0a7ca4b4b981f9ff8c0ebc0f677119dfa8e32cb5746fca18b261ae01f0f18e7e'</span></li><li><span style=white-space:pre-wrap>'  Stored in directory: /root/.cache/pip/wheels/11/f0/23/347d6d8e59787c2bc272162d18223dc3b45bd6dc40aceee6af'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for pynndescent (setup.py): started'</span></li><li><span style=white-space:pre-wrap>'  Building wheel for pynndescent (setup.py): finished with status \'done\''</span></li><li><span style=white-space:pre-wrap>'  Created wheel for pynndescent: filename=pynndescent-0.5.2-cp37-none-any.whl size=51351 sha256=46d4d6df8cbd2dabd69c0f051626ff9b8978fb9141ff68278e7c7d83ec086277'</span></li><li><span style=white-space:pre-wrap>'  Stored in directory: /root/.cache/pip/wheels/ba/52/4e/4c28d04d144a28f89e2575fb63628df6e6d49b56c5ddd0c74e'</span></li><li>'Successfully built loompy numpy-groupies umap-learn sinfo pynndescent'</li><li>'Installing collected packages: numpy-groupies, loompy, anndata, pynndescent, umap-learn, stdlib-list, sinfo, get-version, legacy-api-wrap, scanpy, plotly, kb-python'</li><li><span style=white-space:pre-wrap>'  Found existing installation: plotly 4.4.1'</span></li><li><span style=white-space:pre-wrap>'    Uninstalling plotly-4.4.1:'</span></li><li><span style=white-space:pre-wrap>'      Successfully uninstalled plotly-4.4.1'</span></li><li>'Successfully installed anndata-0.7.5 get-version-2.1 kb-python-0.25.1 legacy-api-wrap-1.2 loompy-3.0.6 numpy-groupies-0.9.13 plotly-4.14.3 pynndescent-0.5.2 scanpy-1.7.1 sinfo-0.3.1 stdlib-list-0.8.0 umap-learn-0.5.1'</li></ol>

<h3 id="download-the-data">Download the data<a class="headerlink" href="#download-the-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Download the data from the 10x website</span>
<span class="nf">system</span><span class="p">(</span><span class="s">&quot;wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">intern</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">system</span><span class="p">(</span><span class="s">&quot;tar -xvf pbmc_1k_v3_fastqs.tar&quot;</span><span class="p">,</span> <span class="n">intern</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'pbmc_1k_v3_fastqs/'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_I1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_I1_001.fastq.gz'</li></ol>

<h3 id="download-an-index">Download an index<a class="headerlink" href="#download-an-index" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;kb ref -d human -i index.idx -g t2g.txt -f1 transcriptome.fasta&quot;</span><span class="p">,</span><span class="n">intern</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="pseudoalignment-and-counting">Pseudoalignment and counting<a class="headerlink" href="#pseudoalignment-and-counting" title="Permanent link">&para;</a></h2>
<h3 id="run-kallisto-and-bustools">Run kallisto and bustools<a class="headerlink" href="#run-kallisto-and-bustools" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;kb count -i index.idx -g t2g.txt -x 10xv3 -o output --filter bustools -t 2 pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz&quot;</span><span class="p">,</span><span class="n">intern</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="basic-qc">Basic QC<a class="headerlink" href="#basic-qc" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">patchwork</span><span class="p">)</span>
<span class="nf">theme_set</span><span class="p">(</span><span class="nf">theme_bw</span><span class="p">())</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Attaching SeuratObject

Warning message in system(&quot;timedatectl&quot;, intern = TRUE):
“running command &#39;timedatectl&#39; had status 1”
Registered S3 method overwritten by &#39;cli&#39;:
  method     from         
  print.boxx spatstat.geom

── [1mAttaching packages[22m ─────────────────────────────────────── tidyverse 1.3.0 ──

[32m✔[39m [34mggplot2[39m 3.3.3     [32m✔[39m [34mpurrr  [39m 0.3.4
[32m✔[39m [34mtibble [39m 3.1.0     [32m✔[39m [34mdplyr  [39m 1.0.5
[32m✔[39m [34mtidyr  [39m 1.1.3     [32m✔[39m [34mstringr[39m 1.4.0
[32m✔[39m [34mreadr  [39m 1.4.0     [32m✔[39m [34mforcats[39m 0.5.1

── [1mConflicts[22m ────────────────────────────────────────── tidyverse_conflicts() ──
[31m✖[39m [34mtidyr[39m::[32mexpand()[39m masks [34mMatrix[39m::expand()
[31m✖[39m [34mdplyr[39m::[32mfilter()[39m masks [34mstats[39m::filter()
[31m✖[39m [34mdplyr[39m::[32mlag()[39m    masks [34mstats[39m::lag()
[31m✖[39m [34mtidyr[39m::[32mpack()[39m   masks [34mMatrix[39m::pack()
[31m✖[39m [34mtidyr[39m::[32munpack()[39m masks [34mMatrix[39m::unpack()
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span> <span class="n">recursive</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'index.idx'</li><li>'output/10xv3_whitelist.txt'</li><li>'output/counts_filtered/cells_x_genes.barcodes.txt'</li><li>'output/counts_filtered/cells_x_genes.genes.txt'</li><li>'output/counts_filtered/cells_x_genes.mtx'</li><li>'output/counts_unfiltered/cells_x_genes.barcodes.txt'</li><li>'output/counts_unfiltered/cells_x_genes.genes.txt'</li><li>'output/counts_unfiltered/cells_x_genes.mtx'</li><li>'output/filter_barcodes.txt'</li><li>'output/inspect.json'</li><li>'output/kb_info.json'</li><li>'output/matrix.ec'</li><li>'output/output.bus'</li><li>'output/output.filtered.bus'</li><li>'output/output.unfiltered.bus'</li><li>'output/run_info.json'</li><li>'output/transcripts.txt'</li><li>'pbmc_1k_v3_fastqs.tar'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_I1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L001_R2_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_I1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R1_001.fastq.gz'</li><li>'pbmc_1k_v3_fastqs/pbmc_1k_v3_S1_L002_R2_001.fastq.gz'</li><li>'sample_data/anscombe.json'</li><li>'sample_data/california_housing_test.csv'</li><li>'sample_data/california_housing_train.csv'</li><li>'sample_data/mnist_test.csv'</li><li>'sample_data/mnist_train_small.csv'</li><li>'sample_data/README.md'</li><li>'t2g.txt'</li></ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Slightly modified from BUSpaRse, just to avoid installing a few dependencies not used here</span>
<span class="n">read_count_output</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">dir</span> <span class="o">&lt;-</span> <span class="nf">normalizePath</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="n">mustWork</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
  <span class="n">m</span> <span class="o">&lt;-</span> <span class="nf">readMM</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="s">&quot;/&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;.mtx&quot;</span><span class="p">))</span>
  <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">t</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
  <span class="n">m</span> <span class="o">&lt;-</span> <span class="nf">as</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="s">&quot;dgCMatrix&quot;</span><span class="p">)</span>
  <span class="c1"># The matrix read has cells in rows</span>
  <span class="n">ge</span> <span class="o">&lt;-</span> <span class="s">&quot;.genes.txt&quot;</span>
  <span class="n">genes</span> <span class="o">&lt;-</span> <span class="nf">readLines</span><span class="p">(</span><span class="nf">file</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="s">&quot;/&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ge</span><span class="p">)))</span>
  <span class="n">barcodes</span> <span class="o">&lt;-</span> <span class="nf">readLines</span><span class="p">(</span><span class="nf">file</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="s">&quot;/&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;.barcodes.txt&quot;</span><span class="p">)))</span>
  <span class="nf">colnames</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">barcodes</span>
  <span class="nf">rownames</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">genes</span>
  <span class="nf">return</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="nf">read_count_output</span><span class="p">(</span><span class="s">&quot;./output/counts_unfiltered&quot;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;cells_x_genes&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>60623</li><li>259615</li></ol>

<h3 id="test-for-library-saturation">Test for library saturation<a class="headerlink" href="#test-for-library-saturation" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tot_counts</span> <span class="o">&lt;-</span> <span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
<span class="n">lib_sat</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">nCount</span> <span class="o">=</span> <span class="n">tot_counts</span><span class="p">,</span>
                  <span class="n">nGene</span> <span class="o">=</span> <span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">lib_sat</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">nCount</span><span class="p">,</span> <span class="n">nGene</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">alpha</span> <span class="o">=</span> <span class="m">0.1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span> <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span> <span class="nf">annotation_logticks</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Warning message:
“Transformation introduced infinite values in continuous x-axis”
Warning message:
“Transformation introduced infinite values in continuous y-axis”
</code></pre></div>

<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_24_1.png" /></p>
<p>This plot is very misleading, as even the small alpha can't accurately show how many points are stacked at one location.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">lib_sat</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">nCount</span><span class="p">,</span> <span class="n">nGene</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_bin2d</span><span class="p">(</span><span class="n">bins</span> <span class="o">=</span> <span class="m">50</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_fill_distiller</span><span class="p">(</span><span class="n">palette</span> <span class="o">=</span> <span class="s">&quot;Blues&quot;</span><span class="p">,</span> <span class="n">direction</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span> <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span> <span class="nf">annotation_logticks</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Warning message:
“Transformation introduced infinite values in continuous x-axis”
Warning message:
“Transformation introduced infinite values in continuous y-axis”
Warning message:
“Removed 19583 rows containing non-finite values (stat_bin2d).”
</code></pre></div>

<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_26_1.png" /></p>
<p>Lots of points are piled at around 1 gene and 1 count, and those with 0 gene or count were removed for introducing -Inf in log transform. These correspond to empty or near empty droplets.</p>
<h3 id="examine-the-knee-plot">Examine the knee plot<a class="headerlink" href="#examine-the-knee-plot" title="Permanent link">&para;</a></h3>
<p>The "knee plot" was introduced in the Drop-seq paper: 
- Macosko et al., <a href="https://www.cell.com/fulltext/S0092-8674(15)00549-8">Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets</a>, 2015. DOI:10.1016/j.cell.2015.05.002</p>
<p>In this plot cells are ordered by the number of UMI counts associated to them (shown on the <em>x</em>-axis), and the fraction of droplets with at least that number of cells is shown on the <em>y</em>-axis:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">summary</span><span class="p">(</span><span class="n">tot_counts</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    0.00     1.00     1.00    43.64     6.00 60120.00
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#&#39; @rdname knee_plot</span>
<span class="c1">#&#39; @param mat Gene count matrix, a dgCMatrix.</span>
<span class="c1">#&#39; @return `get_knee_df` returns a tibble with two columns: \code{total} for </span>
<span class="c1">#&#39; total UMI counts for each barcode, and \code{rank} for rank of the total </span>
<span class="c1">#&#39; counts, with number 1 for the barcode with the most counts.</span>
<span class="c1">#&#39; @export</span>
<span class="c1">#&#39; @importFrom dplyr row_number desc arrange</span>
<span class="c1">#&#39; @importFrom Matrix colSums</span>
<span class="n">get_knee_df</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">total</span> <span class="o">&lt;-</span> <span class="n">rank</span> <span class="o">&lt;-</span> <span class="kc">NULL</span>
  <span class="nf">tibble</span><span class="p">(</span><span class="n">total</span> <span class="o">=</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">mat</span><span class="p">),</span>
         <span class="n">rank</span> <span class="o">=</span> <span class="nf">row_number</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">total</span><span class="p">)))</span> <span class="o">%&gt;%</span>
    <span class="nf">distinct</span><span class="p">()</span> <span class="o">%&gt;%</span>
    <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">total</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span> <span class="o">%&gt;%</span> 
    <span class="nf">arrange</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1">#&#39; @rdname knee_plot</span>
<span class="c1">#&#39; @param df The data frame from \code{\link{get_knee_df}}.</span>
<span class="c1">#&#39; @param lower Minimum total UMI counts for barcode for it to be considered</span>
<span class="c1">#&#39; when calculating the inflection point; this helps to avoid the noisy part of</span>
<span class="c1">#&#39; the curve for barcodes with very few counts.</span>
<span class="c1">#&#39; @return `get_inflection` returns a \code{numeric(1)} for the total UMI count </span>
<span class="c1">#&#39; at the inflection point.</span>
<span class="c1">#&#39; @note Code in part adapted from \code{barcodeRanks} from \code{DropetUtils}.</span>
<span class="c1">#&#39; @export</span>
<span class="c1">#&#39; @importFrom dplyr transmute</span>
<span class="c1">#&#39; </span>
<span class="n">get_inflection</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">lower</span> <span class="o">=</span> <span class="m">100</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">log_total</span> <span class="o">&lt;-</span> <span class="n">log_rank</span> <span class="o">&lt;-</span> <span class="n">total</span> <span class="o">&lt;-</span>  <span class="kc">NULL</span>
  <span class="n">df_fit</span> <span class="o">&lt;-</span> <span class="n">df</span> <span class="o">%&gt;%</span> 
    <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">total</span> <span class="o">&gt;</span> <span class="n">lower</span><span class="p">)</span> <span class="o">%&gt;%</span> 
    <span class="nf">transmute</span><span class="p">(</span><span class="n">log_total</span> <span class="o">=</span> <span class="nf">log10</span><span class="p">(</span><span class="n">total</span><span class="p">),</span>
              <span class="n">log_rank</span> <span class="o">=</span> <span class="nf">log10</span><span class="p">(</span><span class="n">rank</span><span class="p">))</span>
  <span class="n">d1n</span> <span class="o">&lt;-</span> <span class="nf">diff</span><span class="p">(</span><span class="n">df_fit</span><span class="o">$</span><span class="n">log_total</span><span class="p">)</span><span class="o">/</span><span class="nf">diff</span><span class="p">(</span><span class="n">df_fit</span><span class="o">$</span><span class="n">log_rank</span><span class="p">)</span>
  <span class="n">right.edge</span> <span class="o">&lt;-</span> <span class="nf">which.min</span><span class="p">(</span><span class="n">d1n</span><span class="p">)</span>
  <span class="m">10</span><span class="o">^</span><span class="p">(</span><span class="n">df_fit</span><span class="o">$</span><span class="n">log_total</span><span class="p">[</span><span class="n">right.edge</span><span class="p">])</span>
<span class="p">}</span>

<span class="c1">#&#39; Plot the transposed knee plot and inflection point</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; Plot a transposed knee plot, showing the inflection point and</span>
<span class="c1">#&#39; the number of remaining cells after inflection point filtering. It&#39;s</span>
<span class="c1">#&#39; transposed since it&#39;s more generalizable to multi-modal data. Taken from the </span>
<span class="c1">#&#39; BUSpaRse package.</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; @param df The data frame from \code{\link{get_knee_df}}.</span>
<span class="c1">#&#39; @param inflection Output of \code{\link{get_inflection}}.</span>
<span class="c1">#&#39; @return `knee_plot` returns a \code{ggplot2} object.</span>
<span class="c1">#&#39; @export</span>
<span class="c1">#&#39; @importFrom ggplot2 ggplot aes geom_path geom_vline geom_hline </span>
<span class="c1">#&#39; scale_x_log10 scale_y_log10 labs annotation_logticks geom_text</span>
<span class="n">knee_plot</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">inflection</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">total</span> <span class="o">&lt;-</span> <span class="n">rank_cutoff</span> <span class="o">&lt;-</span> <span class="kc">NULL</span>
  <span class="n">annot</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">inflection</span> <span class="o">=</span> <span class="n">inflection</span><span class="p">,</span>
                  <span class="n">rank_cutoff</span> <span class="o">=</span> <span class="nf">max</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">rank</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">total</span> <span class="o">&gt;</span> <span class="n">inflection</span><span class="p">]))</span>
  <span class="nf">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">total</span><span class="p">,</span> <span class="n">rank</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_path</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">geom_vline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="n">inflection</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> 
               <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;gray40&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_hline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="n">rank_cutoff</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> 
               <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;gray40&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">inflection</span><span class="p">,</span> <span class="n">rank_cutoff</span><span class="p">,</span> 
                  <span class="n">label</span> <span class="o">=</span> <span class="nf">paste</span><span class="p">(</span><span class="n">rank_cutoff</span><span class="p">,</span> <span class="s">&quot;&#39;cells&#39;&quot;</span><span class="p">)),</span>
              <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">vjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">labs</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Rank&quot;</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Total UMIs&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">annotation_logticks</span><span class="p">()</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="n">knee_df</span> <span class="o">&lt;-</span> <span class="nf">get_knee_df</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
<span class="n">inflection</span> <span class="o">&lt;-</span> <span class="nf">get_inflection</span><span class="p">(</span><span class="n">knee_df</span><span class="p">)</span>
<span class="nf">knee_plot</span><span class="p">(</span><span class="n">knee_df</span><span class="p">,</span> <span class="n">inflection</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_31_0.png" /></p>
<h2 id="analysis">Analysis<a class="headerlink" href="#analysis" title="Permanent link">&para;</a></h2>
<p>We begin by asking for genes with the highest proportions in droplets (prior to filtering out empty droplets).</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tr2g</span> <span class="o">&lt;-</span> <span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;t2g.txt&quot;</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;transcript&quot;</span><span class="p">,</span> <span class="s">&quot;gene&quot;</span><span class="p">,</span> <span class="s">&quot;gene_name&quot;</span><span class="p">))</span>
<span class="n">tr2g</span> <span class="o">&lt;-</span> <span class="nf">distinct</span><span class="p">(</span><span class="n">tr2g</span><span class="p">[,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;gene&quot;</span><span class="p">,</span> <span class="s">&quot;gene_name&quot;</span><span class="p">)])</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[36m──[39m [1m[1mColumn specification[1m[22m [36m────────────────────────────────────────────────────────[39m
cols(
  transcript = [31mcol_character()[39m,
  gene = [31mcol_character()[39m,
  gene_name = [31mcol_character()[39m
)
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">plot_pct_genes</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">tr2g</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">symbol</span> <span class="o">=</span> <span class="s">&quot;ensembl&quot;</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">pct_tx</span> <span class="o">&lt;-</span> <span class="nf">rowSums</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
  <span class="n">gs</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">mat</span><span class="p">)[</span><span class="nf">order</span><span class="p">(</span><span class="o">-</span><span class="n">pct_tx</span><span class="p">)]</span>
  <span class="n">df</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">mat</span><span class="p">[</span><span class="n">gs</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">],]))</span>
  <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">df</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate_all</span><span class="p">(</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">x</span><span class="o">/</span><span class="nf">colSums</span><span class="p">(</span><span class="n">mat</span><span class="p">))</span> <span class="o">%&gt;%</span>
    <span class="nf">pivot_longer</span><span class="p">(</span><span class="nf">everything</span><span class="p">(),</span> <span class="n">names_to</span> <span class="o">=</span> <span class="s">&quot;gene&quot;</span><span class="p">)</span>
  <span class="nf">if </span><span class="p">(</span><span class="n">symbol</span> <span class="o">==</span> <span class="s">&quot;ensembl&quot;</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">df</span> <span class="o">&lt;-</span> <span class="nf">left_join</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">tr2g</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s">&quot;gene&quot;</span><span class="p">)</span>
  <span class="p">}</span> <span class="n">else</span> <span class="p">{</span>
    <span class="n">df</span> <span class="o">&lt;-</span> <span class="nf">rename</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene</span><span class="p">)</span>
  <span class="p">}</span>
    <span class="n">df</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="nf">fct_reorder</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">.fun</span> <span class="o">=</span> <span class="n">median</span><span class="p">))</span> <span class="o">%&gt;%</span>
    <span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">value</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_boxplot</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Proportion of total counts&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">coord_flip</span><span class="p">()</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">6</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">10</span><span class="p">)</span>
<span class="nf">plot_pct_genes</span><span class="p">(</span><span class="n">res_mat</span><span class="p">,</span> <span class="n">tr2g</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Warning message:
“Removed 391660 rows containing non-finite values (stat_boxplot).”
</code></pre></div>

<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_35_1.png" /></p>
<p>For many barcodes, the top genes by proportion of all counts are ribosomal or mitochondrial genes. Also, the proportions plotted above seem to have some discrete values; this effect is a result of computing fractions with small denominator, which happens when droplets produce very few UMI counts.</p>
<h3 id="filter">Filter<a class="headerlink" href="#filter" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[,</span> <span class="n">tot_counts</span> <span class="o">&gt;</span> <span class="n">inflection</span><span class="p">]</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">,]</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>31861</li><li>1322</li></ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Convert from Ensembl gene ID to gene symbol</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">tr2g</span><span class="o">$</span><span class="n">gene_name</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">),</span> <span class="n">tr2g</span><span class="o">$</span><span class="n">gene</span><span class="p">)]</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="p">(</span><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">res_mat</span><span class="p">,</span> <span class="n">project</span> <span class="o">=</span> <span class="s">&quot;pbmc1k&quot;</span><span class="p">,</span> <span class="n">min.cells</span> <span class="o">=</span> <span class="m">3</span><span class="p">,</span> <span class="n">min.features</span> <span class="o">=</span> <span class="m">200</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Warning</span> <span class="n">message</span><span class="p">:</span>
<span class="err">“</span><span class="n">Non</span><span class="o">-</span><span class="n">unique</span> <span class="n">features</span> <span class="p">(</span><span class="n">rownames</span><span class="p">)</span> <span class="n">present</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">input</span> <span class="n">matrix</span><span class="p">,</span> <span class="n">making</span> <span class="n">unique</span><span class="err">”</span>
<span class="n">Warning</span> <span class="n">message</span><span class="p">:</span>
<span class="err">“</span><span class="n">Feature</span> <span class="n">names</span> <span class="n">cannot</span> <span class="n">have</span> <span class="n">underscores</span> <span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">),</span> <span class="n">replacing</span> <span class="n">with</span> <span class="n">dashes</span> <span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span><span class="err">”</span>



<span class="n">An</span> <span class="n">object</span> <span class="n">of</span> <span class="k">class</span> <span class="n">Seurat</span> 
<span class="mi">25966</span> <span class="n">features</span> <span class="n">across</span> <span class="mi">1208</span> <span class="n">samples</span> <span class="n">within</span> <span class="mi">1</span> <span class="n">assay</span> 
<span class="n">Active</span> <span class="n">assay</span><span class="p">:</span> <span class="n">RNA</span> <span class="p">(</span><span class="mi">25966</span> <span class="n">features</span><span class="p">,</span> <span class="mi">0</span> <span class="n">variable</span> <span class="n">features</span><span class="p">)</span>
</code></pre></div>

<p>The steps below constitute a standard analysis worklow for single-cell RNA-seq data. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span>
<span class="n">pbmc</span><span class="p">[[</span><span class="s">&quot;percent.mt&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">PercentageFeatureSet</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s">&quot;^MT-&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>The number of unique genes and total molecules are automatically calculated when running the <code>CreateSeuratObject</code> command.
The associated data is stored in the object metadata.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Show QC metrics for the first 5 cells</span>
<span class="nf">head</span><span class="p">(</span><span class="n">pbmc</span><span class="o">@</span><span class="n">meta.data</span><span class="p">,</span> <span class="m">5</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="dataframe">
<caption>A data.frame: 5 × 4</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mt</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>AAACCCAAGGAGAGTA</th><td>pbmc1k</td><td>9289</td><td>3198</td><td>11.271396</td></tr>
    <tr><th scope=row>AAACGCTTCAGCCCAG</th><td>pbmc1k</td><td>6483</td><td>2513</td><td> 8.252352</td></tr>
    <tr><th scope=row>AAAGAACAGACGACTG</th><td>pbmc1k</td><td>5011</td><td>2082</td><td> 6.166434</td></tr>
    <tr><th scope=row>AAAGAACCAATGGCAG</th><td>pbmc1k</td><td>3264</td><td>1555</td><td> 6.893382</td></tr>
    <tr><th scope=row>AAAGAACGTCTGCAAT</th><td>pbmc1k</td><td>7488</td><td>2508</td><td> 6.610577</td></tr>
</tbody>
</table>

<p>Next, we visualize some QC metrics and use the results to set filtering criteria.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Visualize QC metrics as a violin plot</span>
<span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">12</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">VlnPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;nFeature_RNA&quot;</span><span class="p">,</span> <span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span> <span class="s">&quot;percent.mt&quot;</span><span class="p">),</span> <span class="n">ncol</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_46_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used</span>
<span class="c1"># for anything calculated by the object, i.e. columns in object metadata, PC scores etc.</span>
<span class="n">plot1</span> <span class="o">&lt;-</span> <span class="nf">FeatureScatter</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">feature1</span> <span class="o">=</span> <span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span> <span class="n">feature2</span> <span class="o">=</span> <span class="s">&quot;percent.mt&quot;</span><span class="p">)</span>
<span class="n">plot2</span> <span class="o">&lt;-</span> <span class="nf">FeatureScatter</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">feature1</span> <span class="o">=</span> <span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span> <span class="n">feature2</span> <span class="o">=</span> <span class="s">&quot;nFeature_RNA&quot;</span><span class="p">)</span>
<span class="n">plot1</span> <span class="o">+</span> <span class="n">plot2</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_47_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">subset</span> <span class="o">=</span> <span class="n">nFeature_RNA</span> <span class="o">&gt;</span> <span class="m">200</span> <span class="o">&amp;</span> <span class="n">nFeature_RNA</span> <span class="o">&lt;</span> <span class="m">6500</span> <span class="o">&amp;</span> <span class="n">percent.mt</span> <span class="o">&lt;</span> <span class="m">20</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h3 id="normalize">Normalize<a class="headerlink" href="#normalize" title="Permanent link">&para;</a></h3>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data. A standard choice is <code>LogNormalize</code> which normalizes the UMI counts for each cell by the total counts, multiplies this by a scale factor (10,000 by default), and finally log-transforms the result. </p>
<p>We recommend the preprint 
- Breda, J., Zavolan, M. and van Nimwegen, E. Bayesian inference of the gene expression states of single cells from scRNA-seq data. bioRxiv (2019). doi.org/10.1101/2019.12.28.889956 </p>
<p>for a thorough discussion of normalization.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">NormalizeData</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">normalization.method</span> <span class="o">=</span> <span class="s">&quot;LogNormalize&quot;</span><span class="p">,</span> <span class="n">scale.factor</span> <span class="o">=</span> <span class="m">10000</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>For clarity, in this previous line of code (and in future commands), we provide the default values for certain parameters in the function call. However, this isn’t required and the same behavior can be achieved with:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># pbmc &lt;- NormalizeData(pbmc)</span>
</code></pre></div>
</td></tr></table>
<h3 id="highly-expressed-genes">Highly expressed genes<a class="headerlink" href="#highly-expressed-genes" title="Permanent link">&para;</a></h3>
<p>To identify a subset of genes that exhibit high cell-to-cell variation in the dataset we apply a procedure implemented in the <code>FindVariableFeatures</code> function. By default, it returns 2,000 genes per dataset. These will be used in downstream analysis.</p>
<p>Seurat documentation describes the method used to find highly variable genes here as such:</p>
<blockquote>
<p>First, fits a line to the relationship of log(variance) and log(mean) using local polynomial regression (loess). Then standardizes the feature values using the observed mean and expected variance (given by the fitted line). Feature variance is then calculated on the standardized values after clipping to a maximum (see clip.max parameter).</p>
</blockquote>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">6</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">10</span><span class="p">)</span>
<span class="nf">plot_pct_genes</span><span class="p">(</span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span> <span class="s">&quot;counts&quot;</span><span class="p">),</span> <span class="n">tr2g</span><span class="p">,</span> <span class="n">symbol</span> <span class="o">=</span> <span class="s">&quot;symbol&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_56_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span>
<span class="normal">9</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">selection.method</span> <span class="o">=</span> <span class="s">&quot;vst&quot;</span><span class="p">,</span> <span class="n">nfeatures</span> <span class="o">=</span> <span class="m">2000</span><span class="p">)</span>

<span class="c1"># Identify the 10 most highly variable genes</span>
<span class="n">top10</span> <span class="o">&lt;-</span> <span class="nf">head</span><span class="p">(</span><span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">pbmc</span><span class="p">),</span> <span class="m">10</span><span class="p">)</span>

<span class="c1"># plot variable features with and without labels</span>
<span class="n">plot1</span> <span class="o">&lt;-</span> <span class="nf">VariableFeaturePlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">log</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">LabelPoints</span><span class="p">(</span><span class="n">plot</span> <span class="o">=</span> <span class="n">plot1</span><span class="p">,</span> <span class="n">points</span> <span class="o">=</span> <span class="n">top10</span><span class="p">,</span> <span class="n">repel</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">When</span> <span class="nv">using</span> <span class="nv">repel</span>, <span class="nv">set</span> <span class="nv">xnudge</span> <span class="nv">and</span> <span class="nv">ynudge</span> <span class="nv">to</span> <span class="mi">0</span> <span class="k">for</span> <span class="nv">optimal</span> <span class="nv">results</span>
</code></pre></div>

<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_57_1.png" /></p>
<h3 id="scaling-the-data">Scaling the data<a class="headerlink" href="#scaling-the-data" title="Permanent link">&para;</a></h3>
<p>Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData function shifts the expression of each gene, so that the mean expression across cells is 0 and the variance across cells is 1
This step gives equal weight to genes in downstream analyses, so that highly-expressed genes do not dominate.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">all.genes</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>
<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">ScaleData</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="n">all.genes</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Centering and scaling data matrix
</code></pre></div>

<p>We apply this only to the genes identified as highly variable:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># pbmc &lt;- ScaleData(pbmc)</span>
</code></pre></div>
</td></tr></table>
<p>The scaling does not affect PCA or clustering results. However, Seurat heatmaps (produced as shown below with DoHeatmap) require genes in the heatmap to be scaled so that highly-expressed genes don’t dominate. To make sure we don’t leave any genes out of the heatmap later, we are scaling all genes in this tutorial.</p>
<p>In Seurat v2 we also use the ScaleData function to remove unwanted sources of variation from a single-cell dataset. For example, we could ‘regress out’ heterogeneity associated with (for example) cell cycle stage, or mitochondrial contamination. These features are still supported in ScaleData in Seurat v3, i.e.:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># pbmc &lt;- ScaleData(pbmc, vars.to.regress = &quot;percent.mt&quot;)</span>
</code></pre></div>
</td></tr></table>
<h3 id="principal-component-analysis">Principal component analysis<a class="headerlink" href="#principal-component-analysis" title="Permanent link">&para;</a></h3>
<p>Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">RunPCA</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">pbmc</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>PC_ 1 
Positive:  S100A9, FCN1, MNDA, FGL2, S100A8, CTSS, CST3, SERPINA1, PSAP, NCF2 
       LYZ, AIF1, TYMP, VCAN, KLF4, GRN, CSTA, MPEG1, CPVL, CLEC7A 
       LST1, MS4A6A, CD14, LGALS1, S100A12, TYROBP, TNFAIP2, FCER1G, CD36, CSF3R 
Negative:  LTB, TRAC, TRBC2, CD3D, IL32, BCL11B, CD3G, IL7R, TCF7, CD69 
       ISG20, CD247, CD27, SPOCK2, ARL4C, CD7, CD2, GZMM, TRBC1, CD6 
       PRKCQ-AS1, NOSIP, AC058791.1, RORA, CTSW, CCR7, AQP3, ITM2A, PEBP1, SAMD3 
PC_ 2 
Positive:  CD79A, MS4A1, IGHM, BANK1, BCL11A, LINC00926, CD79B, TNFRSF13C, IGHD, CD74 
       HLA-DQB1, CD22, HLA-DQA1, HLA-DRB1, HLA-DRA, HLA-DPA1, HLA-DPB1, TCL1A, FCER2, AFF3 
       PAX5, IGKC, VPREB3, SPIB, MEF2C, RALGPS2, HVCN1, FCRL1, HLA-DOB, HLA-DMA 
Negative:  IL32, CD247, GZMM, CD7, CTSW, CD3D, GZMA, NKG7, S100A4, TRAC 
       ANXA1, BCL11B, PRF1, CST7, KLRB1, CD3G, IL7R, ARL4C, SAMD3, CD2 
       TRBC1, CCL5, KLRG1, A2M, MT2A, RORA, ITGB2, GNLY, TCF7, MATK 
PC_ 3 
Positive:  CAVIN2, GP9, PF4, GNG11, PPBP, CD9, TREML1, CMTM5, TUBB1, SPARC 
       CLU, HIST1H2AC, ACRBP, PTCRA, PRKAR2B, NRGN, ITGA2B, CTTN, TMEM40, TSC22D1 
       AC147651.1, GMPR, PF4V1, CLDN5, CA2, MAP3K7CL, PGRMC1, CXCR2P1, HIST1H3H, MMD 
Negative:  CYBA, VIM, FOS, ITGB2, NEAT1, HNRNPU, CALR, LSP1, LCP1, DUSP1 
       S100A10, S100A6, KLF6, CD74, PLAC8, LTB, ZFP36L1, S100A4, IFITM2, ISG20 
       SPCS1, SEC61B, ANXA1, MCL1, EVI2B, HSPA5, APOBEC3G, HSP90B1, PEBP1, AC020916.1 
PC_ 4 
Positive:  LEF1, TCF7, IL7R, MAL, CCR7, BCL11B, CD3D, NOSIP, LTB, CD3G 
       TRAC, CAMK4, NELL2, PASK, CD27, EGR1, SLC2A3, RGCC, FHIT, RGS10 
       CD6, CD40LG, VIM, INPP4B, ADTRP, TRAT1, NOG, TSHZ2, PRKCQ-AS1, TESPA1 
Negative:  GZMB, GNLY, CLIC3, NKG7, KLRF1, PRF1, CST7, SPON2, FGFBP2, KLRD1 
       GZMA, ADGRG1, CCL4, TRDC, HOPX, MATK, IL2RB, TTC38, APOBEC3G, CTSW 
       TBX21, RHOC, C12orf75, S1PR5, FCGR3A, SH2D1B, PTGDR, MYOM2, CMC1, GZMH 
PC_ 5 
Positive:  LILRA4, SCT, PACSIN1, SMPD3, LRRC26, SERPINF1, TPM2, AL096865.1, IL3RA, DNASE1L3 
       TNFRSF21, CUX2, PLD4, ITM2C, GAS6, MYBL2, CLEC4C, PPP1R14B, EPHA2, UGCG 
       PPP1R14B-AS1, CUEDC1, LAMP5, RUNX2, PPM1J, SERPINF2, NRP1, DERL3, LINC02812, CIB2 
Negative:  GNLY, FGFBP2, KLRF1, PRF1, NKG7, CCL4, CST7, KLRD1, MS4A1, IGHD 
       LINC00926, ADGRG1, CD79B, CD79A, TRDC, CD22, GZMA, TBX21, SPON2, TNFRSF13C 
       MATK, FCER2, IL2RB, PAX5, MYOM2, HOPX, S1PR5, TTC38, SH2D1B, BANK1
</code></pre></div>

<p>Seurat provides several useful ways of visualizing both cells and features that define the PCA, including VizDimReduction, DimPlot, and DimHeatmap</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Examine and visualize PCA results a few different ways</span>
<span class="nf">print</span><span class="p">(</span><span class="n">pbmc</span><span class="p">[[</span><span class="s">&quot;pca&quot;</span><span class="p">]],</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">nfeatures</span> <span class="o">=</span> <span class="m">5</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>PC_ 1 
Positive:  S100A9, FCN1, MNDA, FGL2, S100A8 
Negative:  LTB, TRAC, TRBC2, CD3D, IL32 
PC_ 2 
Positive:  CD79A, MS4A1, IGHM, BANK1, BCL11A 
Negative:  IL32, CD247, GZMM, CD7, CTSW 
PC_ 3 
Positive:  CAVIN2, GP9, PF4, GNG11, PPBP 
Negative:  CYBA, VIM, FOS, ITGB2, NEAT1 
PC_ 4 
Positive:  LEF1, TCF7, IL7R, MAL, CCR7 
Negative:  GZMB, GNLY, CLIC3, NKG7, KLRF1 
PC_ 5 
Positive:  LILRA4, SCT, PACSIN1, SMPD3, LRRC26 
Negative:  GNLY, FGFBP2, KLRF1, PRF1, NKG7
</code></pre></div>

<p>Which genes are contributing the most to the first 2 PCs?</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">6</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">8</span><span class="p">)</span>
<span class="nf">VizDimLoadings</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&quot;pca&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_70_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">7</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&quot;pca&quot;</span><span class="p">,</span> <span class="n">feature</span> <span class="o">=</span> <span class="s">&quot;CST3&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_71_0.png" /></p>
<h3 id="determining-dimensionality">Determining dimensionality<a class="headerlink" href="#determining-dimensionality" title="Permanent link">&para;</a></h3>
<p>To overcome the extensive technical noise in any single feature for scRNA-seq data, one can cluster cells based on their PCA projections, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set.</p>
<p>A common heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (ElbowPlot function). In this example, we can observe an ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">ElbowPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_74_0.png" /></p>
<h3 id="the-neighborhood-graph">The neighborhood graph<a class="headerlink" href="#the-neighborhood-graph" title="Permanent link">&para;</a></h3>
<p>We cluster cells using the Louvain algorithm (a default in Seurat), which iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents function.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="n">k.param</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">FindClusters</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="m">0.6</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">Computing</span> <span class="nv">nearest</span> <span class="nv">neighbor</span> <span class="nv">graph</span>

<span class="nv">Computing</span> <span class="nv">SNN</span>



<span class="nv">Modularity</span> <span class="nv">Optimizer</span> <span class="nv">version</span> <span class="mi">1</span>.<span class="mi">3</span>.<span class="mi">0</span> <span class="nv">by</span> <span class="nv">Ludo</span> <span class="nv">Waltman</span> <span class="nv">and</span> <span class="nv">Nees</span> <span class="nv">Jan</span> <span class="nv">van</span> <span class="nv">Eck</span>

<span class="nv">Number</span> <span class="nv">of</span> <span class="nv">nodes</span>: <span class="mi">1138</span>
<span class="nv">Number</span> <span class="nv">of</span> <span class="nv">edges</span>: <span class="mi">35679</span>

<span class="nv">Running</span> <span class="nv">Louvain</span> <span class="nv">algorithm</span>...
<span class="nv">Maximum</span> <span class="nv">modularity</span> <span class="nv">in</span> <span class="mi">10</span> <span class="k">random</span> <span class="nv">starts</span>: <span class="mi">0</span>.<span class="mi">8720</span>
<span class="nv">Number</span> <span class="nv">of</span> <span class="nv">communities</span>: <span class="mi">9</span>
<span class="nv">Elapsed</span> <span class="nv">time</span>: <span class="mi">0</span> <span class="nv">seconds</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Look at cluster IDs of the first 5 cells</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">Idents</span><span class="p">(</span><span class="n">pbmc</span><span class="p">),</span> <span class="m">5</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.dl-inline {width: auto; margin:0; padding: 0}
.dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}
.dl-inline>dt::after {content: ":\0020"; padding-right: .5ex}
.dl-inline>dt:not(:first-of-type) {padding-left: .5ex}
</style>
<dl class=dl-inline><dt>AAACCCAAGGAGAGTA</dt><dd>0</dd><dt>AAACGCTTCAGCCCAG</dt><dd>3</dd><dt>AAAGAACAGACGACTG</dt><dd>5</dd><dt>AAAGAACCAATGGCAG</dt><dd>5</dd><dt>AAAGAACGTCTGCAAT</dt><dd>1</dd></dl>

<details>
    <summary style=display:list-item;cursor:pointer>
        <strong>Levels</strong>:
    </summary>
    <style>
    .list-inline {list-style: none; margin:0; padding: 0}
    .list-inline>li {display: inline-block}
    .list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
    </style>
    <ol class=list-inline><li>'0'</li><li>'1'</li><li>'2'</li><li>'3'</li><li>'4'</li><li>'5'</li><li>'6'</li><li>'7'</li><li>'8'</li></ol>
</details>

<h3 id="umap-and-t-sne">UMAP and t-SNE<a class="headerlink" href="#umap-and-t-sne" title="Permanent link">&para;</a></h3>
<p>tSNE and UMAP can be used to visualize and explore non-linear aspects of high-dimensional data. Here we apply these methods to the PC projection of the data (with same dimension as used for clustering).</p>
<p><a href="https://en.wikipedia.org/wiki/Nonlinear_dimensionality_reduction">UMAP</a> (UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction) is a manifold learning technique that can also be used to visualize cells. It was published in:</p>
<ul>
<li>McInnes, Leland, John Healy, and James Melville. "Umap: Uniform manifold approximation and projection for dimension reduction." arXiv preprint arXiv:1802.03426 (2018).</li>
</ul>
<p><a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding">t-SNE</a> is a non-linear dimensionality reduction technique described in:</p>
<ul>
<li>Maaten, Laurens van der, and Geoffrey Hinton. "Visualizing data using t-SNE." Journal of machine learning research 9.Nov (2008): 2579-2605.</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">RunUMAP</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">Warning</span> <span class="nv">message</span>:
“<span class="nv">The</span> <span class="nv">default</span> <span class="nv">method</span> <span class="k">for</span> <span class="nv">RunUMAP</span> <span class="nv">has</span> <span class="nv">changed</span> <span class="nv">from</span> <span class="nv">calling</span> <span class="nv">Python</span> <span class="nv">UMAP</span> <span class="nv">via</span> <span class="nv">reticulate</span> <span class="nv">to</span> <span class="nv">the</span> <span class="nv">R</span><span class="o">-</span><span class="nv">native</span> <span class="nv">UWOT</span> <span class="nv">using</span> <span class="nv">the</span> <span class="nv">cosine</span> <span class="nv">metric</span>
<span class="nv">To</span> <span class="nv">use</span> <span class="nv">Python</span> <span class="nv">UMAP</span> <span class="nv">via</span> <span class="nv">reticulate</span>, <span class="nv">set</span> <span class="nv">umap</span>.<span class="nv">method</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">umap-learn</span><span class="s1">&#39;</span> <span class="nv">and</span> <span class="nv">metric</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">correlation</span><span class="s1">&#39;</span>
<span class="nv">This</span> <span class="nv">message</span> <span class="nv">will</span> <span class="nv">be</span> <span class="nv">shown</span> <span class="nv">once</span> <span class="nv">per</span> <span class="nv">session</span>”
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># note that you can set `label = TRUE` or use the LabelClusters function to help label</span>
<span class="c1"># individual clusters</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&quot;umap&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_81_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">16</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">5</span><span class="p">)</span>
<span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&quot;umap&quot;</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;CST3&quot;</span><span class="p">,</span> <span class="s">&quot;NKG7&quot;</span><span class="p">,</span> <span class="s">&quot;PPBP&quot;</span><span class="p">),</span>
<span class="n">ncol</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_82_0.png" /></p>
<h3 id="finding-differentially-expressed-features-cluster-biomarkers">Finding differentially expressed features (cluster biomarkers)<a class="headerlink" href="#finding-differentially-expressed-features-cluster-biomarkers" title="Permanent link">&para;</a></h3>
<p>A key follow-up step to clustering cells is to find gene markers that are associated with them. We used Seurat's FindAllMarkers function which automates the process for all clusters.</p>
<p>The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Scanpy style gene rank plot</span>
<span class="n">plot_gene_rank</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">markers</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">df_plot</span> <span class="o">&lt;-</span> <span class="n">markers</span> <span class="o">%&gt;%</span>
    <span class="nf">group_by</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">top_n</span><span class="p">(</span><span class="m">25</span><span class="p">,</span> <span class="n">avg_log2FC</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">rank</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="nf">row_number</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">avg_log2FC</span><span class="p">))))</span>
  <span class="nf">ggplot</span><span class="p">(</span><span class="n">df_plot</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">rank</span><span class="p">,</span> <span class="n">avg_log2FC</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">label</span> <span class="o">=</span> <span class="n">gene</span><span class="p">),</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">-90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span> <span class="n">cluster</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">expand</span> <span class="o">=</span> <span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">0.05</span><span class="p">,</span> <span class="m">0.25</span><span class="p">)))</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<p>Several methods for differential expression are supported by Seurat. The default is Wilcoxon rank sum test.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># find markers for every cluster compared to all remaining cells, report only the positive ones</span>
<span class="n">pbmc.markers</span> <span class="o">&lt;-</span> <span class="nf">FindAllMarkers</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">test.use</span> <span class="o">=</span> <span class="s">&quot;wilcox&quot;</span><span class="p">,</span> <span class="n">only.pos</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> 
                               <span class="n">min.pct</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">,</span> <span class="n">logfc.threshold</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">0</span>

<span class="k">For</span> <span class="nv">a</span> <span class="nv">more</span> <span class="nv">efficient</span> <span class="nv">implementation</span> <span class="nv">of</span> <span class="nv">the</span> <span class="nv">Wilcoxon</span> <span class="nv">Rank</span> <span class="nv">Sum</span> <span class="nv">Test</span>,
<span class="ss">(</span><span class="nv">default</span> <span class="nv">method</span> <span class="k">for</span> <span class="nv">FindMarkers</span><span class="ss">)</span> <span class="nv">please</span> <span class="nv">install</span> <span class="nv">the</span> <span class="nv">limma</span> <span class="nv">package</span>
<span class="o">--------------------------------------------</span>
<span class="nv">install</span>.<span class="nv">packages</span><span class="ss">(</span><span class="s1">&#39;</span><span class="s">BiocManager</span><span class="s1">&#39;</span><span class="ss">)</span>
<span class="nv">BiocManager</span>::<span class="nv">install</span><span class="ss">(</span><span class="s1">&#39;</span><span class="s">limma</span><span class="s1">&#39;</span><span class="ss">)</span>
<span class="o">--------------------------------------------</span>
<span class="nv">After</span> <span class="nv">installation</span> <span class="nv">of</span> <span class="nv">limma</span>, <span class="nv">Seurat</span> <span class="nv">will</span> <span class="nv">automatically</span> <span class="nv">use</span> <span class="nv">the</span> <span class="nv">more</span> 
<span class="nv">efficient</span> <span class="nv">implementation</span> <span class="ss">(</span><span class="nv">no</span> <span class="nv">further</span> <span class="nv">action</span> <span class="nv">necessary</span><span class="ss">)</span>.
<span class="nv">This</span> <span class="nv">message</span> <span class="nv">will</span> <span class="nv">be</span> <span class="nv">shown</span> <span class="nv">once</span> <span class="nv">per</span> <span class="nv">session</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">1</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">2</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">3</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">4</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">5</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">6</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">7</span>

<span class="nv">Calculating</span> <span class="nv">cluster</span> <span class="mi">8</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">pbmc.markers</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="dataframe">
<caption>A data.frame: 6 × 7</caption>
<thead>
    <tr><th></th><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th></tr>
    <tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>S100A12</th><td>4.064804e-197</td><td>4.213263</td><td>0.978</td><td>0.076</td><td>1.055467e-192</td><td>0</td><td>S100A12</td></tr>
    <tr><th scope=row>VCAN</th><td>3.352042e-189</td><td>3.484980</td><td>0.996</td><td>0.104</td><td>8.703913e-185</td><td>0</td><td>VCAN   </td></tr>
    <tr><th scope=row>S100A8</th><td>4.318649e-186</td><td>5.520490</td><td>1.000</td><td>0.143</td><td>1.121380e-181</td><td>0</td><td>S100A8 </td></tr>
    <tr><th scope=row>CD14</th><td>7.936720e-173</td><td>2.574749</td><td>0.944</td><td>0.088</td><td>2.060849e-168</td><td>0</td><td>CD14   </td></tr>
    <tr><th scope=row>S100A9</th><td>7.548587e-171</td><td>4.954452</td><td>0.996</td><td>0.197</td><td>1.960066e-166</td><td>0</td><td>S100A9 </td></tr>
    <tr><th scope=row>MNDA</th><td>4.262766e-170</td><td>3.039318</td><td>0.981</td><td>0.130</td><td>1.106870e-165</td><td>0</td><td>MNDA   </td></tr>
</tbody>
</table>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">12</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">8</span><span class="p">)</span>
<span class="nf">plot_gene_rank</span><span class="p">(</span><span class="n">pbmc.markers</span><span class="p">,</span> <span class="m">25</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_88_0.png" /></p>
<p>Student's t test is also supported</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc.markers.t</span> <span class="o">&lt;-</span> <span class="nf">FindAllMarkers</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">test.use</span> <span class="o">=</span> <span class="s">&quot;t&quot;</span><span class="p">,</span> <span class="n">only.pos</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> 
                               <span class="n">min.pct</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">,</span> <span class="n">logfc.threshold</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Calculating cluster 0

Calculating cluster 1

Calculating cluster 2

Calculating cluster 3

Calculating cluster 4

Calculating cluster 5

Calculating cluster 6

Calculating cluster 7

Calculating cluster 8
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plot_gene_rank</span><span class="p">(</span><span class="n">pbmc.markers.t</span><span class="p">,</span> <span class="m">25</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_91_0.png" /></p>
<p>Also logistic regression to test how good each gene is for deciding whether a cell is in a cluster.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pbmc.markers.lr</span> <span class="o">&lt;-</span> <span class="nf">FindAllMarkers</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">test.use</span> <span class="o">=</span> <span class="s">&quot;LR&quot;</span><span class="p">,</span> <span class="n">only.pos</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> 
                               <span class="n">min.pct</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">,</span> <span class="n">logfc.threshold</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Calculating cluster 0

Calculating cluster 1

Calculating cluster 2

Calculating cluster 3

Calculating cluster 4

Calculating cluster 5

Calculating cluster 6

Calculating cluster 7

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Calculating cluster 8

Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: algorithm did not converge”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">plot_gene_rank</span><span class="p">(</span><span class="n">pbmc.markers.lr</span><span class="p">,</span> <span class="m">25</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_94_0.png" /></p>
<p>Seurat includes several tools for visualizing marker expression. VlnPlot (shows expression probability distributions across clusters), and FeaturePlot (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring RidgePlot, CellScatter, and DotPlot as additional methods to view your dataset.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">marker_genes</span> <span class="o">&lt;-</span> <span class="nf">sort</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s">&#39;CD79A&#39;</span><span class="p">,</span> <span class="s">&#39;MS4A1&#39;</span><span class="p">,</span> <span class="s">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s">&#39;CD8B&#39;</span><span class="p">,</span> <span class="s">&#39;LYZ&#39;</span><span class="p">,</span> <span class="s">&#39;CD14&#39;</span><span class="p">,</span>
                <span class="s">&#39;LGALS3&#39;</span><span class="p">,</span> <span class="s">&#39;S100A8&#39;</span><span class="p">,</span> <span class="s">&#39;GNLY&#39;</span><span class="p">,</span> <span class="s">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s">&#39;KLRB1&#39;</span><span class="p">,</span>  
                <span class="s">&#39;FCGR3A&#39;</span><span class="p">,</span> <span class="s">&#39;MS4A7&#39;</span><span class="p">,</span> <span class="s">&#39;FCER1A&#39;</span><span class="p">,</span> <span class="s">&#39;CST3&#39;</span><span class="p">,</span> <span class="s">&#39;PPBP&#39;</span><span class="p">,</span> <span class="s">&#39;CCR7&#39;</span><span class="p">,</span>
                <span class="s">&#39;S100A4&#39;</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">16</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
<span class="nf">VlnPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="m">4</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_97_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">16</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
<span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">ncol</span> <span class="o">=</span> <span class="m">4</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_98_0.png" /></p>
<h3 id="assigning-cell-type-identity-to-clusters">Assigning cell type identity to clusters<a class="headerlink" href="#assigning-cell-type-identity-to-clusters" title="Permanent link">&para;</a></h3>
<p>In this dataset, we can use canonical markers to easily match the <em>de novo</em> clustering to known cell types:</p>
<table>
<thead>
<tr>
<th>Cluster ID</th>
<th>Markers</th>
<th>Cell Type</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr>
<td>1</td>
<td>IL7R, S100A4</td>
<td>Memory CD4+ T</td>
</tr>
<tr>
<td>2</td>
<td>IL7R, CCR7</td>
<td>Naive CD4+</td>
</tr>
<tr>
<td>3</td>
<td>MS4A1, CD79A</td>
<td>B</td>
</tr>
<tr>
<td>4</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr>
<td>5</td>
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr>
<td>6</td>
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr>
<td>7</td>
<td>MS4A1, CD79A</td>
<td>B</td>
</tr>
<tr>
<td>8</td>
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">6</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">7</span><span class="p">)</span>
<span class="nf">DotPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">assay</span> <span class="o">=</span> <span class="s">&quot;RNA&quot;</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">scale.by</span> <span class="o">=</span> <span class="s">&quot;size&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">coord_flip</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_101_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="n">new.cluster.ids</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD14+ Mono&quot;</span><span class="p">,</span> <span class="s">&quot;Memory CD4 T&quot;</span><span class="p">,</span> <span class="s">&quot;Naive CD4 T&quot;</span><span class="p">,</span> <span class="s">&quot;B1&quot;</span><span class="p">,</span> <span class="s">&quot;FCGR3A+ Mono&quot;</span><span class="p">,</span> 
    <span class="s">&quot;NK&quot;</span><span class="p">,</span> <span class="s">&quot;CD8+ T&quot;</span><span class="p">,</span> <span class="s">&quot;B2&quot;</span><span class="p">,</span> <span class="s">&quot;Platelet&quot;</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">new.cluster.ids</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">levels</span><span class="p">(</span><span class="n">pbmc</span><span class="p">)</span>
<span class="n">pbmc</span> <span class="o">&lt;-</span> <span class="nf">RenameIdents</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">new.cluster.ids</span><span class="p">)</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">pbmc</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&quot;umap&quot;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">pt.size</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span> <span class="n">label.size</span> <span class="o">=</span> <span class="m">6</span><span class="p">)</span> <span class="o">+</span> <span class="nf">NoLegend</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_analysis_0_R_files/kb_analysis_0_R_102_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">Sys.time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Time difference of 50.73962 mins
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">sessionInfo</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">R</span> <span class="n">version</span> <span class="mf">4.0</span><span class="o">.</span><span class="mi">4</span> <span class="p">(</span><span class="mi">2021</span><span class="o">-</span><span class="mi">02</span><span class="o">-</span><span class="mi">15</span><span class="p">)</span>
<span class="n">Platform</span><span class="p">:</span> <span class="n">x86_64</span><span class="o">-</span><span class="n">pc</span><span class="o">-</span><span class="n">linux</span><span class="o">-</span><span class="n">gnu</span> <span class="p">(</span><span class="mi">64</span><span class="o">-</span><span class="n">bit</span><span class="p">)</span>
<span class="n">Running</span> <span class="n">under</span><span class="p">:</span> <span class="n">Ubuntu</span> <span class="mf">18.04</span><span class="o">.</span><span class="mi">5</span> <span class="n">LTS</span>

<span class="n">Matrix</span> <span class="n">products</span><span class="p">:</span> <span class="n">default</span>
<span class="n">BLAS</span><span class="p">:</span>   <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">x86_64</span><span class="o">-</span><span class="n">linux</span><span class="o">-</span><span class="n">gnu</span><span class="o">/</span><span class="n">openblas</span><span class="o">/</span><span class="n">libblas</span><span class="o">.</span><span class="n">so</span><span class="o">.</span><span class="mi">3</span>
<span class="n">LAPACK</span><span class="p">:</span> <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">x86_64</span><span class="o">-</span><span class="n">linux</span><span class="o">-</span><span class="n">gnu</span><span class="o">/</span><span class="n">libopenblasp</span><span class="o">-</span><span class="n">r0</span><span class="o">.</span><span class="mf">2.20</span><span class="o">.</span><span class="n">so</span>

<span class="n">locale</span><span class="p">:</span>
 <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="n">LC_CTYPE</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span>       <span class="n">LC_NUMERIC</span><span class="o">=</span><span class="n">C</span>              
 <span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="n">LC_TIME</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span>        <span class="n">LC_COLLATE</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span>    
 <span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="n">LC_MONETARY</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span>    <span class="n">LC_MESSAGES</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span>   
 <span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="n">LC_PAPER</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span>       <span class="n">LC_NAME</span><span class="o">=</span><span class="n">C</span>                 
 <span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="n">LC_ADDRESS</span><span class="o">=</span><span class="n">C</span>               <span class="n">LC_TELEPHONE</span><span class="o">=</span><span class="n">C</span>            
<span class="p">[</span><span class="mi">11</span><span class="p">]</span> <span class="n">LC_MEASUREMENT</span><span class="o">=</span><span class="n">en_US</span><span class="o">.</span><span class="n">UTF</span><span class="o">-</span><span class="mi">8</span> <span class="n">LC_IDENTIFICATION</span><span class="o">=</span><span class="n">C</span>

<span class="n">attached</span> <span class="n">base</span> <span class="n">packages</span><span class="p">:</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="n">stats</span>     <span class="n">graphics</span>  <span class="n">grDevices</span> <span class="n">utils</span>     <span class="n">datasets</span>  <span class="n">methods</span>   <span class="n">base</span>

<span class="n">other</span> <span class="n">attached</span> <span class="n">packages</span><span class="p">:</span>
 <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="n">patchwork_1</span><span class="o">.</span><span class="mf">1.1</span>    <span class="n">forcats_0</span><span class="o">.</span><span class="mf">5.1</span>      <span class="n">stringr_1</span><span class="o">.</span><span class="mf">4.0</span>      <span class="n">dplyr_1</span><span class="o">.</span><span class="mf">0.5</span>       
 <span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="n">purrr_0</span><span class="o">.</span><span class="mf">3.4</span>        <span class="n">readr_1</span><span class="o">.</span><span class="mf">4.0</span>        <span class="n">tidyr_1</span><span class="o">.</span><span class="mf">1.3</span>        <span class="n">tibble_3</span><span class="o">.</span><span class="mf">1.0</span>      
 <span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="n">ggplot2_3</span><span class="o">.</span><span class="mf">3.3</span>      <span class="n">tidyverse_1</span><span class="o">.</span><span class="mf">3.0</span>    <span class="n">Matrix_1</span><span class="o">.</span><span class="mi">3</span><span class="o">-</span><span class="mi">2</span>       <span class="n">SeuratObject_4</span><span class="o">.</span><span class="mf">0.0</span>
<span class="p">[</span><span class="mi">13</span><span class="p">]</span> <span class="n">Seurat_4</span><span class="o">.</span><span class="mf">0.1</span>

<span class="n">loaded</span> <span class="n">via</span> <span class="n">a</span> <span class="n">namespace</span> <span class="p">(</span><span class="ow">and</span> <span class="ow">not</span> <span class="n">attached</span><span class="p">):</span>
  <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="n">Rtsne_0</span><span class="o">.</span><span class="mi">15</span>            <span class="n">colorspace_2</span><span class="o">.</span><span class="mi">0</span><span class="o">-</span><span class="mi">0</span>      <span class="n">deldir_0</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="mi">10</span>        
  <span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="n">ellipsis_0</span><span class="o">.</span><span class="mf">3.1</span>        <span class="n">ggridges_0</span><span class="o">.</span><span class="mf">5.3</span>        <span class="n">IRdisplay_1</span><span class="o">.</span><span class="mi">0</span>        
  <span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="n">fs_1</span><span class="o">.</span><span class="mf">5.0</span>              <span class="n">base64enc_0</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">3</span>       <span class="n">rstudioapi_0</span><span class="o">.</span><span class="mi">13</span>      
 <span class="p">[</span><span class="mi">10</span><span class="p">]</span> <span class="n">spatstat</span><span class="o">.</span><span class="n">data_2</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">0</span>   <span class="n">farver_2</span><span class="o">.</span><span class="mf">1.0</span>          <span class="n">leiden_0</span><span class="o">.</span><span class="mf">3.7</span>         
 <span class="p">[</span><span class="mi">13</span><span class="p">]</span> <span class="n">listenv_0</span><span class="o">.</span><span class="mf">8.0</span>         <span class="n">ggrepel_0</span><span class="o">.</span><span class="mf">9.1</span>         <span class="n">RSpectra_0</span><span class="o">.</span><span class="mi">16</span><span class="o">-</span><span class="mi">0</span>      
 <span class="p">[</span><span class="mi">16</span><span class="p">]</span> <span class="n">lubridate_1</span><span class="o">.</span><span class="mf">7.10</span>      <span class="n">fansi_0</span><span class="o">.</span><span class="mf">4.2</span>           <span class="n">xml2_1</span><span class="o">.</span><span class="mf">3.2</span>           
 <span class="p">[</span><span class="mi">19</span><span class="p">]</span> <span class="n">codetools_0</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="mi">18</span>      <span class="n">splines_4</span><span class="o">.</span><span class="mf">0.4</span>         <span class="n">polyclip_1</span><span class="o">.</span><span class="mi">10</span><span class="o">-</span><span class="mi">0</span>      
 <span class="p">[</span><span class="mi">22</span><span class="p">]</span> <span class="n">IRkernel_1</span><span class="o">.</span><span class="mf">1.1</span>        <span class="n">jsonlite_1</span><span class="o">.</span><span class="mf">7.2</span>        <span class="n">broom_0</span><span class="o">.</span><span class="mf">7.5</span>          
 <span class="p">[</span><span class="mi">25</span><span class="p">]</span> <span class="n">ica_1</span><span class="o">.</span><span class="mi">0</span><span class="o">-</span><span class="mi">2</span>             <span class="n">cluster_2</span><span class="o">.</span><span class="mf">1.1</span>         <span class="n">dbplyr_2</span><span class="o">.</span><span class="mf">1.0</span>         
 <span class="p">[</span><span class="mi">28</span><span class="p">]</span> <span class="n">png_0</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">7</span>             <span class="n">uwot_0</span><span class="o">.</span><span class="mf">1.10</span>           <span class="n">shiny_1</span><span class="o">.</span><span class="mf">6.0</span>          
 <span class="p">[</span><span class="mi">31</span><span class="p">]</span> <span class="n">sctransform_0</span><span class="o">.</span><span class="mf">3.2</span>     <span class="n">spatstat</span><span class="o">.</span><span class="n">sparse_2</span><span class="o">.</span><span class="mi">0</span><span class="o">-</span><span class="mi">0</span> <span class="n">compiler_4</span><span class="o">.</span><span class="mf">0.4</span>       
 <span class="p">[</span><span class="mi">34</span><span class="p">]</span> <span class="n">httr_1</span><span class="o">.</span><span class="mf">4.2</span>            <span class="n">backports_1</span><span class="o">.</span><span class="mf">2.1</span>       <span class="n">assertthat_0</span><span class="o">.</span><span class="mf">2.1</span>     
 <span class="p">[</span><span class="mi">37</span><span class="p">]</span> <span class="n">fastmap_1</span><span class="o">.</span><span class="mf">1.0</span>         <span class="n">lazyeval_0</span><span class="o">.</span><span class="mf">2.2</span>        <span class="n">cli_2</span><span class="o">.</span><span class="mf">3.1</span>            
 <span class="p">[</span><span class="mi">40</span><span class="p">]</span> <span class="n">later_1</span><span class="o">.</span><span class="mf">1.0</span><span class="o">.</span><span class="mi">1</span>         <span class="n">htmltools_0</span><span class="o">.</span><span class="mf">5.1</span><span class="o">.</span><span class="mi">1</span>     <span class="n">tools_4</span><span class="o">.</span><span class="mf">0.4</span>          
 <span class="p">[</span><span class="mi">43</span><span class="p">]</span> <span class="n">igraph_1</span><span class="o">.</span><span class="mf">2.6</span>          <span class="n">gtable_0</span><span class="o">.</span><span class="mf">3.0</span>          <span class="n">glue_1</span><span class="o">.</span><span class="mf">4.2</span>           
 <span class="p">[</span><span class="mi">46</span><span class="p">]</span> <span class="n">RANN_2</span><span class="o">.</span><span class="mf">6.1</span>            <span class="n">reshape2_1</span><span class="o">.</span><span class="mf">4.4</span>        <span class="n">Rcpp_1</span><span class="o">.</span><span class="mf">0.6</span>           
 <span class="p">[</span><span class="mi">49</span><span class="p">]</span> <span class="n">scattermore_0</span><span class="o">.</span><span class="mi">7</span>       <span class="n">cellranger_1</span><span class="o">.</span><span class="mf">1.0</span>      <span class="n">vctrs_0</span><span class="o">.</span><span class="mf">3.6</span>          
 <span class="p">[</span><span class="mi">52</span><span class="p">]</span> <span class="n">nlme_3</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">152</span>          <span class="n">lmtest_0</span><span class="o">.</span><span class="mi">9</span><span class="o">-</span><span class="mi">38</span>         <span class="n">ps_1</span><span class="o">.</span><span class="mf">6.0</span>             
 <span class="p">[</span><span class="mi">55</span><span class="p">]</span> <span class="n">globals_0</span><span class="o">.</span><span class="mf">14.0</span>        <span class="n">rvest_1</span><span class="o">.</span><span class="mf">0.0</span>           <span class="n">mime_0</span><span class="o">.</span><span class="mi">10</span>            
 <span class="p">[</span><span class="mi">58</span><span class="p">]</span> <span class="n">miniUI_0</span><span class="o">.</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">1</span>        <span class="n">lifecycle_1</span><span class="o">.</span><span class="mf">0.0</span>       <span class="n">irlba_2</span><span class="o">.</span><span class="mf">3.3</span>          
 <span class="p">[</span><span class="mi">61</span><span class="p">]</span> <span class="n">goftest_1</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="mi">2</span>         <span class="n">future_1</span><span class="o">.</span><span class="mf">21.0</span>         <span class="n">MASS_7</span><span class="o">.</span><span class="mi">3</span><span class="o">-</span><span class="mf">53.1</span>        
 <span class="p">[</span><span class="mi">64</span><span class="p">]</span> <span class="n">zoo_1</span><span class="o">.</span><span class="mi">8</span><span class="o">-</span><span class="mi">9</span>             <span class="n">scales_1</span><span class="o">.</span><span class="mf">1.1</span>          <span class="n">spatstat</span><span class="o">.</span><span class="n">core_2</span><span class="o">.</span><span class="mi">0</span><span class="o">-</span><span class="mi">0</span>  
 <span class="p">[</span><span class="mi">67</span><span class="p">]</span> <span class="n">hms_1</span><span class="o">.</span><span class="mf">0.0</span>             <span class="n">promises_1</span><span class="o">.</span><span class="mf">2.0</span><span class="o">.</span><span class="mi">1</span>      <span class="n">spatstat</span><span class="o">.</span><span class="n">utils_2</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">0</span> 
 <span class="p">[</span><span class="mi">70</span><span class="p">]</span> <span class="n">parallel_4</span><span class="o">.</span><span class="mf">0.4</span>        <span class="n">RColorBrewer_1</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">2</span>    <span class="n">reticulate_1</span><span class="o">.</span><span class="mi">18</span>      
 <span class="p">[</span><span class="mi">73</span><span class="p">]</span> <span class="n">pbapply_1</span><span class="o">.</span><span class="mi">4</span><span class="o">-</span><span class="mi">3</span>         <span class="n">gridExtra_2</span><span class="o">.</span><span class="mi">3</span>         <span class="n">rpart_4</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">15</span>         
 <span class="p">[</span><span class="mi">76</span><span class="p">]</span> <span class="n">stringi_1</span><span class="o">.</span><span class="mf">5.3</span>         <span class="n">repr_1</span><span class="o">.</span><span class="mf">1.3</span>            <span class="n">rlang_0</span><span class="o">.</span><span class="mf">4.10</span>         
 <span class="p">[</span><span class="mi">79</span><span class="p">]</span> <span class="n">pkgconfig_2</span><span class="o">.</span><span class="mf">0.3</span>       <span class="n">matrixStats_0</span><span class="o">.</span><span class="mf">58.0</span>    <span class="n">evaluate_0</span><span class="o">.</span><span class="mi">14</span>        
 <span class="p">[</span><span class="mi">82</span><span class="p">]</span> <span class="n">lattice_0</span><span class="o">.</span><span class="mi">20</span><span class="o">-</span><span class="mi">41</span>       <span class="n">ROCR_1</span><span class="o">.</span><span class="mi">0</span><span class="o">-</span><span class="mi">11</span>           <span class="n">tensor_1</span><span class="o">.</span><span class="mi">5</span>           
 <span class="p">[</span><span class="mi">85</span><span class="p">]</span> <span class="n">labeling_0</span><span class="o">.</span><span class="mf">4.2</span>        <span class="n">htmlwidgets_1</span><span class="o">.</span><span class="mf">5.3</span>     <span class="n">cowplot_1</span><span class="o">.</span><span class="mf">1.1</span>        
 <span class="p">[</span><span class="mi">88</span><span class="p">]</span> <span class="n">tidyselect_1</span><span class="o">.</span><span class="mf">1.0</span>      <span class="n">parallelly_1</span><span class="o">.</span><span class="mf">24.0</span>     <span class="n">RcppAnnoy_0</span><span class="o">.</span><span class="mf">0.18</span>     
 <span class="p">[</span><span class="mi">91</span><span class="p">]</span> <span class="n">plyr_1</span><span class="o">.</span><span class="mf">8.6</span>            <span class="n">magrittr_2</span><span class="o">.</span><span class="mf">0.1</span>        <span class="n">R6_2</span><span class="o">.</span><span class="mf">5.0</span>             
 <span class="p">[</span><span class="mi">94</span><span class="p">]</span> <span class="n">generics_0</span><span class="o">.</span><span class="mf">1.0</span>        <span class="n">pbdZMQ_0</span><span class="o">.</span><span class="mi">3</span><span class="o">-</span><span class="mi">5</span>          <span class="n">DBI_1</span><span class="o">.</span><span class="mf">1.1</span>            
 <span class="p">[</span><span class="mi">97</span><span class="p">]</span> <span class="n">withr_2</span><span class="o">.</span><span class="mf">4.1</span>           <span class="n">haven_2</span><span class="o">.</span><span class="mf">3.1</span>           <span class="n">pillar_1</span><span class="o">.</span><span class="mf">5.1</span>         
<span class="p">[</span><span class="mi">100</span><span class="p">]</span> <span class="n">mgcv_1</span><span class="o">.</span><span class="mi">8</span><span class="o">-</span><span class="mi">34</span>           <span class="n">fitdistrplus_1</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">3</span>    <span class="n">survival_3</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="mi">10</span>      
<span class="p">[</span><span class="mi">103</span><span class="p">]</span> <span class="n">abind_1</span><span class="o">.</span><span class="mi">4</span><span class="o">-</span><span class="mi">5</span>           <span class="n">future</span><span class="o">.</span><span class="n">apply_1</span><span class="o">.</span><span class="mf">7.0</span>    <span class="n">modelr_0</span><span class="o">.</span><span class="mf">1.8</span>         
<span class="p">[</span><span class="mi">106</span><span class="p">]</span> <span class="n">crayon_1</span><span class="o">.</span><span class="mf">4.1</span>          <span class="n">uuid_0</span><span class="o">.</span><span class="mi">1</span><span class="o">-</span><span class="mi">4</span>            <span class="n">KernSmooth_2</span><span class="o">.</span><span class="mi">23</span><span class="o">-</span><span class="mi">18</span>   
<span class="p">[</span><span class="mi">109</span><span class="p">]</span> <span class="n">utf8_1</span><span class="o">.</span><span class="mf">2.1</span>            <span class="n">spatstat</span><span class="o">.</span><span class="n">geom_2</span><span class="o">.</span><span class="mi">0</span><span class="o">-</span><span class="mi">1</span>   <span class="n">plotly_4</span><span class="o">.</span><span class="mf">9.3</span>         
<span class="p">[</span><span class="mi">112</span><span class="p">]</span> <span class="n">readxl_1</span><span class="o">.</span><span class="mf">3.1</span>          <span class="n">grid_4</span><span class="o">.</span><span class="mf">0.4</span>            <span class="n">data</span><span class="o">.</span><span class="n">table_1</span><span class="o">.</span><span class="mf">14.0</span>    
<span class="p">[</span><span class="mi">115</span><span class="p">]</span> <span class="n">reprex_1</span><span class="o">.</span><span class="mf">0.0</span>          <span class="n">digest_0</span><span class="o">.</span><span class="mf">6.27</span>         <span class="n">xtable_1</span><span class="o">.</span><span class="mi">8</span><span class="o">-</span><span class="mi">4</span>         
<span class="p">[</span><span class="mi">118</span><span class="p">]</span> <span class="n">httpuv_1</span><span class="o">.</span><span class="mf">5.5</span>          <span class="n">munsell_0</span><span class="o">.</span><span class="mf">5.0</span>         <span class="n">viridisLite_0</span><span class="o">.</span><span class="mf">3.0</span>
</code></pre></div>

<p><strong>Feedback</strong>: please report any issues, or submit pull requests for improvements, in the <a href="https://github.com/pachterlab/kallistobustools/blob/master/notebooks/kb_analysis_0_R.ipynb">Github repository where this notebook is located</a>.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../python/kb_analysis_0_python.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Python" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Python
            </div>
          </div>
        </a>
      
      
        
        <a href="../../kb_pseudotime/R/slingshot/kb_slingshot.html" class="md-footer__link md-footer__link--next" aria-label="Next: R" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              R
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../assets/javascripts/workers/search.477d984a.min.js", "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.82b56eb2.min.js"></script>
      
    
  </body>
</html>