
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.1, mkdocs-material-7.1.9">
    
    
      
        <title>R - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.ca7ac06f.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.f1a3b89f.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../css/custom.css">
    
    
      
  



  
  



  <script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
  <script async src="https://www.google-analytics.com/analytics.js"></script>


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    <script>function __prefix(e){return new URL("../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#introduction-to-single-cell-rna-seq-ii-getting-started-with-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../index.html" title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              R
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../index.html" title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../index.html" class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../applications.html" class="md-nav__link">
        Applications
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Using kb
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_usage.html" class="md-nav__link">
        Introduction
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_ref.html" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_usage/kb_count.html" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      <label class="md-nav__link" for="__nav_4">
        Tutorials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_1" type="checkbox" id="__nav_4_1" >
      
      <label class="md-nav__link" for="__nav_4_1">
        Introduction to single-cell RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../scRNA-seq_intro/python/scRNA-seq_intro.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_2" type="checkbox" id="__nav_4_2" >
      
      <label class="md-nav__link" for="__nav_4_2">
        Preprocessing & Quality Control
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/python/kb_intro_1_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_quality_control/R/kb_intro_1_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_3" type="checkbox" id="__nav_4_3" checked>
      
      <label class="md-nav__link" for="__nav_4_3">
        Getting started with analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../python/kb_intro_2_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          R
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="kb_intro_2_R.html" class="md-nav__link md-nav__link--active">
        R
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-r-packages" class="md-nav__link">
    Install R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb-python" class="md-nav__link">
    Install kb-python
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filter-empty-droplets" class="md-nav__link">
    Filter empty droplets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-count-distributions" class="md-nav__link">
    Visualizing count distributions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#identify-highly-variable-genes" class="md-nav__link">
    Identify highly variable genes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pca" class="md-nav__link">
    PCA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering-and-visualization" class="md-nav__link">
    Clustering and visualization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tsne" class="md-nav__link">
    tSNE
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umap" class="md-nav__link">
    UMAP
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" >
      
      <label class="md-nav__link" for="__nav_4_4">
        Building & annotating an atlas
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_building_atlas/python/kb_analysis_0_python.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_building_atlas/R/kb_analysis_0_R.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      <label class="md-nav__link" for="__nav_4_5">
        Pseudotime analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/slingshot/kb_slingshot.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      <label class="md-nav__link" for="__nav_4_6">
        Single-nuclei RNA-seq
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_nucleus/python/kb_single_nucleus.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      <label class="md-nav__link" for="__nav_4_7">
        RNA Velocity
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity/python/kb_velocity.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" >
      
      <label class="md-nav__link" for="__nav_4_8">
        Feature barcode & multimodal data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_kite/python/kb_kite.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../download_data/data_download.html" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      <label class="md-nav__link" for="__nav_4_10">
        Build custom index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_custom_index/python/kb_transcriptome_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_11" type="checkbox" id="__nav_4_11" >
      
      <label class="md-nav__link" for="__nav_4_11">
        Build RNA velocity index
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_velocity_index/python/kb_velocity_index.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_12" type="checkbox" id="__nav_4_12" >
      
      <label class="md-nav__link" for="__nav_4_12">
        Aggregate multiple matrices
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_aggregate/python/kb_aggregating_count_matrices.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_13" type="checkbox" id="__nav_4_13" >
      
      <label class="md-nav__link" for="__nav_4_13">
        Process multiple FASTQs
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_multiple_fastqs/python/kb_multiple_files.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_14" type="checkbox" id="__nav_4_14" >
      
      <label class="md-nav__link" for="__nav_4_14">
        Processing multi-species data
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_species_mixing/python/kb_species_mixing.html" class="md-nav__link">
        Python
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_species_mixing/R/kb_mixed_species_10x_v2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
            
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_15" type="checkbox" id="__nav_4_15" >
      
      <label class="md-nav__link" for="__nav_4_15">
        Pseudotime analysis (Monocle2)
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../kb_pseudotime/R/monocle2/kb_monocle2.html" class="md-nav__link">
        R
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-r-packages" class="md-nav__link">
    Install R packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb-python" class="md-nav__link">
    Install kb-python
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filter-empty-droplets" class="md-nav__link">
    Filter empty droplets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-count-distributions" class="md-nav__link">
    Visualizing count distributions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#identify-highly-variable-genes" class="md-nav__link">
    Identify highly variable genes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pca" class="md-nav__link">
    PCA
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering-and-visualization" class="md-nav__link">
    Clustering and visualization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tsne" class="md-nav__link">
    tSNE
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umap" class="md-nav__link">
    UMAP
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/tree/master/tutorials/docs/tutorials/kb_getting_started/R/kb_intro_2_R.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/tutorials/docs/tutorials/kb_getting_started/R/kb_intro_2_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="introduction-to-single-cell-rna-seq-ii-getting-started-with-analysis">Introduction to single-cell RNA-seq II: getting started with analysis<a class="headerlink" href="#introduction-to-single-cell-rna-seq-ii-getting-started-with-analysis" title="Permanent link">&para;</a></h1>
<p>This notebook demonstrates pre-processing and basic analysis of the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE126783">mouse retinal cells GSE126783</a> dataset from <a href="https://doi.org/10.1016/j.immuni.2019.02.007">Koren et al., 2019</a>. Following pre-processing using kallisto and bustools and basic QC, the notebook demonstrates some initial analysis. The approximate running time of the notebook is 12 minutes.</p>
<p>The notebook was written by Kyung Hoi (Joseph) Min, Lambda Lu, A. Sina Booeshaghi and Lior Pachter. If you use the methods in this notebook for your analysis please cite the following publications which describe the tools used in the notebook, as well as specific methods they run (these are cited inline in the notebook):</p>
<ul>
<li>Melsted, P., Booeshaghi, A.S. et al. Modular and efficient pre-processing of single-cell RNA-seq. bioRxiv (2019). doi:10.1101/673285</li>
<li>McCarthy, D.J., Campbell, K.R., Lun, A.T. and Wills, Q.F. Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R (2017). doi.org/10.1093/bioinformatics/btw777</li>
</ul>
<p>A Python notebook implementing the same analysis is available <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_intro_2_python.ipynb">here</a>. See the <a href="https://www.kallistobus.tools/tutorials">kallistobus.tools tutorials</a> site for additional notebooks demonstrating other analyses.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># This is  used to time the running of the notebook</span>
<span class="n">start_time</span> <span class="o">&lt;-</span> <span class="nf">Sys.time</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<h3 id="install-r-packages">Install R packages<a class="headerlink" href="#install-r-packages" title="Permanent link">&para;</a></h3>
<p>There are several packages in R built for scRNA-seq data analysis. Here we use Seurat.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system.time</span><span class="p">({</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
  <span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;multtest&#39;</span><span class="p">,</span> <span class="s">&quot;DropletUtils&quot;</span><span class="p">),</span> <span class="n">Ncpus</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
  <span class="nf">install.packages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Seurat&quot;</span><span class="p">,</span> <span class="s">&quot;scico&quot;</span><span class="p">,</span> <span class="s">&quot;ggpointdensity&quot;</span><span class="p">),</span> <span class="n">Ncpus</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
<span class="p">})</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Installing</span> <span class="n">package</span> <span class="n">into</span> <span class="err">‘</span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">library</span><span class="err">’</span>
<span class="p">(</span><span class="k">as</span> <span class="err">‘</span><span class="n">lib</span><span class="err">’</span> <span class="k">is</span> <span class="n">unspecified</span><span class="p">)</span>

<span class="s1">&#39;getOption(&quot;repos&quot;)&#39;</span> <span class="n">replaces</span> <span class="n">Bioconductor</span> <span class="n">standard</span> <span class="n">repositories</span><span class="p">,</span> <span class="n">see</span>
<span class="s1">&#39;?repositories&#39;</span> <span class="k">for</span> <span class="n">details</span>

<span class="n">replacement</span> <span class="n">repositories</span><span class="p">:</span>
    <span class="n">CRAN</span><span class="p">:</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">cran</span><span class="o">.</span><span class="n">rstudio</span><span class="o">.</span><span class="n">com</span>


<span class="n">Bioconductor</span> <span class="n">version</span> <span class="mf">3.12</span> <span class="p">(</span><span class="n">BiocManager</span> <span class="mf">1.30</span><span class="o">.</span><span class="mi">12</span><span class="p">),</span> <span class="n">R</span> <span class="mf">4.0</span><span class="o">.</span><span class="mi">4</span> <span class="p">(</span><span class="mi">2021</span><span class="o">-</span><span class="mi">02</span><span class="o">-</span><span class="mi">15</span><span class="p">)</span>

<span class="n">Installing</span> <span class="n">package</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="s1">&#39;BiocVersion&#39;</span><span class="p">,</span> <span class="s1">&#39;multtest&#39;</span><span class="p">,</span> <span class="s1">&#39;DropletUtils&#39;</span>

<span class="n">also</span> <span class="n">installing</span> <span class="n">the</span> <span class="n">dependencies</span> <span class="err">‘</span><span class="n">zlibbioc</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bitops</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">formatR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">matrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">XVector</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RCurl</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomeInfoDbData</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lambda</span><span class="o">.</span><span class="n">r</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">futile</span><span class="o">.</span><span class="n">options</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sparseMatrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">MatrixGenerics</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomicRanges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">IRanges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">GenomeInfoDb</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">futile</span><span class="o">.</span><span class="n">logger</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">snow</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rhdf5filters</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">limma</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">locfit</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">oo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">methodsS3</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sitmo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DelayedMatrixStats</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocGenerics</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Biobase</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SingleCellExperiment</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">S4Vectors</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SummarizedExperiment</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">BiocParallel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">DelayedArray</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">HDF5Array</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">rhdf5</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">edgeR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">R</span><span class="o">.</span><span class="n">utils</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">dqrng</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">beachmat</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">scuttle</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rhdf5lib</span><span class="err">’</span>


<span class="n">Old</span> <span class="n">packages</span><span class="p">:</span> <span class="s1">&#39;callr&#39;</span><span class="p">,</span> <span class="s1">&#39;cpp11&#39;</span><span class="p">,</span> <span class="s1">&#39;gert&#39;</span><span class="p">,</span> <span class="s1">&#39;tinytex&#39;</span><span class="p">,</span> <span class="s1">&#39;vctrs&#39;</span>

<span class="n">Installing</span> <span class="n">packages</span> <span class="n">into</span> <span class="err">‘</span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">R</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">library</span><span class="err">’</span>
<span class="p">(</span><span class="k">as</span> <span class="err">‘</span><span class="n">lib</span><span class="err">’</span> <span class="k">is</span> <span class="n">unspecified</span><span class="p">)</span>

<span class="n">also</span> <span class="n">installing</span> <span class="n">the</span> <span class="n">dependencies</span> <span class="err">‘</span><span class="n">gtools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">caTools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sass</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">jquerylib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">globals</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">listenv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">parallelly</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plyr</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">zoo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gplots</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">reshape2</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">gridExtra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppArmadillo</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">httpuv</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">xtable</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sourcetools</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">bslib</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">data</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">utils</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">sparse</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">abind</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">tensor</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">goftest</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">deldir</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">polyclip</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">FNN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RSpectra</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">cowplot</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">fitdistrplus</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">future</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">future</span><span class="o">.</span><span class="n">apply</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggrepel</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ggridges</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ica</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">igraph</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">irlba</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">leiden</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">lmtest</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">miniUI</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">patchwork</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">pbapply</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">plotly</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">png</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RANN</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppAnnoy</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">reticulate</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">ROCR</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">Rtsne</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">scattermore</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">sctransform</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">SeuratObject</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">shiny</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">core</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">spatstat</span><span class="o">.</span><span class="n">geom</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">uwot</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppEigen</span><span class="err">’</span><span class="p">,</span> <span class="err">‘</span><span class="n">RcppProgress</span><span class="err">’</span>





    <span class="n">user</span>   <span class="n">system</span>  <span class="n">elapsed</span> 
<span class="mf">2827.978</span>  <span class="mf">267.740</span> <span class="mf">1771.560</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">DropletUtils</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggpointdensity</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">scico</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">scales</span><span class="p">)</span>
<span class="nf">theme_set</span><span class="p">(</span><span class="nf">theme_bw</span><span class="p">())</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">SingleCellExperiment</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">SummarizedExperiment</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">MatrixGenerics</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">matrixStats</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">MatrixGenerics</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">matrixStats</span><span class="err">’</span><span class="p">:</span>

    <span class="n">colAlls</span><span class="p">,</span> <span class="n">colAnyNAs</span><span class="p">,</span> <span class="n">colAnys</span><span class="p">,</span> <span class="n">colAvgsPerRowSet</span><span class="p">,</span> <span class="n">colCollapse</span><span class="p">,</span>
    <span class="n">colCounts</span><span class="p">,</span> <span class="n">colCummaxs</span><span class="p">,</span> <span class="n">colCummins</span><span class="p">,</span> <span class="n">colCumprods</span><span class="p">,</span> <span class="n">colCumsums</span><span class="p">,</span>
    <span class="n">colDiffs</span><span class="p">,</span> <span class="n">colIQRDiffs</span><span class="p">,</span> <span class="n">colIQRs</span><span class="p">,</span> <span class="n">colLogSumExps</span><span class="p">,</span> <span class="n">colMadDiffs</span><span class="p">,</span>
    <span class="n">colMads</span><span class="p">,</span> <span class="n">colMaxs</span><span class="p">,</span> <span class="n">colMeans2</span><span class="p">,</span> <span class="n">colMedians</span><span class="p">,</span> <span class="n">colMins</span><span class="p">,</span> <span class="n">colOrderStats</span><span class="p">,</span>
    <span class="n">colProds</span><span class="p">,</span> <span class="n">colQuantiles</span><span class="p">,</span> <span class="n">colRanges</span><span class="p">,</span> <span class="n">colRanks</span><span class="p">,</span> <span class="n">colSdDiffs</span><span class="p">,</span> <span class="n">colSds</span><span class="p">,</span>
    <span class="n">colSums2</span><span class="p">,</span> <span class="n">colTabulates</span><span class="p">,</span> <span class="n">colVarDiffs</span><span class="p">,</span> <span class="n">colVars</span><span class="p">,</span> <span class="n">colWeightedMads</span><span class="p">,</span>
    <span class="n">colWeightedMeans</span><span class="p">,</span> <span class="n">colWeightedMedians</span><span class="p">,</span> <span class="n">colWeightedSds</span><span class="p">,</span>
    <span class="n">colWeightedVars</span><span class="p">,</span> <span class="n">rowAlls</span><span class="p">,</span> <span class="n">rowAnyNAs</span><span class="p">,</span> <span class="n">rowAnys</span><span class="p">,</span> <span class="n">rowAvgsPerColSet</span><span class="p">,</span>
    <span class="n">rowCollapse</span><span class="p">,</span> <span class="n">rowCounts</span><span class="p">,</span> <span class="n">rowCummaxs</span><span class="p">,</span> <span class="n">rowCummins</span><span class="p">,</span> <span class="n">rowCumprods</span><span class="p">,</span>
    <span class="n">rowCumsums</span><span class="p">,</span> <span class="n">rowDiffs</span><span class="p">,</span> <span class="n">rowIQRDiffs</span><span class="p">,</span> <span class="n">rowIQRs</span><span class="p">,</span> <span class="n">rowLogSumExps</span><span class="p">,</span>
    <span class="n">rowMadDiffs</span><span class="p">,</span> <span class="n">rowMads</span><span class="p">,</span> <span class="n">rowMaxs</span><span class="p">,</span> <span class="n">rowMeans2</span><span class="p">,</span> <span class="n">rowMedians</span><span class="p">,</span> <span class="n">rowMins</span><span class="p">,</span>
    <span class="n">rowOrderStats</span><span class="p">,</span> <span class="n">rowProds</span><span class="p">,</span> <span class="n">rowQuantiles</span><span class="p">,</span> <span class="n">rowRanges</span><span class="p">,</span> <span class="n">rowRanks</span><span class="p">,</span>
    <span class="n">rowSdDiffs</span><span class="p">,</span> <span class="n">rowSds</span><span class="p">,</span> <span class="n">rowSums2</span><span class="p">,</span> <span class="n">rowTabulates</span><span class="p">,</span> <span class="n">rowVarDiffs</span><span class="p">,</span> <span class="n">rowVars</span><span class="p">,</span>
    <span class="n">rowWeightedMads</span><span class="p">,</span> <span class="n">rowWeightedMeans</span><span class="p">,</span> <span class="n">rowWeightedMedians</span><span class="p">,</span>
    <span class="n">rowWeightedSds</span><span class="p">,</span> <span class="n">rowWeightedVars</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">GenomicRanges</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">stats4</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">BiocGenerics</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">parallel</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">BiocGenerics</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">parallel</span><span class="err">’</span><span class="p">:</span>

    <span class="n">clusterApply</span><span class="p">,</span> <span class="n">clusterApplyLB</span><span class="p">,</span> <span class="n">clusterCall</span><span class="p">,</span> <span class="n">clusterEvalQ</span><span class="p">,</span>
    <span class="n">clusterExport</span><span class="p">,</span> <span class="n">clusterMap</span><span class="p">,</span> <span class="n">parApply</span><span class="p">,</span> <span class="n">parCapply</span><span class="p">,</span> <span class="n">parLapply</span><span class="p">,</span>
    <span class="n">parLapplyLB</span><span class="p">,</span> <span class="n">parRapply</span><span class="p">,</span> <span class="n">parSapply</span><span class="p">,</span> <span class="n">parSapplyLB</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">stats</span><span class="err">’</span><span class="p">:</span>

    <span class="n">IQR</span><span class="p">,</span> <span class="n">mad</span><span class="p">,</span> <span class="n">sd</span><span class="p">,</span> <span class="k">var</span><span class="p">,</span> <span class="n">xtabs</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">base</span><span class="err">’</span><span class="p">:</span>

    <span class="n">anyDuplicated</span><span class="p">,</span> <span class="n">append</span><span class="p">,</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">,</span> <span class="n">basename</span><span class="p">,</span> <span class="n">cbind</span><span class="p">,</span> <span class="n">colnames</span><span class="p">,</span>
    <span class="n">dirname</span><span class="p">,</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">,</span> <span class="n">duplicated</span><span class="p">,</span> <span class="n">eval</span><span class="p">,</span> <span class="n">evalq</span><span class="p">,</span> <span class="n">Filter</span><span class="p">,</span> <span class="n">Find</span><span class="p">,</span> <span class="n">get</span><span class="p">,</span> <span class="n">grep</span><span class="p">,</span>
    <span class="n">grepl</span><span class="p">,</span> <span class="n">intersect</span><span class="p">,</span> <span class="k">is</span><span class="o">.</span><span class="n">unsorted</span><span class="p">,</span> <span class="n">lapply</span><span class="p">,</span> <span class="n">Map</span><span class="p">,</span> <span class="n">mapply</span><span class="p">,</span> <span class="k">match</span><span class="p">,</span> <span class="n">mget</span><span class="p">,</span>
    <span class="n">order</span><span class="p">,</span> <span class="n">paste</span><span class="p">,</span> <span class="n">pmax</span><span class="p">,</span> <span class="n">pmax</span><span class="o">.</span><span class="n">int</span><span class="p">,</span> <span class="n">pmin</span><span class="p">,</span> <span class="n">pmin</span><span class="o">.</span><span class="n">int</span><span class="p">,</span> <span class="n">Position</span><span class="p">,</span> <span class="n">rank</span><span class="p">,</span>
    <span class="n">rbind</span><span class="p">,</span> <span class="n">Reduce</span><span class="p">,</span> <span class="n">rownames</span><span class="p">,</span> <span class="n">sapply</span><span class="p">,</span> <span class="n">setdiff</span><span class="p">,</span> <span class="n">sort</span><span class="p">,</span> <span class="n">table</span><span class="p">,</span> <span class="n">tapply</span><span class="p">,</span>
    <span class="n">union</span><span class="p">,</span> <span class="n">unique</span><span class="p">,</span> <span class="n">unsplit</span><span class="p">,</span> <span class="n">which</span><span class="o">.</span><span class="n">max</span><span class="p">,</span> <span class="n">which</span><span class="o">.</span><span class="n">min</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">S4Vectors</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">S4Vectors</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">base</span><span class="err">’</span><span class="p">:</span>

    <span class="n">expand</span><span class="o">.</span><span class="n">grid</span>


<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">IRanges</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">GenomeInfoDb</span>

<span class="n">Loading</span> <span class="n">required</span> <span class="n">package</span><span class="p">:</span> <span class="n">Biobase</span>

<span class="n">Welcome</span> <span class="n">to</span> <span class="n">Bioconductor</span>

    <span class="n">Vignettes</span> <span class="n">contain</span> <span class="n">introductory</span> <span class="n">material</span><span class="p">;</span> <span class="n">view</span> <span class="n">with</span>
    <span class="s1">&#39;browseVignettes()&#39;</span><span class="o">.</span> <span class="n">To</span> <span class="n">cite</span> <span class="n">Bioconductor</span><span class="p">,</span> <span class="n">see</span>
    <span class="s1">&#39;citation(&quot;Biobase&quot;)&#39;</span><span class="p">,</span> <span class="ow">and</span> <span class="k">for</span> <span class="n">packages</span> <span class="s1">&#39;citation(&quot;pkgname&quot;)&#39;</span><span class="o">.</span>



<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">Biobase</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">MatrixGenerics</span><span class="err">’</span><span class="p">:</span>

    <span class="n">rowMedians</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">objects</span> <span class="n">are</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">matrixStats</span><span class="err">’</span><span class="p">:</span>

    <span class="n">anyMissing</span><span class="p">,</span> <span class="n">rowMedians</span>



<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">Matrix</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">S4Vectors</span><span class="err">’</span><span class="p">:</span>

    <span class="n">expand</span>


<span class="n">Warning</span> <span class="n">message</span> <span class="ow">in</span> <span class="n">system</span><span class="p">(</span><span class="s2">&quot;timedatectl&quot;</span><span class="p">,</span> <span class="n">intern</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">):</span>
<span class="err">“</span><span class="n">running</span> <span class="n">command</span> <span class="s1">&#39;timedatectl&#39;</span> <span class="n">had</span> <span class="n">status</span> <span class="mi">1</span><span class="err">”</span>
<span class="err">──</span> <span class="err"></span><span class="p">[</span><span class="mi">1</span><span class="n">mAttaching</span> <span class="n">packages</span><span class="err"></span><span class="p">[</span><span class="mi">22</span><span class="n">m</span> <span class="err">───────────────────────────────────────</span> <span class="n">tidyverse</span> <span class="mf">1.3</span><span class="o">.</span><span class="mi">0</span> <span class="err">──</span>

<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mggplot2</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">3.3</span><span class="o">.</span><span class="mi">3</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mpurrr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">0.3</span><span class="o">.</span><span class="mi">4</span>
<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtibble</span> <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">3.1</span><span class="o">.</span><span class="mi">0</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.0</span><span class="o">.</span><span class="mi">5</span>
<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.1</span><span class="o">.</span><span class="mi">3</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mstringr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.4</span><span class="o">.</span><span class="mi">0</span>
<span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mreadr</span>  <span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">1.4</span><span class="o">.</span><span class="mi">0</span>     <span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">m</span><span class="err">✔</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mforcats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="mf">0.5</span><span class="o">.</span><span class="mi">1</span>

<span class="err">──</span> <span class="err"></span><span class="p">[</span><span class="mi">1</span><span class="n">mConflicts</span><span class="err"></span><span class="p">[</span><span class="mi">22</span><span class="n">m</span> <span class="err">──────────────────────────────────────────</span> <span class="n">tidyverse_conflicts</span><span class="p">()</span> <span class="err">──</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mcollapse</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>   <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">collapse</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mcombine</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>    <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mBiobase</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">combine</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mBiocGenerics</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">combine</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mcount</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>      <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mmatrixStats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">count</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mdesc</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>       <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">desc</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mexpand</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mMatrix</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">expand</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mS4Vectors</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">expand</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mfilter</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mstats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">filter</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mfirst</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>      <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mS4Vectors</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">first</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mlag</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>        <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mstats</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">lag</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mpack</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>       <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mMatrix</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">pack</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mggplot2</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mPosition</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mBiocGenerics</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">Position</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mbase</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">Position</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mpurrr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mreduce</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mGenomicRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">reduce</span><span class="p">(),</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">reduce</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mrename</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mS4Vectors</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">rename</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mdplyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">mslice</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>      <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mIRanges</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">slice</span><span class="p">()</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">m</span><span class="err">✖</span><span class="p">[</span><span class="mi">39</span><span class="n">m</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mtidyr</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="err"></span><span class="p">[</span><span class="mi">32</span><span class="n">munpack</span><span class="p">()</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span>     <span class="n">masks</span> <span class="err"></span><span class="p">[</span><span class="mi">34</span><span class="n">mMatrix</span><span class="err"></span><span class="p">[</span><span class="mi">39</span><span class="n">m</span><span class="p">::</span><span class="n">unpack</span><span class="p">()</span>

<span class="n">Registered</span> <span class="n">S3</span> <span class="n">method</span> <span class="n">overwritten</span> <span class="n">by</span> <span class="s1">&#39;spatstat.geom&#39;</span><span class="p">:</span>
  <span class="n">method</span>     <span class="n">from</span>
  <span class="nb">print</span><span class="o">.</span><span class="n">boxx</span> <span class="n">cli</span>

<span class="n">Attaching</span> <span class="n">SeuratObject</span>


<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">Seurat</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">SummarizedExperiment</span><span class="err">’</span><span class="p">:</span>

    <span class="n">Assays</span>



<span class="n">Attaching</span> <span class="n">package</span><span class="p">:</span> <span class="err">‘</span><span class="n">scales</span><span class="err">’</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">purrr</span><span class="err">’</span><span class="p">:</span>

    <span class="n">discard</span>


<span class="n">The</span> <span class="n">following</span> <span class="n">object</span> <span class="k">is</span> <span class="n">masked</span> <span class="n">from</span> <span class="err">‘</span><span class="n">package</span><span class="p">:</span><span class="n">readr</span><span class="err">’</span><span class="p">:</span>

    <span class="n">col_factor</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Slightly modified from BUSpaRse, just to avoid installing a few dependencies not used here</span>
<span class="n">read_count_output</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">dir</span> <span class="o">&lt;-</span> <span class="nf">normalizePath</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="n">mustWork</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
  <span class="n">m</span> <span class="o">&lt;-</span> <span class="nf">readMM</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="s">&quot;/&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;.mtx&quot;</span><span class="p">))</span>
  <span class="n">m</span> <span class="o">&lt;-</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">t</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
  <span class="n">m</span> <span class="o">&lt;-</span> <span class="nf">as</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="s">&quot;dgCMatrix&quot;</span><span class="p">)</span>
  <span class="c1"># The matrix read has cells in rows</span>
  <span class="n">ge</span> <span class="o">&lt;-</span> <span class="s">&quot;.genes.txt&quot;</span>
  <span class="n">genes</span> <span class="o">&lt;-</span> <span class="nf">readLines</span><span class="p">(</span><span class="nf">file</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="s">&quot;/&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ge</span><span class="p">)))</span>
  <span class="n">barcodes</span> <span class="o">&lt;-</span> <span class="nf">readLines</span><span class="p">(</span><span class="nf">file</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span> <span class="s">&quot;/&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;.barcodes.txt&quot;</span><span class="p">)))</span>
  <span class="nf">colnames</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">barcodes</span>
  <span class="nf">rownames</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">genes</span>
  <span class="nf">return</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<h3 id="install-kb-python">Install kb-python<a class="headerlink" href="#install-kb-python" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;pip3 install kb-python&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h3 id="download-the-data">Download the data<a class="headerlink" href="#download-the-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;https://caltech.box.com/shared/static/w9ww8et5o029s2e3usjzpbq8lpot29rh.gz&quot;</span><span class="p">,</span> 
<span class="n">destfile</span> <span class="o">=</span> <span class="s">&quot;SRR8599150_S1_L001_R1_001.fastq.gz&quot;</span><span class="p">)</span>
<span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;https://caltech.box.com/shared/static/ql00zyvqnpy7bf8ogdoe9zfy907guzy9.gz&quot;</span><span class="p">,</span>
<span class="n">destfile</span> <span class="o">=</span> <span class="s">&quot;SRR8599150_S1_L001_R2_001.fastq.gz&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Download an index</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;kb ref -d mouse -i index.idx -g t2g.txt&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="pseudoalignment-and-counting">Pseudoalignment and counting<a class="headerlink" href="#pseudoalignment-and-counting" title="Permanent link">&para;</a></h2>
<h3 id="run-kallisto-and-bustools">Run kallisto and bustools<a class="headerlink" href="#run-kallisto-and-bustools" title="Permanent link">&para;</a></h3>
<p>The following command will generate an RNA count matrix of cells (rows) by genes (columns) in H5AD format, which is a binary format used to store <a href="https://anndata.readthedocs.io/en/stable/">Anndata</a> objects. Notice that this requires providing the index and transcript-to-gene mapping downloaded in the previous step to the <code>-i</code> and <code>-g</code> arguments respectively. Also, since the reads were generated with the 10x Genomics Chromium Single Cell v2 Chemistry, the <code>-x 10xv2</code> argument is used. To view other supported technologies, run <code>kb --list</code>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">system</span><span class="p">(</span><span class="s">&quot;kb count -i index.idx -g t2g.txt -x 10xv2 -t2 -o . SRR8599150_S1_L001_R1_001.fastq.gz SRR8599150_S1_L001_R2_001.fastq.gz&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span> <span class="n">recursive</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'10xv2_whitelist.txt'</li><li>'counts_unfiltered/cells_x_genes.barcodes.txt'</li><li>'counts_unfiltered/cells_x_genes.genes.txt'</li><li>'counts_unfiltered/cells_x_genes.mtx'</li><li>'index.idx'</li><li>'inspect.json'</li><li>'kb_info.json'</li><li>'matrix.ec'</li><li>'output.bus'</li><li>'output.unfiltered.bus'</li><li>'run_info.json'</li><li>'sample_data/anscombe.json'</li><li>'sample_data/california_housing_test.csv'</li><li>'sample_data/california_housing_train.csv'</li><li>'sample_data/mnist_test.csv'</li><li>'sample_data/mnist_train_small.csv'</li><li>'sample_data/README.md'</li><li>'SRR8599150_S1_L001_R1_001.fastq.gz'</li><li>'SRR8599150_S1_L001_R2_001.fastq.gz'</li><li>'t2g.txt'</li><li>'transcripts.txt'</li></ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Read matrix into R</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="nf">read_count_output</span><span class="p">(</span><span class="s">&quot;counts_unfiltered&quot;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;cells_x_genes&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h2 id="basic-qc">Basic QC<a class="headerlink" href="#basic-qc" title="Permanent link">&para;</a></h2>
<h3 id="filter-empty-droplets">Filter empty droplets<a class="headerlink" href="#filter-empty-droplets" title="Permanent link">&para;</a></h3>
<p>Most barcodes in the matrix correspond to empty droplets. A common way to determine which barcodes are empty droplets and which are real cells is to plot the rank of total UMI counts of each barcode against the total UMI count itself, which is commonly called knee plot. The inflection point in that plot, signifying a change in state, is used as a cutoff for total UMI counts; barcodes below that cutoff are deemed empty droplets and removed. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>55421</li><li>96775</li></ol>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tot_counts</span> <span class="o">&lt;-</span> <span class="n">Matrix</span><span class="o">::</span><span class="nf">colSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">tot_counts</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00    1.00    1.00   25.74    3.00 2753.00
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">bc_rank</span> <span class="o">&lt;-</span> <span class="nf">barcodeRanks</span><span class="p">(</span><span class="n">res_mat</span><span class="p">,</span> <span class="n">lower</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<h3 id="examine-the-knee-plot">Examine the knee plot<a class="headerlink" href="#examine-the-knee-plot" title="Permanent link">&para;</a></h3>
<p>The "knee plot" was introduced in the Drop-seq paper: 
- Macosko et al., <a href="https://www.cell.com/fulltext/S0092-8674(15)00549-8">Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets</a>, 2015. DOI:10.1016/j.cell.2015.05.002</p>
<p>In this plot cells are ordered by the number of UMI counts associated to them (shown on the <em>x</em>-axis), and the fraction of droplets with at least that number of cells is shown on the <em>y</em>-axis:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#&#39; Knee plot for filtering empty droplets</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; Visualizes the inflection point to filter empty droplets. This function plots </span>
<span class="c1">#&#39; different datasets with a different color. Facets can be added after calling</span>
<span class="c1">#&#39; this function with `facet_*` functions. Will be added to the next release</span>
<span class="c1">#&#39; version of BUSpaRse.</span>
<span class="c1">#&#39; </span>
<span class="c1">#&#39; @param bc_rank A `DataFrame` output from `DropletUtil::barcodeRanks`.</span>
<span class="c1">#&#39; @return A ggplot2 object.</span>
<span class="n">knee_plot</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span> <span class="p">{</span>
  <span class="n">knee_plt</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">rank</span> <span class="o">=</span> <span class="n">bc_rank</span><span class="p">[[</span><span class="s">&quot;rank&quot;</span><span class="p">]],</span>
                     <span class="n">total</span> <span class="o">=</span> <span class="n">bc_rank</span><span class="p">[[</span><span class="s">&quot;total&quot;</span><span class="p">]])</span> <span class="o">%&gt;%</span> 
    <span class="nf">distinct</span><span class="p">()</span> <span class="o">%&gt;%</span> 
    <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">total</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span>
  <span class="n">annot</span> <span class="o">&lt;-</span> <span class="nf">tibble</span><span class="p">(</span><span class="n">inflection</span> <span class="o">=</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)[[</span><span class="s">&quot;inflection&quot;</span><span class="p">]],</span>
                  <span class="n">rank_cutoff</span> <span class="o">=</span> <span class="nf">max</span><span class="p">(</span><span class="n">bc_rank</span><span class="o">$</span><span class="n">rank</span><span class="p">[</span><span class="n">bc_rank</span><span class="o">$</span><span class="n">total</span> <span class="o">&gt;</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)[[</span><span class="s">&quot;inflection&quot;</span><span class="p">]]]))</span>
  <span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">knee_plt</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">total</span><span class="p">,</span> <span class="n">rank</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_line</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">geom_hline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="n">rank_cutoff</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_vline</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="n">inflection</span><span class="p">),</span> <span class="n">data</span> <span class="o">=</span> <span class="n">annot</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">scale_x_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">scale_y_log10</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">annotation_logticks</span><span class="p">()</span> <span class="o">+</span>
    <span class="nf">labs</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Rank&quot;</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Total UMIs&quot;</span><span class="p">)</span>
  <span class="nf">return</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">knee_plot</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_27_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[,</span> <span class="n">tot_counts</span> <span class="o">&gt;</span> <span class="nf">metadata</span><span class="p">(</span><span class="n">bc_rank</span><span class="p">)</span><span class="o">$</span><span class="n">inflection</span><span class="p">]</span>
<span class="n">res_mat</span> <span class="o">&lt;-</span> <span class="n">res_mat</span><span class="p">[</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">,]</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>21574</li><li>3655</li></ol>

<h3 id="visualizing-count-distributions">Visualizing count distributions<a class="headerlink" href="#visualizing-count-distributions" title="Permanent link">&para;</a></h3>
<p>Percentage of transcripts from mitochondrially encoded genes</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tr2g</span> <span class="o">&lt;-</span> <span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;t2g.txt&quot;</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;transcript&quot;</span><span class="p">,</span> <span class="s">&quot;gene&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">))</span> <span class="o">%&gt;%</span>
  <span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">transcript</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="nf">distinct</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[36m──[39m [1m[1mColumn specification[1m[22m [36m────────────────────────────────────────────────────────[39m
cols(
  transcript = [31mcol_character()[39m,
  gene = [31mcol_character()[39m,
  gene_symbol = [31mcol_character()[39m
)
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Convert from Ensembl gene ID to gene symbol</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">tr2g</span><span class="o">$</span><span class="n">gene_symbol</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">res_mat</span><span class="p">),</span> <span class="n">tr2g</span><span class="o">$</span><span class="n">gene</span><span class="p">)]</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">res_mat</span><span class="p">,</span> <span class="n">min.cells</span> <span class="o">=</span> <span class="m">3</span><span class="p">,</span> <span class="n">min.features</span> <span class="o">=</span> <span class="m">200</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Warning message:
“Non-unique features (rownames) present in the input matrix, making unique”
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span><span class="p">[[</span><span class="s">&quot;percent.mt&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">PercentageFeatureSet</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s">&quot;^mt-&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Visualize QC metrics as a violin plot</span>
<span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">12</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">VlnPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;nFeature_RNA&quot;</span><span class="p">,</span> <span class="s">&quot;nCount_RNA&quot;</span><span class="p">,</span> <span class="s">&quot;percent.mt&quot;</span><span class="p">),</span> <span class="n">ncol</span> <span class="o">=</span> <span class="m">3</span><span class="p">,</span> <span class="n">pt.size</span> <span class="o">=</span> <span class="m">0.1</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_35_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="o">=</span><span class="m">9</span><span class="p">,</span> <span class="n">repr.plot.height</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">seu</span><span class="o">@</span><span class="n">meta.data</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">nCount_RNA</span><span class="p">,</span> <span class="n">nFeature_RNA</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_hex</span><span class="p">(</span><span class="n">bins</span> <span class="o">=</span> <span class="m">100</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_fill_scico</span><span class="p">(</span><span class="n">palette</span> <span class="o">=</span> <span class="s">&quot;devon&quot;</span><span class="p">,</span> <span class="n">direction</span> <span class="o">=</span> <span class="m">-1</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="m">0.9</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_x_log10</span><span class="p">(</span><span class="n">breaks</span> <span class="o">=</span> <span class="nf">breaks_log</span><span class="p">(</span><span class="m">12</span><span class="p">))</span> <span class="o">+</span> 
  <span class="nf">scale_y_log10</span><span class="p">(</span><span class="n">breaks</span> <span class="o">=</span> <span class="nf">breaks_log</span><span class="p">(</span><span class="m">12</span><span class="p">))</span> <span class="o">+</span> <span class="nf">annotation_logticks</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Total UMI counts&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Number of genes detected&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">panel.grid.minor</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">())</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Warning</span> <span class="n">message</span><span class="o">:</span>
<span class="n">“Computation</span> <span class="n">failed</span> <span class="k">in</span> <span class="n n-Quoted">`stat_binhex()`</span><span class="o">:</span>
  <span class="n">Package</span> <span class="n n-Quoted">`hexbin`</span> <span class="n">required</span> <span class="k">for</span> <span class="n n-Quoted">`stat_binhex`</span><span class="p">.</span>
  <span class="n">Please</span> <span class="k">install</span> <span class="k">and</span> <span class="n">try</span> <span class="n">again</span><span class="p">.</span><span class="n">”</span>
</code></pre></div>

<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_36_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">seu</span><span class="o">@</span><span class="n">meta.data</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">nCount_RNA</span><span class="p">,</span> <span class="n">percent.mt</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_pointdensity</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">scale_color_scico</span><span class="p">(</span><span class="n">palette</span> <span class="o">=</span> <span class="s">&quot;devon&quot;</span><span class="p">,</span> <span class="n">direction</span> <span class="o">=</span> <span class="m">-1</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="m">0.9</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;Total UMI counts&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Percentage mitochondrial&quot;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_37_0.png" /></p>
<p>The color shows density of points, as the density is not apparent when many points are stacked on top of each other. Cells with high percentage of mitochondrially encoded transcripts are often removed in QC, as those are likely to be low quality cells. If a cell is lysed in sample preparation, transcripts in the mitochondria are less likely to be lost than transcripts in the cytoplasm due to the double membrane of the mitochondria, so cells that lysed tend to have a higher percentage of mitochondrially encoded transcripts.</p>
<p>We filter cells with more than 3% mitochondrial content based on the plot above.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">subset</span> <span class="o">=</span> <span class="n">percent.mt</span> <span class="o">&lt;</span> <span class="m">3</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">NormalizeData</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">ScaleData</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Centering and scaling data matrix
</code></pre></div>

<h2 id="analysis">Analysis<a class="headerlink" href="#analysis" title="Permanent link">&para;</a></h2>
<h3 id="identify-highly-variable-genes">Identify highly variable genes<a class="headerlink" href="#identify-highly-variable-genes" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">nfeatures</span> <span class="o">=</span> <span class="m">3000</span><span class="p">)</span>
<span class="n">top10</span> <span class="o">&lt;-</span> <span class="nf">head</span><span class="p">(</span><span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">seu</span><span class="p">),</span> <span class="m">10</span><span class="p">)</span>
<span class="n">plot1</span> <span class="o">&lt;-</span> <span class="nf">VariableFeaturePlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">log</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">LabelPoints</span><span class="p">(</span><span class="n">plot</span> <span class="o">=</span> <span class="n">plot1</span><span class="p">,</span> <span class="n">points</span> <span class="o">=</span> <span class="n">top10</span><span class="p">,</span> <span class="n">repel</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">When</span> <span class="nv">using</span> <span class="nv">repel</span>, <span class="nv">set</span> <span class="nv">xnudge</span> <span class="nv">and</span> <span class="nv">ynudge</span> <span class="nv">to</span> <span class="mi">0</span> <span class="k">for</span> <span class="nv">optimal</span> <span class="nv">results</span>
</code></pre></div>

<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_44_1.png" /></p>
<h3 id="pca">PCA<a class="headerlink" href="#pca" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">RunPCA</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">npcs</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span> <span class="c1"># uses HVG by default</span>
<span class="nf">ElbowPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">ndims</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_46_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">PCAPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_47_0.png" /></p>
<h3 id="clustering-and-visualization">Clustering and visualization<a class="headerlink" href="#clustering-and-visualization" title="Permanent link">&para;</a></h3>
<p>There are many algorithms for clustering cells, and while they have been compared in detail in various benchmarks (see e.g., <a href="https://f1000research.com/articles/7-1141/v2">Duo et al. 2018</a>), there is no univerally agreed upon method. Here we demonstrate clustering using <a href="https://en.wikipedia.org/wiki/Louvain_modularity">Louvain clustering</a>, which is a popular method for clustering single-cell RNA-seq data. The method was published in </p>
<ul>
<li>Blondel, Vincent D; Guillaume, Jean-Loup; Lambiotte, Renaud; Lefebvre, Etienne (9 October 2008). "Fast unfolding of communities in large networks". Journal of Statistical Mechanics: Theory and Experiment. 2008 (10): P10008.</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)</span>
<span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">FindClusters</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">Computing</span> <span class="nv">nearest</span> <span class="nv">neighbor</span> <span class="nv">graph</span>

<span class="nv">Computing</span> <span class="nv">SNN</span>



<span class="nv">Modularity</span> <span class="nv">Optimizer</span> <span class="nv">version</span> <span class="mi">1</span>.<span class="mi">3</span>.<span class="mi">0</span> <span class="nv">by</span> <span class="nv">Ludo</span> <span class="nv">Waltman</span> <span class="nv">and</span> <span class="nv">Nees</span> <span class="nv">Jan</span> <span class="nv">van</span> <span class="nv">Eck</span>

<span class="nv">Number</span> <span class="nv">of</span> <span class="nv">nodes</span>: <span class="mi">3507</span>
<span class="nv">Number</span> <span class="nv">of</span> <span class="nv">edges</span>: <span class="mi">112378</span>

<span class="nv">Running</span> <span class="nv">Louvain</span> <span class="nv">algorithm</span>...
<span class="nv">Maximum</span> <span class="nv">modularity</span> <span class="nv">in</span> <span class="mi">10</span> <span class="k">random</span> <span class="nv">starts</span>: <span class="mi">0</span>.<span class="mi">7206</span>
<span class="nv">Number</span> <span class="nv">of</span> <span class="nv">communities</span>: <span class="mi">8</span>
<span class="nv">Elapsed</span> <span class="nv">time</span>: <span class="mi">0</span> <span class="nv">seconds</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">PCAPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_50_0.png" /></p>
<h3 id="tsne">tSNE<a class="headerlink" href="#tsne" title="Permanent link">&para;</a></h3>
<p><a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding">t-SNE</a> is a non-linear dimensionality reduction technique described in:</p>
<ul>
<li>Maaten, Laurens van der, and Geoffrey Hinton. "Visualizing data using t-SNE." Journal of machine learning research 9.Nov (2008): 2579-2605.</li>
</ul>
<p>Here it is applied to the 10-dimensional PCA projection of the cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">RunTSNE</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)</span>
<span class="nf">TSNEPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_52_0.png" /></p>
<h3 id="umap">UMAP<a class="headerlink" href="#umap" title="Permanent link">&para;</a></h3>
<p>UMAP (UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction) is a manifold learning technique that can also be used to visualize cells. It was published in:</p>
<ul>
<li>McInnes, Leland, John Healy, and James Melville. "Umap: Uniform manifold approximation and projection for dimension reduction." arXiv preprint arXiv:1802.03426 (2018).</li>
</ul>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">seu</span> <span class="o">&lt;-</span> <span class="nf">RunUMAP</span><span class="p">(</span><span class="n">seu</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">Warning</span> <span class="nv">message</span>:
“<span class="nv">The</span> <span class="nv">default</span> <span class="nv">method</span> <span class="k">for</span> <span class="nv">RunUMAP</span> <span class="nv">has</span> <span class="nv">changed</span> <span class="nv">from</span> <span class="nv">calling</span> <span class="nv">Python</span> <span class="nv">UMAP</span> <span class="nv">via</span> <span class="nv">reticulate</span> <span class="nv">to</span> <span class="nv">the</span> <span class="nv">R</span><span class="o">-</span><span class="nv">native</span> <span class="nv">UWOT</span> <span class="nv">using</span> <span class="nv">the</span> <span class="nv">cosine</span> <span class="nv">metric</span>
<span class="nv">To</span> <span class="nv">use</span> <span class="nv">Python</span> <span class="nv">UMAP</span> <span class="nv">via</span> <span class="nv">reticulate</span>, <span class="nv">set</span> <span class="nv">umap</span>.<span class="nv">method</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">umap-learn</span><span class="s1">&#39;</span> <span class="nv">and</span> <span class="nv">metric</span> <span class="nv">to</span> <span class="s1">&#39;</span><span class="s">correlation</span><span class="s1">&#39;</span>
<span class="nv">This</span> <span class="nv">message</span> <span class="nv">will</span> <span class="nv">be</span> <span class="nv">shown</span> <span class="nv">once</span> <span class="nv">per</span> <span class="nv">session</span>”
</code></pre></div>

<p><img alt="png" src="kb_intro_2_R_files/kb_intro_2_R_54_1.png" /></p>
<h1 id="discussion">Discussion<a class="headerlink" href="#discussion" title="Permanent link">&para;</a></h1>
<p>This notebook has demonstrated visualization of cells following pre-processing of single-cell RNA-seq data.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="nf">Sys.time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start_time</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>Time difference of 38.41409 mins
</code></pre></div>

<p>Installing packages took about 26 minutes, which is a drawback of Rcpp. The QC and  analysis post-installation takes about 10 minutes from reads to results. This includes downloading the data, filtering, clustering and visualization.</p>
<p><strong>Feedback</strong>: please report any issues, or submit pull requests for improvements, in the <a href="https://github.com/pachterlab/kallistobustools/">Github repository where this notebook is located</a>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>
</code></pre></div>
</td></tr></table>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../python/kb_intro_2_python.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Python" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Python
            </div>
          </div>
        </a>
      
      
        
        <a href="../../kb_building_atlas/python/kb_analysis_0_python.html" class="md-footer__link md-footer__link--next" aria-label="Next: Python" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Python
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../assets/javascripts/workers/search.477d984a.min.js", "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.82b56eb2.min.js"></script>
      
    
  </body>
</html>