
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Short and simple bioinformatics tutorials">
      
      
      
        <meta name="author" content="A. Sina Booeshaghi">
      
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.2, mkdocs-material-7.3.0">
    
    
      
        <title>Python - kallisto | bustools</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.8b42a75e.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.3f5d1f46.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../../../css/custom.css">
    
    
      
  



  
  



  <script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-142178789-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
  <script async src="https://www.google-analytics.com/analytics.js"></script>


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    <script>function __prefix(e){return new URL("../../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#analysis-of-single-cell-rna-seq-data-building-and-annotating-an-atlas" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="kallisto | bustools" class="md-header__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            kallisto | bustools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Python
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="kallisto | bustools" class="md-nav__button md-logo" aria-label="kallisto | bustools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    kallisto | bustools
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/kallistobustools/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    pachterlab/kallistobustools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../applications/" class="md-nav__link">
        Applications
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_3">
          Using kb
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Using kb" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Using kb
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_usage/kb_usage/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_usage/kb_ref/" class="md-nav__link">
        Building a reference
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../kb_usage/kb_count/" class="md-nav__link">
        Quantifying a dataset
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Tutorials
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_1" type="checkbox" id="__nav_4_1" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_1">
          Introduction to single-cell RNA-seq
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Introduction to single-cell RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Introduction to single-cell RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../scRNA-seq_intro/python/scRNA-seq_intro/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_2" type="checkbox" id="__nav_4_2" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_2">
          Preprocessing & Quality Control
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Preprocessing & Quality Control" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing & Quality Control
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_quality_control/python/kb_intro_1_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_quality_control/R/kb_intro_1_R/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_3" type="checkbox" id="__nav_4_3" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_3">
          Getting started with analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Getting started with analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_3">
          <span class="md-nav__icon md-icon"></span>
          Getting started with analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_getting_started/python/kb_intro_2_python/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_getting_started/R/kb_intro_2_R/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4_4">
          Building & annotating an atlas
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Building & annotating an atlas" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          Building & annotating an atlas
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Python
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Python
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-python-packages" class="md-nav__link">
    Install python packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb-python" class="md-nav__link">
    Install kb-python
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-an-index" class="md-nav__link">
    Download an index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-for-library-saturation" class="md-nav__link">
    Test for library saturation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filter" class="md-nav__link">
    Filter
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#normalize-counts" class="md-nav__link">
    Normalize counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identify-highly-variable-genes" class="md-nav__link">
    Identify highly-variable genes.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scaling-the-data" class="md-nav__link">
    Scaling the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#principal-component-analysis" class="md-nav__link">
    Principal component analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute-the-neighborhood-graph" class="md-nav__link">
    Compute the neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#embed-the-neighborhood-graph" class="md-nav__link">
    Embed the neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umap" class="md-nav__link">
    UMAP
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cluster-the-neighborhood-graph" class="md-nav__link">
    Cluster the neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering" class="md-nav__link">
    Clustering
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-marker-genes" class="md-nav__link">
    Find marker genes
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../R/kb_analysis_0_R/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_5">
          Pseudotime analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Pseudotime analysis" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_pseudotime/R/slingshot/kb_slingshot/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_6">
          Single-nuclei RNA-seq
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Single-nuclei RNA-seq" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          Single-nuclei RNA-seq
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_nucleus/python/kb_single_nucleus/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_7">
          RNA Velocity
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="RNA Velocity" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          RNA Velocity
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_velocity/python/kb_velocity/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_8">
          Feature barcode & multimodal data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Feature barcode & multimodal data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          Feature barcode & multimodal data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_kite/python/kb_kite/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../download_data/data_download/" class="md-nav__link">
        Find & download data
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_10">
          Build custom index
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Build custom index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          Build custom index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_custom_index/python/kb_transcriptome_index/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_11" type="checkbox" id="__nav_4_11" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_11">
          Build RNA velocity index
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Build RNA velocity index" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_11">
          <span class="md-nav__icon md-icon"></span>
          Build RNA velocity index
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_velocity_index/python/kb_velocity_index/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_12" type="checkbox" id="__nav_4_12" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_12">
          Aggregate multiple matrices
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Aggregate multiple matrices" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_12">
          <span class="md-nav__icon md-icon"></span>
          Aggregate multiple matrices
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_aggregate/python/kb_aggregating_count_matrices/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_13" type="checkbox" id="__nav_4_13" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_13">
          Process multiple FASTQs
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Process multiple FASTQs" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_13">
          <span class="md-nav__icon md-icon"></span>
          Process multiple FASTQs
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_multiple_fastqs/python/kb_multiple_files/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_14" type="checkbox" id="__nav_4_14" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_14">
          Processing multi-species data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Processing multi-species data" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_14">
          <span class="md-nav__icon md-icon"></span>
          Processing multi-species data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_species_mixing/python/kb_species_mixing/" class="md-nav__link">
        Python
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_species_mixing/R/kb_mixed_species_10x_v2/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_15" type="checkbox" id="__nav_4_15" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_15">
          Pseudotime analysis (Monocle2)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Pseudotime analysis (Monocle2)" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_15">
          <span class="md-nav__icon md-icon"></span>
          Pseudotime analysis (Monocle2)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../kb_pseudotime/R/monocle2/kb_monocle2/" class="md-nav__link">
        R
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-python-packages" class="md-nav__link">
    Install python packages
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-kb-python" class="md-nav__link">
    Install kb-python
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-the-data" class="md-nav__link">
    Download the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#download-an-index" class="md-nav__link">
    Download an index
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pseudoalignment-and-counting" class="md-nav__link">
    Pseudoalignment and counting
  </a>
  
    <nav class="md-nav" aria-label="Pseudoalignment and counting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-kallisto-and-bustools" class="md-nav__link">
    Run kallisto and bustools
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-qc" class="md-nav__link">
    Basic QC
  </a>
  
    <nav class="md-nav" aria-label="Basic QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-for-library-saturation" class="md-nav__link">
    Test for library saturation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examine-the-knee-plot" class="md-nav__link">
    Examine the knee plot
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analysis" class="md-nav__link">
    Analysis
  </a>
  
    <nav class="md-nav" aria-label="Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#filter" class="md-nav__link">
    Filter
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#normalize-counts" class="md-nav__link">
    Normalize counts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identify-highly-variable-genes" class="md-nav__link">
    Identify highly-variable genes.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scaling-the-data" class="md-nav__link">
    Scaling the data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#principal-component-analysis" class="md-nav__link">
    Principal component analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute-the-neighborhood-graph" class="md-nav__link">
    Compute the neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#embed-the-neighborhood-graph" class="md-nav__link">
    Embed the neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umap" class="md-nav__link">
    UMAP
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cluster-the-neighborhood-graph" class="md-nav__link">
    Cluster the neighborhood graph
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering" class="md-nav__link">
    Clustering
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find-marker-genes" class="md-nav__link">
    Find marker genes
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/kallistobustools/tree/master/tutorials/docs/tutorials/kb_building_atlas/python/kb_analysis_0_python.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p><a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/docs/tutorials/kb_building_atlas/python/kb_analysis_0_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h1 id="analysis-of-single-cell-rna-seq-data-building-and-annotating-an-atlas">Analysis of single-cell RNA-seq data: building and annotating an atlas<a class="headerlink" href="#analysis-of-single-cell-rna-seq-data-building-and-annotating-an-atlas" title="Permanent link">&para;</a></h1>
<p>This Python notebook pre-processes the <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3">pbmc_1k v3 dataset</a> from 10X Genomics with kallisto and bustools using <code>kb</code>, and then performs an analysis of the cell types and their marker genes.</p>
<p>The notebook was written by A. Sina Booeshaghi and Lior Pachter and is based on three noteboks:
- The kallisto | bustools <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_1_minute_intro.ipynb#scrollTo=wtwMjIjjCMcD">Introduction to single-cell RNA-seq I</a> notebook.
- The kallisto | bustools <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_standard.ipynb#scrollTo=ijU_u6uj3Sio">Introduction to single-cell RNA-seq II</a> notebook.
- The Scanpy <a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">Preprocessing and clustering 3k PBMCs" notebook</a>.</p>
<p>If you use the methods in this notebook for your analysis please cite the following publications which describe the tools used in the notebook:</p>
<ul>
<li>Melsted, P., Booeshaghi, A.S. et al. Modular and efficient pre-processing of single-cell RNA-seq. bioRxiv (2019). doi:10.1101/673285</li>
<li>Wolf, F. A., Angere, P. and Theis, F.J. SCANPY: large-scale single-cell gene expression data analysis. Genome Biology (2018). doi:10.1186/s13059-017-1382-0</li>
</ul>
<p>An R notebook implementing the same analysis is available <a href="https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/notebooks/kb_analysis_0_R.ipynb">here</a>. See the <a href="https://www.kallistobus.tools/tutorials">kallistobus.tools tutorials</a> site for additional notebooks demonstrating other analyses.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># This is  used to time the running of the notebook</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<h3 id="install-python-packages">Install python packages<a class="headerlink" href="#install-python-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="o">%%</span><span class="n">time</span>
<span class="c1"># These packages are pre-installed on Google Colab, but are included here to simplify running this notebook locally</span>
<span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="o">--</span><span class="n">quiet</span> <span class="n">matplotlib</span> <span class="n">scikit</span><span class="o">-</span><span class="n">learn</span> <span class="n">numpy</span> <span class="n">scipy</span>
<span class="err">!</span><span class="n">pip3</span> <span class="n">install</span> <span class="o">--</span><span class="n">quiet</span> <span class="n">leidenalg</span>
<span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="o">--</span><span class="n">quiet</span> <span class="n">louvain</span> <span class="n">scanpy</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">1</span>.<span class="mi">4</span><span class="nv">MB</span> <span class="mi">4</span>.<span class="mi">2</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">3</span>.<span class="mi">2</span><span class="nv">MB</span> <span class="mi">34</span>.<span class="mi">9</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">2</span>.<span class="mi">2</span><span class="nv">MB</span> <span class="mi">4</span>.<span class="mi">3</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">2</span>.<span class="mi">0</span><span class="nv">MB</span> <span class="mi">36</span>.<span class="mi">8</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">81</span><span class="nv">kB</span> <span class="mi">7</span>.<span class="mi">9</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">133</span><span class="nv">kB</span> <span class="mi">43</span>.<span class="mi">0</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">71</span><span class="nv">kB</span> <span class="mi">7</span>.<span class="mi">5</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[<span class="nv">K</span>     <span class="o">|</span>â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ<span class="o">|</span> <span class="mi">1</span>.<span class="mi">1</span><span class="nv">MB</span> <span class="mi">40</span>.<span class="mi">5</span><span class="nv">MB</span><span class="o">/</span><span class="nv">s</span> 
[?<span class="mi">25</span><span class="nv">h</span>  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">sinfo</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">umap</span><span class="o">-</span><span class="nv">learn</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
  <span class="nv">Building</span> <span class="nv">wheel</span> <span class="k">for</span> <span class="nv">pynndescent</span> <span class="ss">(</span><span class="nv">setup</span>.<span class="nv">py</span><span class="ss">)</span> ... [?<span class="mi">25</span><span class="nv">l</span>[?<span class="mi">25</span><span class="nv">hdone</span>
<span class="nv">CPU</span> <span class="nv">times</span>: <span class="nv">user</span> <span class="mi">164</span> <span class="nv">ms</span>, <span class="nv">sys</span>: <span class="mi">38</span>.<span class="mi">8</span> <span class="nv">ms</span>, <span class="nv">total</span>: <span class="mi">203</span> <span class="nv">ms</span>
<span class="nv">Wall</span> <span class="nv">time</span>: <span class="mi">17</span>.<span class="mi">1</span> <span class="nv">s</span>
</code></pre></div>

<h3 id="install-kb-python">Install kb-python<a class="headerlink" href="#install-kb-python" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="o">%%</span><span class="n">time</span>
<span class="c1"># install kb</span>
<span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="o">--</span><span class="n">quiet</span> <span class="n">kb</span><span class="o">-</span><span class="n">python</span> 
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mf">7.4</span><span class="n">MB</span> <span class="mf">4.5</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mi">51</span><span class="n">kB</span> <span class="mf">3.5</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mf">51.7</span><span class="n">MB</span> <span class="mi">48</span><span class="n">kB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mf">20.6</span><span class="n">MB</span> <span class="mf">1.5</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mi">112</span><span class="n">kB</span> <span class="mf">52.1</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mf">9.9</span><span class="n">MB</span> <span class="mf">47.1</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="n">K</span>     <span class="o">|</span><span class="err">â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ</span><span class="o">|</span> <span class="mi">276</span><span class="n">kB</span> <span class="mf">40.5</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span> 
<span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">h</span>  <span class="n">Installing</span> <span class="n">build</span> <span class="n">dependencies</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
  <span class="n">Getting</span> <span class="n">requirements</span> <span class="n">to</span> <span class="n">build</span> <span class="n">wheel</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
    <span class="n">Preparing</span> <span class="n">wheel</span> <span class="n">metadata</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
  <span class="n">Building</span> <span class="n">wheel</span> <span class="k">for</span> <span class="n">pyseq</span><span class="o">-</span><span class="n">align</span> <span class="p">(</span><span class="n">PEP</span> <span class="mi">517</span><span class="p">)</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
  <span class="n">Building</span> <span class="n">wheel</span> <span class="k">for</span> <span class="n">kb</span><span class="o">-</span><span class="n">python</span> <span class="p">(</span><span class="n">setup</span><span class="o">.</span><span class="n">py</span><span class="p">)</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
  <span class="n">Building</span> <span class="n">wheel</span> <span class="k">for</span> <span class="n">loompy</span> <span class="p">(</span><span class="n">setup</span><span class="o">.</span><span class="n">py</span><span class="p">)</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
  <span class="n">Building</span> <span class="n">wheel</span> <span class="k">for</span> <span class="n">ngs</span><span class="o">-</span><span class="n">tools</span> <span class="p">(</span><span class="n">setup</span><span class="o">.</span><span class="n">py</span><span class="p">)</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
  <span class="n">Building</span> <span class="n">wheel</span> <span class="k">for</span> <span class="n">numpy</span><span class="o">-</span><span class="n">groupies</span> <span class="p">(</span><span class="n">setup</span><span class="o">.</span><span class="n">py</span><span class="p">)</span> <span class="o">...</span> <span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">l</span><span class="err"></span><span class="p">[</span><span class="err">?</span><span class="mi">25</span><span class="n">hdone</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">mERROR</span><span class="p">:</span> <span class="n">ngs</span><span class="o">-</span><span class="n">tools</span> <span class="mf">1.4</span><span class="o">.</span><span class="mi">0</span> <span class="n">has</span> <span class="n">requirement</span> <span class="n">numba</span><span class="o">&gt;=</span><span class="mf">0.53</span><span class="o">.</span><span class="mi">1</span><span class="p">,</span> <span class="n">but</span> <span class="n">you</span><span class="s1">&#39;ll have numba 0.51.2 which is incompatible.[0m</span>
<span class="err"></span><span class="p">[</span><span class="mi">31</span><span class="n">mERROR</span><span class="p">:</span> <span class="n">ngs</span><span class="o">-</span><span class="n">tools</span> <span class="mf">1.4</span><span class="o">.</span><span class="mi">0</span> <span class="n">has</span> <span class="n">requirement</span> <span class="n">tqdm</span><span class="o">&gt;=</span><span class="mf">4.50</span><span class="o">.</span><span class="mi">0</span><span class="p">,</span> <span class="n">but</span> <span class="n">you</span><span class="s1">&#39;ll have tqdm 4.41.1 which is incompatible.[0m</span>
<span class="n">CPU</span> <span class="n">times</span><span class="p">:</span> <span class="n">user</span> <span class="mi">570</span> <span class="n">ms</span><span class="p">,</span> <span class="n">sys</span><span class="p">:</span> <span class="mf">95.8</span> <span class="n">ms</span><span class="p">,</span> <span class="n">total</span><span class="p">:</span> <span class="mi">666</span> <span class="n">ms</span>
<span class="n">Wall</span> <span class="n">time</span><span class="p">:</span> <span class="mi">1</span><span class="nb">min</span> <span class="mi">1</span><span class="n">s</span>
</code></pre></div>

<h3 id="download-the-data">Download the data<a class="headerlink" href="#download-the-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="o">%%</span><span class="n">time</span>
<span class="c1"># Download the data from the 10x website</span>
<span class="err">!</span><span class="n">wget</span> <span class="o">-</span><span class="n">q</span> <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">cf</span><span class="mf">.10</span><span class="n">xgenomics</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">samples</span><span class="o">/</span><span class="n">cell</span><span class="o">-</span><span class="n">exp</span><span class="o">/</span><span class="mf">3.0.0</span><span class="o">/</span><span class="n">pbmc_1k_v3</span><span class="o">/</span><span class="n">pbmc_1k_v3_fastqs</span><span class="o">.</span><span class="n">tar</span>

<span class="c1"># unpack the downloaded files</span>
<span class="err">!</span><span class="n">tar</span> <span class="o">-</span><span class="n">xf</span> <span class="n">pbmc_1k_v3_fastqs</span><span class="o">.</span><span class="n">tar</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>CPU times: user 1.94 s, sys: 318 ms, total: 2.26 s
Wall time: 4min 59s
</code></pre></div>

<h3 id="download-an-index">Download an index<a class="headerlink" href="#download-an-index" title="Permanent link">&para;</a></h3>
<p>This data consists of peripheral blood mononuclear cells from a human, so we download the human index.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="err">!</span><span class="n">kb</span> <span class="n">ref</span> <span class="o">-</span><span class="n">d</span> <span class="n">human</span> <span class="o">-</span><span class="n">i</span> <span class="n">index</span><span class="o">.</span><span class="n">idx</span> <span class="o">-</span><span class="n">g</span> <span class="n">t2g</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">f1</span> <span class="n">transcriptome</span><span class="o">.</span><span class="n">fasta</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">[</span><span class="n">2021-07-07 18:47:56,003</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">download</span><span class="o">]</span><span class="w"> </span><span class="n">Downloading</span><span class="w"> </span><span class="n">files</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">human</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="nl">https</span><span class="p">:</span><span class="o">//</span><span class="n">caltech</span><span class="p">.</span><span class="n">box</span><span class="p">.</span><span class="n">com</span><span class="o">/</span><span class="n">shared</span><span class="o">/</span><span class="k">static</span><span class="o">/</span><span class="n">v1nm7lpnqz5syh8dyzdk2zs8bglncfib</span><span class="p">.</span><span class="n">gz</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">tmp</span><span class="o">/</span><span class="n">v1nm7lpnqz5syh8dyzdk2zs8bglncfib</span><span class="p">.</span><span class="n">gz</span><span class="w"></span>
<span class="mi">100</span><span class="o">%</span><span class="w"> </span><span class="mf">2.23</span><span class="n">G</span><span class="o">/</span><span class="mf">2.23</span><span class="n">G</span><span class="w"> </span><span class="o">[</span><span class="n">02:48&lt;00:00, 14.2MB/s</span><span class="o">]</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 18:50:44,475</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">download</span><span class="o">]</span><span class="w"> </span><span class="n">Extracting</span><span class="w"> </span><span class="n">files</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">tmp</span><span class="o">/</span><span class="n">v1nm7lpnqz5syh8dyzdk2zs8bglncfib</span><span class="p">.</span><span class="n">gz</span><span class="w"></span>
</code></pre></div>

<h2 id="pseudoalignment-and-counting">Pseudoalignment and counting<a class="headerlink" href="#pseudoalignment-and-counting" title="Permanent link">&para;</a></h2>
<h3 id="run-kallisto-and-bustools">Run kallisto and bustools<a class="headerlink" href="#run-kallisto-and-bustools" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="o">%%</span><span class="n">time</span>
<span class="err">!</span><span class="n">kb</span> <span class="n">count</span> <span class="o">--</span><span class="n">h5ad</span> <span class="o">-</span><span class="n">i</span> <span class="n">index</span><span class="o">.</span><span class="n">idx</span> <span class="o">-</span><span class="n">g</span> <span class="n">t2g</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">x</span> <span class="mi">10</span><span class="n">xv3</span> <span class="o">-</span><span class="n">o</span> <span class="n">output</span> <span class="o">--</span><span class="nb">filter</span> <span class="n">bustools</span> <span class="o">-</span><span class="n">t</span> <span class="mi">2</span> \
<span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> \
<span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">[</span><span class="n">2021-07-07 18:51:36,108</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="k">Using</span><span class="w"> </span><span class="k">index</span><span class="w"> </span><span class="k">index</span><span class="p">.</span><span class="n">idx</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">generate</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="w"> </span><span class="k">from</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 18:51:36,109</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w">         </span><span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L001_R1_001</span><span class="p">.</span><span class="n">fastq</span><span class="p">.</span><span class="n">gz</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 18:51:36,109</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w">         </span><span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L001_R2_001</span><span class="p">.</span><span class="n">fastq</span><span class="p">.</span><span class="n">gz</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 18:51:36,109</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w">         </span><span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L002_R1_001</span><span class="p">.</span><span class="n">fastq</span><span class="p">.</span><span class="n">gz</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 18:51:36,109</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w">         </span><span class="n">pbmc_1k_v3_fastqs</span><span class="o">/</span><span class="n">pbmc_1k_v3_S1_L002_R2_001</span><span class="p">.</span><span class="n">fastq</span><span class="p">.</span><span class="n">gz</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:04:53,157</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Sorting</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">s</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:05:21,205</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Whitelist</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">provided</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:05:21,205</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Copying</span><span class="w"> </span><span class="n">pre</span><span class="o">-</span><span class="n">packaged</span><span class="w"> </span><span class="mi">10</span><span class="n">XV3</span><span class="w"> </span><span class="n">whitelist</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:05:22,196</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Inspecting</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">s</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:05:40,875</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Correcting</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="n">records</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">s</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">s</span><span class="p">.</span><span class="n">c</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">with</span><span class="w"> </span><span class="n">whitelist</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="mi">10</span><span class="n">x_version3_whitelist</span><span class="p">.</span><span class="n">txt</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:01,625</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Sorting</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">s</span><span class="p">.</span><span class="n">c</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">unfiltered</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:22,834</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Generating</span><span class="w"> </span><span class="nf">count</span><span class="w"> </span><span class="n">matrix</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">counts_unfiltered</span><span class="o">/</span><span class="n">cells_x_genes</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">unfiltered</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:38,641</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Reading</span><span class="w"> </span><span class="n">matrix</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">counts_unfiltered</span><span class="o">/</span><span class="n">cells_x_genes</span><span class="p">.</span><span class="n">mtx</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:46,637</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Writing</span><span class="w"> </span><span class="n">matrix</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">h5ad</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">counts_unfiltered</span><span class="o">/</span><span class="n">adata</span><span class="p">.</span><span class="n">h5ad</span><span class="w"></span>
<span class="p">...</span><span class="w"> </span><span class="n">storing</span><span class="w"> </span><span class="s1">&#39;gene_name&#39;</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">categorical</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:47,631</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Filtering</span><span class="w"> </span><span class="k">with</span><span class="w"> </span><span class="n">bustools</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:47,631</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Generating</span><span class="w"> </span><span class="n">whitelist</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">filter_barcodes</span><span class="p">.</span><span class="n">txt</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">unfiltered</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:47,815</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Correcting</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="n">records</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">unfiltered</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">unfiltered</span><span class="p">.</span><span class="n">c</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">with</span><span class="w"> </span><span class="n">whitelist</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">filter_barcodes</span><span class="p">.</span><span class="n">txt</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:06:56,813</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Sorting</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">unfiltered</span><span class="p">.</span><span class="n">c</span><span class="p">.</span><span class="n">bus</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">filtered</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:07:13,988</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Generating</span><span class="w"> </span><span class="nf">count</span><span class="w"> </span><span class="n">matrix</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">counts_filtered</span><span class="o">/</span><span class="n">cells_x_genes</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">BUS</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="k">output</span><span class="p">.</span><span class="n">filtered</span><span class="p">.</span><span class="n">bus</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:07:27,060</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Reading</span><span class="w"> </span><span class="n">matrix</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">counts_filtered</span><span class="o">/</span><span class="n">cells_x_genes</span><span class="p">.</span><span class="n">mtx</span><span class="w"></span>
<span class="o">[</span><span class="n">2021-07-07 19:07:32,997</span><span class="o">]</span><span class="w">    </span><span class="n">INFO</span><span class="w"> </span><span class="o">[</span><span class="n">count</span><span class="o">]</span><span class="w"> </span><span class="n">Writing</span><span class="w"> </span><span class="n">matrix</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="n">h5ad</span><span class="w"> </span><span class="k">output</span><span class="o">/</span><span class="n">counts_filtered</span><span class="o">/</span><span class="n">adata</span><span class="p">.</span><span class="n">h5ad</span><span class="w"></span>
<span class="p">...</span><span class="w"> </span><span class="n">storing</span><span class="w"> </span><span class="s1">&#39;gene_name&#39;</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">categorical</span><span class="w"></span>
<span class="n">CPU</span><span class="w"> </span><span class="nl">times</span><span class="p">:</span><span class="w"> </span><span class="k">user</span><span class="w"> </span><span class="mf">6.71</span><span class="w"> </span><span class="n">s</span><span class="p">,</span><span class="w"> </span><span class="nl">sys</span><span class="p">:</span><span class="w"> </span><span class="mi">889</span><span class="w"> </span><span class="n">ms</span><span class="p">,</span><span class="w"> </span><span class="nl">total</span><span class="p">:</span><span class="w"> </span><span class="mf">7.6</span><span class="w"> </span><span class="n">s</span><span class="w"></span>
<span class="n">Wall</span><span class="w"> </span><span class="nc">time</span><span class="err">:</span><span class="w"> </span><span class="mi">16</span><span class="nf">min</span><span class="w"></span>
</code></pre></div>

<h2 id="basic-qc">Basic QC<a class="headerlink" href="#basic-qc" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">TruncatedSVD</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">3</span>             <span class="c1"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># load the unfiltered matrix</span>
<span class="n">results_file</span> <span class="o">=</span> <span class="s1">&#39;pbmc1k.h5ad&#39;</span>  <span class="c1"># the file that will store the analysis results</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;output/counts_unfiltered/adata.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

<span class="n">t2g</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;t2g.txt&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tid&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">t2g</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">t2g</span><span class="o">.</span><span class="n">gene_id</span>
<span class="n">t2g</span> <span class="o">=</span> <span class="n">t2g</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">t2g</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)]</span>

<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">t2g</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">])</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span>

<span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>  <span class="c1"># this is unnecessary if using `var_names=&#39;gene_ids&#39;` in `sc.read_10x_mtx`</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>/usr/local/lib/python3.7/dist-packages/anndata/utils.py:117: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: [&#39;SNORD116-1&#39;, &#39;SNORD116-2&#39;, &#39;SNORD116-3&#39;, &#39;SNORD116-4&#39;, &#39;SNORD116-5&#39;]
  + str(example_colliding_values)
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">259615</span> <span class="err">Ã—</span> <span class="mi">60623</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span>
</code></pre></div>

<h3 id="test-for-library-saturation">Test for library saturation<a class="headerlink" href="#test-for-library-saturation" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Create a plot showing genes detected as a function of UMI counts.</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))[:,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))[:,</span><span class="mi">0</span><span class="p">]</span>

<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;green&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.25</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;UMI Counts&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Genes Detected&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_22_0.png" /></p>
<p>This plot is very misleading, as even the small alpha can't accurately show how many points are stacked at one location (This takes about a minute to run since there are a lot of points)</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">))</span>

<span class="c1">#histogram definition</span>
<span class="n">bins</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span> <span class="mi">1500</span><span class="p">]</span> <span class="c1"># number of bins</span>

<span class="c1"># histogram the data</span>
<span class="n">hh</span><span class="p">,</span> <span class="n">locx</span><span class="p">,</span> <span class="n">locy</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram2d</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>

<span class="c1"># Sort the points by density, so that the densest points are plotted last</span>
<span class="n">z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">hh</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">a</span><span class="o">&lt;=</span><span class="n">locx</span><span class="p">[</span><span class="mi">1</span><span class="p">:]),</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">b</span><span class="o">&lt;=</span><span class="n">locy</span><span class="p">[</span><span class="mi">1</span><span class="p">:])]</span> <span class="k">for</span> <span class="n">a</span><span class="p">,</span><span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">)])</span>
<span class="n">idx</span> <span class="o">=</span> <span class="n">z</span><span class="o">.</span><span class="n">argsort</span><span class="p">()</span>
<span class="n">x2</span><span class="p">,</span> <span class="n">y2</span><span class="p">,</span> <span class="n">z2</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span> <span class="n">y</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span> <span class="n">z</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>


<span class="n">s</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x2</span><span class="p">,</span> <span class="n">y2</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">z2</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Greens&#39;</span><span class="p">)</span>  
<span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;UMI Counts&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Genes Detected&quot;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="o">**</span><span class="mi">5</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_24_0.png" /></p>
<h3 id="examine-the-knee-plot">Examine the knee plot<a class="headerlink" href="#examine-the-knee-plot" title="Permanent link">&para;</a></h3>
<p>The "knee plot" was introduced in the Drop-seq paper: 
- Macosko et al., <a href="https://www.cell.com/fulltext/S0092-8674(15)00549-8">Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets</a>, 2015. DOI:10.1016/j.cell.2015.05.002</p>
<p>In this plot cells are ordered by the number of UMI counts associated to them (shown on the <em>x</em>-axis), and the fraction of droplets with at least that number of cells is shown on the <em>y</em>-axis:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#@title Threshold cells according to knee plot { run: &quot;auto&quot;, vertical-output: true }</span>
<span class="n">expected_num_cells</span> <span class="o">=</span>  <span class="mi">1178</span><span class="c1">#@param {type:&quot;integer&quot;}</span>
<span class="n">knee</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)))</span><span class="o">.</span><span class="n">flatten</span><span class="p">())[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>

<span class="n">ax</span><span class="o">.</span><span class="n">loglog</span><span class="p">(</span><span class="n">knee</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">knee</span><span class="p">)),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">knee</span><span class="p">[</span><span class="n">expected_num_cells</span><span class="p">],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">expected_num_cells</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;UMI Counts&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Set of Barcodes&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s2">&quot;both&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_26_0.png" /></p>
<h2 id="analysis">Analysis<a class="headerlink" href="#analysis" title="Permanent link">&para;</a></h2>
<p>It is useful to examine mitochondrial genes, which are important for quality control. <a href="https://master.bioconductor.org/packages/release/workflows/html/simpleSingleCell.html#examining-gene-level-metrics">(Lun, McCarthy &amp; Marioni, 2017)</a> write that</p>
<blockquote>
<p>High proportions are indicative of poor-quality cells (Islam et al. 2014; Ilicic et al. 2016), possibly because of loss of cytoplasmic RNA from perforated cells. The reasoning is that mitochondria are larger than individual transcript molecules and less likely to escape through tears in the cell membrane.</p>
</blockquote>
<p>Note you can also use the function <code>pp.calculate_qc_metrics</code> to compute the fraction of mitochondrial genes and additional measures.</p>
<p>Show those genes that yield the highest fraction of counts in each single cells, across all cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>normalizing counts per cell
    finished (0:00:00)


/usr/local/lib/python3.7/dist-packages/scanpy/preprocessing/_normalization.py:182: UserWarning: Some cells have zero counts
  warn(UserWarning(&#39;Some cells have zero counts&#39;))
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_31_2.png" /></p>
<h3 id="filter">Filter<a class="headerlink" href="#filter" title="Permanent link">&para;</a></h3>
<p>Begin by filtering cells according to various criteria. First, a filter for genes and cells based on minimum thresholds:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Removes cells with less than 1070 umi counts</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1070</span><span class="p">]</span>

<span class="c1"># Removes genes with 0 umi counts</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1180</span> <span class="err">Ã—</span> <span class="mi">31837</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>filtered out 5 cells that have less than 200 genes expressed


Trying to set attribute `.obs` of view, copying.


filtered out 5884 genes that are detected in less than 3 cells
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1175</span> <span class="err">Ã—</span> <span class="mi">25953</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_genes&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;n_cells&#39;</span>
</code></pre></div>

<p>Next, filter by mitochondrial gene content</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">mito_genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;MT-&#39;</span><span class="p">)</span>
<span class="c1"># for each cell compute fraction of counts in mito genes vs. all genes</span>
<span class="c1"># the `.A1` is only necessary as X is sparse (to transform to a dense array after summing)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;percent_mito&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">[:,</span> <span class="n">mito_genes</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
<span class="c1"># add the total counts per cell as observations-annotation to adata</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
</code></pre></div>
</td></tr></table>
<p>Perform a QC check of the counts post-filtering</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;n_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;percent_mito&#39;</span><span class="p">],</span>
             <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>... storing &#39;gene_name&#39; as categorical
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_41_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#examine mitochondrial content </span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;percent_mito&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;n_genes&#39;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_42_0.png" /></p>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_42_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Create a mask to filter out cells with more than 6500 genes, less than 200 genes or less than 0.2 mitochondrial umi counts</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">((</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">n_genes</span> <span class="o">&lt;</span> <span class="mi">6500</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">n_genes</span> <span class="o">&gt;</span> <span class="mi">200</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">percent_mito</span> <span class="o">&lt;</span> <span class="mf">0.2</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">#filter</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span> <span class="p">:]</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1175</span> <span class="err">Ã—</span> <span class="mi">25953</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;percent_mito&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;n_cells&#39;</span>
</code></pre></div>

<h3 id="normalize-counts">Normalize counts<a class="headerlink" href="#normalize-counts" title="Permanent link">&para;</a></h3>
<p>Total-count normalize (library-size correct) the data matrix $\mathbf{X}$ to 10,000 reads per cell, so that counts become comparable among cells.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># normalize counts in each cell to be equal</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>normalizing counts per cell
    finished (0:00:00)


/usr/local/lib/python3.7/dist-packages/scanpy/preprocessing/_normalization.py:155: UserWarning: Revieved a view of an AnnData. Making a copy.
  view_to_actual(adata)
</code></pre></div>

<p>Log the counts</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Replace raw counts with their logarithm</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p>Lets now look at the highest expressed genes after filtering, normalization, and log</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>normalizing counts per cell
    finished (0:00:00)
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_52_1.png" /></p>
<p>Set the <code>.raw</code> attribute of AnnData object to the normalized and logarithmized raw gene expression for later use in differential testing and visualizations of gene expression. This simply freezes the state of the AnnData object.</p>
<p>The unnormalized data is stored in <code>.raw</code>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="alert alert-info">

**Note**
The result of the following highly-variable-genes detection is stored as an annotation in `.var.highly_variable` and auto-detected by PCA and hence, `sc.pp.neighbors` and subsequent manifold/graph tools.

</div>

<h3 id="identify-highly-variable-genes">Identify highly-variable genes.<a class="headerlink" href="#identify-highly-variable-genes" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># flavor=&quot;cell_ranger&quot; is consistent with Seurat and flavor=&quot;suerat&quot; is not consistent with Seurat</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">If</span> <span class="n">you</span> <span class="k">pass</span> <span class="err">`</span><span class="n">n_top_genes</span><span class="err">`</span><span class="p">,</span> <span class="nb">all</span> <span class="n">cutoffs</span> <span class="n">are</span> <span class="n">ignored</span><span class="o">.</span>
<span class="n">extracting</span> <span class="n">highly</span> <span class="n">variable</span> <span class="n">genes</span>


<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.7</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">statsmodels</span><span class="o">/</span><span class="n">tools</span><span class="o">/</span><span class="n">_testing</span><span class="o">.</span><span class="n">py</span><span class="p">:</span><span class="mi">19</span><span class="p">:</span> <span class="ne">FutureWarning</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">testing</span> <span class="ow">is</span> <span class="n">deprecated</span><span class="o">.</span> <span class="n">Use</span> <span class="n">the</span> <span class="n">functions</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">public</span> <span class="n">API</span> <span class="n">at</span> <span class="n">pandas</span><span class="o">.</span><span class="n">testing</span> <span class="n">instead</span><span class="o">.</span>
  <span class="kn">import</span> <span class="nn">pandas.util.testing</span> <span class="k">as</span> <span class="nn">tm</span>


    <span class="n">finished</span> <span class="p">(</span><span class="mi">0</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">01</span><span class="p">)</span>
<span class="o">--&gt;</span> <span class="n">added</span>
    <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="n">boolean</span> <span class="n">vector</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
    <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="nb">float</span> <span class="n">vector</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
    <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="nb">float</span> <span class="n">vector</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
    <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="nb">float</span> <span class="n">vector</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_58_0.png" /></p>
<p>Regress out effects of total counts per cell and the percentage of mitochondrial genes expressed. Scale the data to unit variance.</p>
<p>We do not regress out as per https://github.com/theislab/scanpy/issues/526</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># sc.pp.regress_out(adata, [&#39;n_counts&#39;, &#39;percent_mito&#39;])</span>
</code></pre></div>
</td></tr></table>
<h3 id="scaling-the-data">Scaling the data<a class="headerlink" href="#scaling-the-data" title="Permanent link">&para;</a></h3>
<p>Scale each gene to unit variance. Clip values exceeding standard deviation 10. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>... as `zero_center=True`, sparse input is densified and may lead to large memory consumption
</code></pre></div>

<h3 id="principal-component-analysis">Principal component analysis<a class="headerlink" href="#principal-component-analysis" title="Permanent link">&para;</a></h3>
<p>Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># We perform PCA on just the highly variable genes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">computing</span> <span class="n">PCA</span>
    <span class="n">on</span> <span class="n">highly</span> <span class="n">variable</span> <span class="n">genes</span>
    <span class="n">with</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">50</span>
    <span class="n">finished</span> <span class="p">(</span><span class="mi">0</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">)</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;CST3&#39;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_67_0.png" /></p>
<p>Let us inspect the contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells, e.g. used in the clustering function  <code>sc.tl.leiden()</code> or tSNE <code>sc.tl.tsne()</code>. In our experience, often, a rough estimate of the number of PCs does fine.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_69_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1175</span> <span class="err">Ã—</span> <span class="mi">25953</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;percent_mito&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;n_cells&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;log1p&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
</code></pre></div>

<h3 id="compute-the-neighborhood-graph">Compute the neighborhood graph<a class="headerlink" href="#compute-the-neighborhood-graph" title="Permanent link">&para;</a></h3>
<p>Next we compute the neighborhood graph of cells using the PCA representation of the data matrix. You might simply use default values here. In order to be consistent with Seurat's results, we use the following values.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nv">computing</span> <span class="nv">neighbors</span>
    <span class="nv">using</span> <span class="s1">&#39;</span><span class="s">X_pca</span><span class="s1">&#39;</span> <span class="nv">with</span> <span class="nv">n_pcs</span> <span class="o">=</span> <span class="mi">10</span>


<span class="o">/</span><span class="nv">usr</span><span class="o">/</span><span class="nv">local</span><span class="o">/</span><span class="nv">lib</span><span class="o">/</span><span class="nv">python3</span>.<span class="mi">7</span><span class="o">/</span><span class="nv">dist</span><span class="o">-</span><span class="nv">packages</span><span class="o">/</span><span class="nv">numba</span><span class="o">/</span><span class="nv">np</span><span class="o">/</span><span class="nv">ufunc</span><span class="o">/</span><span class="nv">parallel</span>.<span class="nv">py</span>:<span class="mi">363</span>: <span class="nv">NumbaWarning</span>: <span class="nv">The</span> <span class="nv">TBB</span> <span class="nv">threading</span> <span class="nv">layer</span> <span class="nv">requires</span> <span class="nv">TBB</span> <span class="nv">version</span> <span class="mi">2019</span>.<span class="mi">5</span> <span class="nv">or</span> <span class="nv">later</span> <span class="nv">i</span>.<span class="nv">e</span>., <span class="nv">TBB_INTERFACE_VERSION</span> <span class="o">&gt;=</span> <span class="mi">11005</span>. <span class="nv">Found</span> <span class="nv">TBB_INTERFACE_VERSION</span> <span class="o">=</span> <span class="mi">9107</span>. <span class="nv">The</span> <span class="nv">TBB</span> <span class="nv">threading</span> <span class="nv">layer</span> <span class="nv">is</span> <span class="nv">disabled</span>.
  <span class="nv">warnings</span>.<span class="nv">warn</span><span class="ss">(</span><span class="nv">problem</span><span class="ss">)</span>


    <span class="nv">finished</span>: <span class="nv">added</span> <span class="nv">to</span> `.<span class="nv">uns</span>[<span class="s1">&#39;</span><span class="s">neighbors</span><span class="s1">&#39;</span>]`
    `.<span class="nv">obsp</span>[<span class="s1">&#39;</span><span class="s">distances</span><span class="s1">&#39;</span>]`, <span class="nv">distances</span> <span class="k">for</span> <span class="nv">each</span> <span class="nv">pair</span> <span class="nv">of</span> <span class="nv">neighbors</span>
    `.<span class="nv">obsp</span>[<span class="s1">&#39;</span><span class="s">connectivities</span><span class="s1">&#39;</span>]`, <span class="nv">weighted</span> <span class="nv">adjacency</span> <span class="nv">matrix</span> <span class="ss">(</span><span class="mi">0</span>:<span class="mi">00</span>:<span class="mi">26</span><span class="ss">)</span>
</code></pre></div>

<h3 id="embed-the-neighborhood-graph">Embed the neighborhood graph<a class="headerlink" href="#embed-the-neighborhood-graph" title="Permanent link">&para;</a></h3>
<h3 id="umap">UMAP<a class="headerlink" href="#umap" title="Permanent link">&para;</a></h3>
<p>UMAP (UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction) is a manifold learning technique that can also be used to visualize cells. It was published in:</p>
<ul>
<li>McInnes, Leland, John Healy, and James Melville. "Umap: Uniform manifold approximation and projection for dimension reduction." arXiv preprint arXiv:1802.03426 (2018).</li>
</ul>
<p>We run that to visualize the results:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>tl.paga(adata)
pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph
tl.umap(adata, init_pos=&#39;paga&#39;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>computing UMAP
    finished: added
    &#39;X_umap&#39;, UMAP coordinates (adata.obsm) (0:00:04)
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CST3&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;PPBP&#39;</span><span class="p">])</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_77_0.png" /></p>
<h3 id="cluster-the-neighborhood-graph">Cluster the neighborhood graph<a class="headerlink" href="#cluster-the-neighborhood-graph" title="Permanent link">&para;</a></h3>
<h3 id="clustering">Clustering<a class="headerlink" href="#clustering" title="Permanent link">&para;</a></h3>
<p>There are many algorithms for clustering cells, and while they have been compared in detail in various benchmarks (see e.g., <a href="https://f1000research.com/articles/7-1141/v2">Duo et al. 2018</a>), there is no univerally agreed upon method. Here we demonstrate clustering using <a href="https://en.wikipedia.org/wiki/Louvain_modularity">Louvain clustering</a>, which is a popular method for clustering single-cell RNA-seq data. The method was published in </p>
<ul>
<li>Blondel, Vincent D; Guillaume, Jean-Loup; Lambiotte, Renaud; Lefebvre, Etienne (9 October 2008). "Fast unfolding of communities in large networks". Journal of Statistical Mechanics: Theory and Experiment. 2008 (10): P10008.</li>
</ul>
<p>Note that Louvain clustering directly clusters the neighborhood graph of cells, which we already computed in the previous section.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">louvain</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">resolution</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>running Louvain clustering
    using the &quot;louvain&quot; package of Traag (2017)
    finished: found 8 clusters and added
    &#39;louvain&#39;, the cluster labels (adata.obs, categorical) (0:00:00)
</code></pre></div>

<p>A plot of the clusters is shown below:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;CST3&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">])</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_82_0.png" /></p>
<h3 id="find-marker-genes">Find marker genes<a class="headerlink" href="#find-marker-genes" title="Permanent link">&para;</a></h3>
<p>A key aspect of annotating a cell atlas is identifying "marker genes". These are genes specific to individual clusters that "mark" them, and are important both for assigning functions to cell clusters, and for designing downstream experiments to probe activity of clusters. </p>
<p>A gene marker analysis begins with ranking genes in each cluster according to how different they are relative to other clusters. Typically the t-test is used for this purpose.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test&#39;</span><span class="p">,</span> <span class="n">corr_method</span><span class="o">=</span><span class="s2">&quot;bonferroni&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>ranking genes
    finished: added to `.uns[&#39;rank_genes_groups&#39;]`
    &#39;names&#39;, sorted np.recarray to be indexed by group ids
    &#39;scores&#39;, sorted np.recarray to be indexed by group ids
    &#39;logfoldchanges&#39;, sorted np.recarray to be indexed by group ids
    &#39;pvals&#39;, sorted np.recarray to be indexed by group ids
    &#39;pvals_adj&#39;, sorted np.recarray to be indexed by group ids (0:00:00)
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_85_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># reduce the verbosity</span>
</code></pre></div>
</td></tr></table>
<p>An alternative to the parametric t-test is the non-parametric <a href="https://de.wikipedia.org/wiki/Wilcoxon-Mann-Whitney-Test">Wilcoxon rank-sum (Mann-Whitney-U)</a> test.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilcoxon&#39;</span><span class="p">,</span> <span class="n">corr_method</span><span class="o">=</span><span class="s2">&quot;bonferroni&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>ranking genes
    finished (0:00:03)
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_88_1.png" /></p>
<p>As an alternative, genes can be ranked using logistic regression (see <a href="https://doi.org/10.1101/258566">Natranos et al. (2018)</a>).</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">tmp</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;logreg&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">ranking</span> <span class="n">genes</span>
    <span class="n">finished</span> <span class="p">(</span><span class="mh">0</span><span class="o">:</span><span class="mh">00</span><span class="o">:</span><span class="mh">18</span><span class="p">)</span>


<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.7</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">sklearn</span><span class="o">/</span><span class="n">linear_model</span><span class="o">/</span><span class="n">_logistic</span><span class="p">.</span><span class="nl">py:</span><span class="mh">940</span><span class="o">:</span> <span class="nl">ConvergenceWarning:</span> <span class="n">lbfgs</span> <span class="n">failed</span> <span class="n">to</span> <span class="n">converge</span> <span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="mh">1</span><span class="p">)</span><span class="o">:</span>
<span class="nl">STOP:</span> <span class="n">TOTAL</span> <span class="n">NO</span><span class="p">.</span> <span class="n">of</span> <span class="n">ITERATIONS</span> <span class="n">REACHED</span> <span class="n">LIMIT</span><span class="p">.</span>

<span class="n">Increase</span> <span class="n">the</span> <span class="n">number</span> <span class="n">of</span> <span class="n">iterations</span> <span class="p">(</span><span class="n">max_iter</span><span class="p">)</span> <span class="k">or</span> <span class="n">scale</span> <span class="n">the</span> <span class="n">data</span> <span class="n">as</span> <span class="n">shown</span> <span class="nl">in:</span>
    <span class="nl">https:</span><span class="c1">//scikit-learn.org/stable/modules/preprocessing.html</span>
<span class="n">Please</span> <span class="n">also</span> <span class="n">refer</span> <span class="n">to</span> <span class="n">the</span> <span class="n">documentation</span> <span class="k">for</span> <span class="n">alternative</span> <span class="n">solver</span> <span class="nl">options:</span>
    <span class="nl">https:</span><span class="c1">//scikit-learn.org/stable/modules/linear_model.html#logistic-regression</span>
  <span class="n">extra_warning_msg</span><span class="o">=</span><span class="n">_LOGISTIC_SOLVER_CONVERGENCE_MSG</span><span class="p">)</span>
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_91_2.png" /></p>
<p>With the exceptions of <em>IL7R</em>, which is only found by the t-test and <em>FCER1A</em>, which is only found by the other two appraoches, all marker genes are recovered in all approaches.</p>
<table>
<thead>
<tr>
<th>Louvain Group</th>
<th>Markers</th>
<th>Cell Type</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>IL7R</td>
<td>CD4 T cells</td>
</tr>
<tr>
<td>1</td>
<td>CD14, LYZ</td>
<td>CD14+ Monocytes</td>
</tr>
<tr>
<td>2</td>
<td>MS4A1</td>
<td>B cells</td>
</tr>
<tr>
<td>3</td>
<td>GNLY, NKG7</td>
<td>NK cells</td>
</tr>
<tr>
<td>4</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Monocytes</td>
</tr>
<tr>
<td>5</td>
<td>CD8A</td>
<td>CD8 T cells</td>
</tr>
<tr>
<td>6</td>
<td>MS4A1</td>
<td>B cells</td>
</tr>
</tbody>
</table>
<p>Let us also define a list of marker genes for later reference.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">marker_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79A&#39;</span><span class="p">,</span> <span class="s1">&#39;MS4A1&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8B&#39;</span><span class="p">,</span> <span class="s1">&#39;LYZ&#39;</span><span class="p">,</span> <span class="s1">&#39;CD14&#39;</span><span class="p">,</span>
                <span class="s1">&#39;LGALS3&#39;</span><span class="p">,</span> <span class="s1">&#39;S100A8&#39;</span><span class="p">,</span> <span class="s1">&#39;GNLY&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;KLRB1&#39;</span><span class="p">,</span>  
                <span class="s1">&#39;FCGR3A&#39;</span><span class="p">,</span> <span class="s1">&#39;MS4A7&#39;</span><span class="p">,</span> <span class="s1">&#39;FCER1A&#39;</span><span class="p">,</span> <span class="s1">&#39;CST3&#39;</span><span class="p">,</span> <span class="s1">&#39;PPBP&#39;</span><span class="p">]</span>
</code></pre></div>
</td></tr></table>
<p>Reload the object that has been save with the Wilcoxon Rank-Sum test result.</p>
<p>Show the 10 top ranked genes per cluster 0, 1, ..., 7 in a dataframe.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>S100A8</td>
      <td>IL7R</td>
      <td>RPL32</td>
      <td>IGHD</td>
      <td>NKG7</td>
      <td>CST3</td>
      <td>KLRB1</td>
      <td>CD79A</td>
    </tr>
    <tr>
      <th>1</th>
      <td>S100A9</td>
      <td>IL32</td>
      <td>RPS12</td>
      <td>IGHM</td>
      <td>CTSW</td>
      <td>HLA-DPA1</td>
      <td>KLRG1</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VCAN</td>
      <td>TRAC</td>
      <td>RPL30</td>
      <td>CD37</td>
      <td>GZMA</td>
      <td>NPC2</td>
      <td>NKG7</td>
      <td>CD79B</td>
    </tr>
    <tr>
      <th>3</th>
      <td>S100A12</td>
      <td>LDHB</td>
      <td>RPL35A</td>
      <td>CD79A</td>
      <td>CST7</td>
      <td>HLA-DPB1</td>
      <td>IL32</td>
      <td>BANK1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FOS</td>
      <td>BCL11B</td>
      <td>RPS15A</td>
      <td>MS4A1</td>
      <td>GZMM</td>
      <td>FCER1G</td>
      <td>GZMA</td>
      <td>TNFRSF13C</td>
    </tr>
  </tbody>
</table>
</div>

<p>Get a table with the scores and groups.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">result</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">]</span>
<span class="n">groups</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span>
<span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span><span class="n">group</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">key</span><span class="p">[:</span><span class="mi">1</span><span class="p">]:</span> <span class="n">result</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">group</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">,</span> <span class="s1">&#39;pvals&#39;</span><span class="p">]})</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0_n</th>
      <th>0_p</th>
      <th>1_n</th>
      <th>1_p</th>
      <th>2_n</th>
      <th>2_p</th>
      <th>3_n</th>
      <th>3_p</th>
      <th>4_n</th>
      <th>4_p</th>
      <th>5_n</th>
      <th>5_p</th>
      <th>6_n</th>
      <th>6_p</th>
      <th>7_n</th>
      <th>7_p</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>S100A8</td>
      <td>4.325218e-141</td>
      <td>IL7R</td>
      <td>3.851653e-57</td>
      <td>RPL32</td>
      <td>7.151780e-62</td>
      <td>IGHD</td>
      <td>1.303809e-75</td>
      <td>NKG7</td>
      <td>9.495682e-64</td>
      <td>CST3</td>
      <td>1.595268e-42</td>
      <td>KLRB1</td>
      <td>8.274802e-41</td>
      <td>CD79A</td>
      <td>3.916006e-28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>S100A9</td>
      <td>1.357507e-140</td>
      <td>IL32</td>
      <td>3.390931e-49</td>
      <td>RPS12</td>
      <td>4.320062e-61</td>
      <td>IGHM</td>
      <td>1.624935e-74</td>
      <td>CTSW</td>
      <td>1.281002e-58</td>
      <td>HLA-DPA1</td>
      <td>9.721044e-42</td>
      <td>KLRG1</td>
      <td>1.245774e-34</td>
      <td>IGKC</td>
      <td>7.115329e-27</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VCAN</td>
      <td>1.698042e-136</td>
      <td>TRAC</td>
      <td>5.952130e-49</td>
      <td>RPL30</td>
      <td>2.823714e-59</td>
      <td>CD37</td>
      <td>1.704719e-73</td>
      <td>GZMA</td>
      <td>5.700633e-57</td>
      <td>NPC2</td>
      <td>5.443673e-33</td>
      <td>NKG7</td>
      <td>7.556990e-32</td>
      <td>CD79B</td>
      <td>6.053017e-26</td>
    </tr>
    <tr>
      <th>3</th>
      <td>S100A12</td>
      <td>2.736086e-136</td>
      <td>LDHB</td>
      <td>6.487053e-46</td>
      <td>RPL35A</td>
      <td>6.098855e-58</td>
      <td>CD79A</td>
      <td>2.184785e-70</td>
      <td>CST7</td>
      <td>1.401230e-56</td>
      <td>HLA-DPB1</td>
      <td>7.522829e-33</td>
      <td>IL32</td>
      <td>1.135243e-30</td>
      <td>BANK1</td>
      <td>6.905314e-26</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FOS</td>
      <td>3.264834e-132</td>
      <td>BCL11B</td>
      <td>4.040622e-42</td>
      <td>RPS15A</td>
      <td>6.949561e-58</td>
      <td>MS4A1</td>
      <td>6.806009e-70</td>
      <td>GZMM</td>
      <td>5.222455e-51</td>
      <td>FCER1G</td>
      <td>3.106843e-32</td>
      <td>GZMA</td>
      <td>1.885113e-30</td>
      <td>TNFRSF13C</td>
      <td>7.946763e-24</td>
    </tr>
  </tbody>
</table>
</div>

<p>Compare to a single cluster. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">],</span> <span class="n">reference</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilcoxon&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">],</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code>ranking genes
    finished (0:00:00)
</code></pre></div>

<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_101_1.png" /></p>
<p>If we want a more detailed view for a certain group, use <code>sc.pl.rank_genes_groups_violin</code>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_103_0.png" /></p>
<p>If you want to compare a certain gene across groups, use the following.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;louvain&#39;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_105_0.png" /></p>
<p>Actually mark the cell types.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>S100A8</td>
      <td>IL7R</td>
      <td>RPL32</td>
      <td>IGHD</td>
      <td>NKG7</td>
      <td>CST3</td>
      <td>KLRB1</td>
      <td>CD79A</td>
    </tr>
    <tr>
      <th>1</th>
      <td>S100A9</td>
      <td>IL32</td>
      <td>RPS12</td>
      <td>IGHM</td>
      <td>CTSW</td>
      <td>HLA-DPA1</td>
      <td>KLRG1</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VCAN</td>
      <td>TRAC</td>
      <td>RPL30</td>
      <td>CD37</td>
      <td>GZMA</td>
      <td>NPC2</td>
      <td>NKG7</td>
      <td>CD79B</td>
    </tr>
    <tr>
      <th>3</th>
      <td>S100A12</td>
      <td>LDHB</td>
      <td>RPL35A</td>
      <td>CD79A</td>
      <td>CST7</td>
      <td>HLA-DPB1</td>
      <td>IL32</td>
      <td>BANK1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>FOS</td>
      <td>BCL11B</td>
      <td>RPS15A</td>
      <td>MS4A1</td>
      <td>GZMM</td>
      <td>FCER1G</td>
      <td>GZMA</td>
      <td>TNFRSF13C</td>
    </tr>
    <tr>
      <th>5</th>
      <td>MNDA</td>
      <td>CD3D</td>
      <td>RPS6</td>
      <td>CD74</td>
      <td>PRF1</td>
      <td>COTL1</td>
      <td>CTSW</td>
      <td>MS4A1</td>
    </tr>
    <tr>
      <th>6</th>
      <td>LYZ</td>
      <td>CD3E</td>
      <td>RPS14</td>
      <td>TCL1A</td>
      <td>HCST</td>
      <td>AIF1</td>
      <td>CCL5</td>
      <td>TCF4</td>
    </tr>
    <tr>
      <th>7</th>
      <td>SRGN</td>
      <td>CD3G</td>
      <td>RPS3A</td>
      <td>CD79B</td>
      <td>KLRD1</td>
      <td>CTSZ</td>
      <td>IL7R</td>
      <td>CD74</td>
    </tr>
    <tr>
      <th>8</th>
      <td>FCN1</td>
      <td>TCF7</td>
      <td>RPS28</td>
      <td>LINC00926</td>
      <td>CCL5</td>
      <td>HLA-DRA</td>
      <td>AQP3</td>
      <td>HLA-DQA1</td>
    </tr>
    <tr>
      <th>9</th>
      <td>DUSP1</td>
      <td>CD2</td>
      <td>RPS27</td>
      <td>BCL11A</td>
      <td>B2M</td>
      <td>HLA-DRB1</td>
      <td>SLC4A10</td>
      <td>LINC00926</td>
    </tr>
  </tbody>
</table>
</div>

<p>Recall</p>
<table>
<thead>
<tr>
<th>Louvain Group</th>
<th>Markers</th>
<th>Cell Type</th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td>IL7R</td>
<td>CD4 T cells</td>
</tr>
<tr>
<td>1</td>
<td>CD14, LYZ</td>
<td>CD14+ Monocytes</td>
</tr>
<tr>
<td>2</td>
<td>MS4A1</td>
<td>B cells</td>
</tr>
<tr>
<td>3</td>
<td>GNLY, NKG7</td>
<td>NK cells</td>
</tr>
<tr>
<td>4</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Monocytes</td>
</tr>
<tr>
<td>5</td>
<td>CD8A</td>
<td>CD8 T cells</td>
</tr>
<tr>
<td>6</td>
<td>MS4A1</td>
<td>B cells</td>
</tr>
</tbody>
</table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1">## Map the gene ids to a cluster label</span>
<span class="c1"># Each cluster (index in the list) corresponds to a cell type</span>
<span class="n">new_cluster_names</span> <span class="o">=</span> <span class="p">[</span>
                     <span class="s2">&quot;CD4 T&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;B Cells&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;NK&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;CD14 Monocytes&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;FCGR3A Monocytes&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;CD8 T&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;B-2&quot;</span><span class="p">,</span>
<span class="p">]</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Relabel the clusters</span>
<span class="c1"># adata.rename_categories(&#39;louvain&#39;, new_cluster_names)</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_111_0.png" /></p>
<p>Now that we annotated the cell types, let us visualize the marker genes.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">ax</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;louvain&#39;</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_113_0.png" /></p>
<p>There is also a very compact violin plot.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">ax</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stacked_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../kb_analysis_0_python_files/kb_analysis_0_python_115_0.png" /></p>
<p>Note that as a result of the analysis the adata object has accumulated several annotations:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="n">adata</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1175</span> <span class="err">Ã—</span> <span class="mi">25953</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;percent_mito&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;n_cells&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;log1p&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;distances&#39;</span><span class="p">,</span> <span class="s1">&#39;connectivities&#39;</span>
</code></pre></div>

<p>If you want to export to "csv", you have the following options:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Export single fields of the annotation of observations</span>
<span class="c1"># adata.obs[[&#39;n_counts&#39;, &#39;louvain_groups&#39;]].to_csv(</span>
<span class="c1">#     &#39;./write/pbmc3k_corrected_louvain_groups.csv&#39;)</span>

<span class="c1"># Export single columns of the multidimensional annotation</span>
<span class="c1"># adata.obsm.to_df()[[&#39;X_pca1&#39;, &#39;X_pca2&#39;]].to_csv(</span>
<span class="c1">#     &#39;./write/pbmc3k_corrected_X_pca.csv&#39;)</span>

<span class="c1"># Or export everything except the data using `.write_csvs`.</span>
<span class="c1"># Set `skip_data=False` if you also want to export the data.</span>
<span class="c1"># adata.write_csvs(results_file[:-5], )</span>
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code><span class="c1"># Running time of the notebook</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{:.2f}</span><span class="s2"> minutes&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">((</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span><span class="o">-</span><span class="n">start_time</span><span class="p">)</span><span class="o">/</span><span class="mi">60</span><span class="p">))</span>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="mf">27.63</span> <span class="n">minutes</span>
</code></pre></div>

<p><strong>Feedback</strong>: please report any issues, or submit pull requests for improvements, in the <a href="https://github.com/pachterlab/kallistobustools/">Github repository where this notebook is located</a>.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>
</code></pre></div>
</td></tr></table>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../../../kb_getting_started/R/kb_intro_2_R/" class="md-footer__link md-footer__link--prev" aria-label="Previous: R" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              R
            </div>
          </div>
        </a>
      
      
        
        <a href="../../R/kb_analysis_0_R/" class="md-footer__link md-footer__link--next" aria-label="Next: R" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              R
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2021 - 2021 Pachterlab
          </div>
        
        
          Made with
          <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
            Material for MkDocs
          </a>
        
        
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/pachterlab" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
      
      
        
        
      
      <a href="https://twitter.com/sinabooeshaghi" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../../..", "features": ["search.suggest", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../../assets/javascripts/workers/search.f8263e09.min.js", "version": null}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.4fc53ad4.min.js"></script>
      
    
  </body>
</html>